
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Analyze simulated library of codon variants &#8212; dms_variants 0.7.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Libraries with secondary target genes" href="codonvariant_sim_data_multi_targets.html" />
    <link rel="prev" title="Examples" href="examples.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput,
div.nbinput div.prompt,
div.nbinput div.input_area,
div.nbinput div[class*=highlight],
div.nbinput div[class*=highlight] pre,
div.nboutput,
div.nbinput div.prompt,
div.nbinput div.output_area,
div.nboutput div[class*=highlight],
div.nboutput div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput,
div.nboutput {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput,
    div.nboutput {
        flex-direction: column;
    }
}

/* input container */
div.nbinput {
    padding-top: 5px;
}

/* last container */
div.nblast {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput div.prompt,
div.nboutput div.prompt {
    min-width: 5ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput div.prompt,
    div.nboutput div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput div.prompt pre,
div.nboutput div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput div.input_area,
div.nboutput div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput div.input_area,
    div.nboutput div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput div.math p {
    text-align: left;
}

/* standard error */
div.nboutput div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Analyze-simulated-library-of-codon-variants">
<h1>Analyze simulated library of codon variants<a class="headerlink" href="#Analyze-simulated-library-of-codon-variants" title="Permalink to this headline">¶</a></h1>
<p>This example does the following three major things:</p>
<ol class="arabic simple">
<li><p>Simulates a deep mutational scanning experiment on a “plausible” data set of barcoded codon variants.</p></li>
<li><p>Analyze the simulated experiment by examining the variant counts and functional scores.</p></li>
<li><p>Fits global epistasis models.</p></li>
</ol>
<div class="section" id="Setup-for-analysis">
<h2>Setup for analysis<a class="headerlink" href="#Setup-for-analysis" title="Permalink to this headline">¶</a></h2>
<p>Import Python modules / packages:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">plotnine</span> <span class="kn">import</span> <span class="o">*</span>

<span class="kn">import</span> <span class="nn">scipy</span>

<span class="kn">import</span> <span class="nn">dmslogo</span>  <span class="c1"># used for preference logo plots</span>

<span class="kn">import</span> <span class="nn">dms_variants.binarymap</span>
<span class="kn">import</span> <span class="nn">dms_variants.codonvarianttable</span>
<span class="kn">import</span> <span class="nn">dms_variants.globalepistasis</span>
<span class="kn">import</span> <span class="nn">dms_variants.plotnine_themes</span>
<span class="kn">import</span> <span class="nn">dms_variants.simulate</span>
<span class="kn">from</span> <span class="nn">dms_variants.constants</span> <span class="kn">import</span> <span class="n">CBPALETTE</span><span class="p">,</span> <span class="n">CODONS_NOSTOP</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/fh/fast/bloom_j/software/conda_v2/envs/BloomLab/lib/python3.6/site-packages/dmslogo/logo.py:40: MatplotlibDeprecationWarning:
The createFontList function was deprecated in Matplotlib 3.2 and will be removed two minor releases later. Use FontManager.addfont instead.
  matplotlib.font_manager.findSystemFonts(_FONT_PATH)))
</pre></div></div>
</div>
<p>Set parameters that define the simulated data. Except for that fact that we use a short gene to reduce the computational run-time of this notebook, these simulation parameters are designed to reflect what we might observe in a real Bloom lab deep mutational scanning experiment:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">seed</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># random number seed</span>
<span class="n">genelength</span> <span class="o">=</span> <span class="mi">30</span>  <span class="c1"># gene length in codons</span>
<span class="n">libs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;lib_1&#39;</span><span class="p">,</span> <span class="s1">&#39;lib_2&#39;</span><span class="p">]</span>  <span class="c1"># distinct libraries of gene</span>
<span class="n">variants_per_lib</span> <span class="o">=</span> <span class="mi">500</span> <span class="o">*</span> <span class="n">genelength</span>  <span class="c1"># variants per library</span>
<span class="n">avgmuts</span> <span class="o">=</span> <span class="mf">2.0</span>  <span class="c1"># average codon mutations per variant</span>
<span class="n">bclen</span> <span class="o">=</span> <span class="mi">16</span>  <span class="c1"># length of nucleotide barcode for each variant</span>
<span class="n">variant_error_rate</span> <span class="o">=</span> <span class="mf">0.005</span>  <span class="c1"># rate at which variant sequence mis-called</span>
<span class="n">avgdepth_per_variant</span> <span class="o">=</span> <span class="mi">200</span>  <span class="c1"># average per-variant sequencing depth</span>
<span class="n">lib_uniformity</span> <span class="o">=</span> <span class="mi">5</span>  <span class="c1"># uniformity of library pre-selection</span>
<span class="n">noise</span> <span class="o">=</span> <span class="mf">0.02</span>  <span class="c1"># random noise in selections</span>
<span class="n">bottlenecks</span> <span class="o">=</span> <span class="p">{</span>  <span class="c1"># bottlenecks from pre- to post-selection</span>
        <span class="s1">&#39;tight_bottle&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="mi">5</span><span class="p">,</span>
        <span class="s1">&#39;loose_bottle&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
        <span class="p">}</span>
</pre></div>
</div>
</div>
<p>Seed random number generator for reproducible output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Set pandas display options to show large chunks of Data Frames in this example:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.width&#39;</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Hide warnings that clutter output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Set the gray grid <a class="reference external" href="https://plotnine.readthedocs.io/en/stable/generated/plotnine.themes.theme.html?highlight=themes">plotnine theme</a> defined in <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.plotnine_themes.html">dms_variants.plotnine_themes</a> as this gives a nice appearance for the plots:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">theme_set</span><span class="p">(</span><span class="n">dms_variants</span><span class="o">.</span><span class="n">plotnine_themes</span><span class="o">.</span><span class="n">theme_graygrid</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Simulate-library-of-variants">
<h2>Simulate library of variants<a class="headerlink" href="#Simulate-library-of-variants" title="Permalink to this headline">¶</a></h2>
<p>Simulate wildtype gene sequence:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">geneseq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">CODONS_NOSTOP</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">genelength</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wildtype gene of </span><span class="si">{genelength}</span><span class="s2"> codons:</span><span class="se">\n</span><span class="si">{geneseq}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Wildtype gene of 30 codons:
AGATCCGTGATTCTGCGTGCTTACACCAACTCACGGGTGAAACGTGTAATCTTATGCAACAACGACTTACCTATCCGCAACATCCGGCTG
</pre></div></div>
</div>
<p>Generate a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a> using <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulate_CodonVariantTable">simulate_CodonVariantTable</a> function:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulate_CodonVariantTable</span><span class="p">(</span>
                <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">,</span>
                <span class="n">bclen</span><span class="o">=</span><span class="n">bclen</span><span class="p">,</span>
                <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span><span class="n">lib</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;avgmuts&#39;</span><span class="p">:</span> <span class="n">avgmuts</span><span class="p">,</span>
                                     <span class="s1">&#39;nvariants&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">}</span>
                               <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span><span class="p">},</span>
                <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
                <span class="p">)</span>
</pre></div>
</div>
</div>
<p>We can get basic information about the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a>, such as the sites, wildtype codons, and wildtype amino acids. Below we do this for the first few sites:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">sites</span><span class="p">[:</span> <span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[1, 2, 3, 4, 5]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">list</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">sites</span><span class="p">[:</span> <span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[&#39;AGA&#39;, &#39;TCC&#39;, &#39;GTG&#39;, &#39;ATT&#39;, &#39;CTG&#39;]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">list</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">aas</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">sites</span><span class="p">[:</span> <span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[&#39;R&#39;, &#39;S&#39;, &#39;V&#39;, &#39;I&#39;, &#39;L&#39;]
</pre></div>
</div>
</div>
<p>The different libraries in the table:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area highlight-none notranslate"><div class="highlight"><pre>
<span></span>[&#39;lib_1&#39;, &#39;lib_2&#39;]
</pre></div>
</div>
</div>
<p>Here is the data frame that contains all of the information on the barcoded variants (just looking at the first few lines)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>variant_call_support</th>
      <th>codon_substitutions</th>
      <th>aa_substitutions</th>
      <th>n_codon_substitutions</th>
      <th>n_aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>AAAAAAAACGTTTTGT</td>
      <td>1</td>
      <td>CTG5TGG TCA11TCT</td>
      <td>L5W</td>
      <td>2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>AAAAAAAGGCTTATAC</td>
      <td>5</td>
      <td>TCA11TGC</td>
      <td>S11C</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>AAAAAAATCGTCCGTG</td>
      <td>2</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>AAAAAAGACAACACCG</td>
      <td>5</td>
      <td>ATC25TTC</td>
      <td>I25F</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>AAAAAAGATGTGGTGG</td>
      <td>3</td>
      <td>TCA11TTA CGG12GGT CGT15TGA ATC25GGA ATC28CGC</td>
      <td>S11L R12G R15* I25G I28R</td>
      <td>5</td>
      <td>5</td>
    </tr>
    <tr>
      <th>5</th>
      <td>lib_1</td>
      <td>AAAAAAGTCATACTCG</td>
      <td>1</td>
      <td>CGG12TCA TGC19ACC AAC27CGC</td>
      <td>R12S C19T N27R</td>
      <td>3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Analyze-library-composition">
<h2>Analyze library composition<a class="headerlink" href="#Analyze-library-composition" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a> has methods for summarizing and plotting properties of the barcoded variants. These methods can be used to analyze the barcoded variants themselves, or to analyze the frequency (or counts) of variants in the library in samples that have undergone some type of selection.</p>
<p>We have not yet added counts of the variants in any specific samples, so we just analyze the composition of the variant library itself. This is done by setting <code class="docutils literal notranslate"><span class="pre">samples=None</span></code> in the method calls below.</p>
<p>Number of variants in each library:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>15000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>15000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>30000</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Plot distribution of variant call supports, grouping together all variants with support <span class="math notranslate nohighlight">\(\ge 8\)</span>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotVariantSupportHistogram</span><span class="p">(</span><span class="n">max_support</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_27_0.png" src="_images/codonvariant_sim_data_27_0.png" />
</div>
</div>
<p>The set of valid barcodes for each library:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;First few barcodes for library </span><span class="si">{lib}</span><span class="s2">:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">(</span><span class="n">lib</span><span class="p">))[</span> <span class="p">:</span> <span class="mi">3</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
First few barcodes for library lib_1:
[&#39;AAAAAAAACGTTTTGT&#39;, &#39;AAAAAAAGGCTTATAC&#39;, &#39;AAAAAAATCGTCCGTG&#39;]
First few barcodes for library lib_2:
[&#39;AAAAAACCCAGACTTA&#39;, &#39;AAAAAACCGGCGGCAG&#39;, &#39;AAAAAATCTACTGCGT&#39;]
</pre></div></div>
</div>
<p>Plot the number of amino-acid mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_31_0.png" src="_images/codonvariant_sim_data_31_0.png" />
</div>
</div>
<p>Number of codon mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_33_0.png" src="_images/codonvariant_sim_data_33_0.png" />
</div>
</div>
<p>Average number of codon mutations per variant of each type of mutation. We make these plots for: 1. Just single-mutant and wildtype variants 2. For all variants</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">mut_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="n">mut_type</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_35_0.png" src="_images/codonvariant_sim_data_35_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_35_1.png" src="_images/codonvariant_sim_data_35_1.png" />
</div>
</div>
<p>Here are the numerical data in the plots above:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">numCodonMutsByType</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>mutation_type</th>
      <th>num_muts_count</th>
      <th>count</th>
      <th>number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>nonsynonymous</td>
      <td>3615</td>
      <td>6112</td>
      <td>0.591</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>synonymous</td>
      <td>194</td>
      <td>6112</td>
      <td>0.032</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>stop</td>
      <td>211</td>
      <td>6112</td>
      <td>0.035</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>nonsynonymous</td>
      <td>3716</td>
      <td>6114</td>
      <td>0.608</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>synonymous</td>
      <td>235</td>
      <td>6114</td>
      <td>0.038</td>
    </tr>
    <tr>
      <th>5</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>stop</td>
      <td>179</td>
      <td>6114</td>
      <td>0.029</td>
    </tr>
    <tr>
      <th>6</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>nonsynonymous</td>
      <td>7331</td>
      <td>12226</td>
      <td>0.600</td>
    </tr>
    <tr>
      <th>7</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>synonymous</td>
      <td>429</td>
      <td>12226</td>
      <td>0.035</td>
    </tr>
    <tr>
      <th>8</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>stop</td>
      <td>390</td>
      <td>12226</td>
      <td>0.032</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">numCodonMutsByType</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>mutation_type</th>
      <th>num_muts_count</th>
      <th>count</th>
      <th>number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>nonsynonymous</td>
      <td>26916</td>
      <td>15000</td>
      <td>1.794</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>synonymous</td>
      <td>1510</td>
      <td>15000</td>
      <td>0.101</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>stop</td>
      <td>1424</td>
      <td>15000</td>
      <td>0.095</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>nonsynonymous</td>
      <td>26960</td>
      <td>15000</td>
      <td>1.797</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>synonymous</td>
      <td>1498</td>
      <td>15000</td>
      <td>0.100</td>
    </tr>
    <tr>
      <th>5</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>stop</td>
      <td>1371</td>
      <td>15000</td>
      <td>0.091</td>
    </tr>
    <tr>
      <th>6</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>nonsynonymous</td>
      <td>53876</td>
      <td>30000</td>
      <td>1.796</td>
    </tr>
    <tr>
      <th>7</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>synonymous</td>
      <td>3008</td>
      <td>30000</td>
      <td>0.100</td>
    </tr>
    <tr>
      <th>8</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>stop</td>
      <td>2795</td>
      <td>30000</td>
      <td>0.093</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Examine how well amino-acid mutations are sampled in the library by looking at the fraction of mutations seen &lt;= some number of times. Here we do that for amino- acid mutations, making separate plots for single mutants and all mutants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">variant_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulMutCoverage</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span>
                                      <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span>
                                      <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_40_0.png" src="_images/codonvariant_sim_data_40_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_40_1.png" src="_images/codonvariant_sim_data_40_1.png" />
</div>
</div>
<p>We can also get the numerical information plotted above (here for single mutants only):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>mutation</th>
      <th>count</th>
      <th>mutation_type</th>
      <th>site</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>V16L</td>
      <td>25</td>
      <td>nonsynonymous</td>
      <td>16</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>C19R</td>
      <td>21</td>
      <td>nonsynonymous</td>
      <td>19</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>R12S</td>
      <td>21</td>
      <td>nonsynonymous</td>
      <td>12</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>R26L</td>
      <td>21</td>
      <td>nonsynonymous</td>
      <td>26</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>S11R</td>
      <td>21</td>
      <td>nonsynonymous</td>
      <td>11</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Here are the frequencies of mutations along the gene, looking both at single mutants and all mutants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">variant_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotMutFreqs</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span>
                              <span class="s1">&#39;codon&#39;</span><span class="p">,</span>
                              <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_44_0.png" src="_images/codonvariant_sim_data_44_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_44_1.png" src="_images/codonvariant_sim_data_44_1.png" />
</div>
</div>
<p>We can also look at mutation frequencies in a heat-map form:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotMutHeatmap</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_46_0.png" src="_images/codonvariant_sim_data_46_0.png" />
</div>
</div>
</div>
<div class="section" id="Simulate-counts-for-samples">
<h2>Simulate counts for samples<a class="headerlink" href="#Simulate-counts-for-samples" title="Permalink to this headline">¶</a></h2>
<p>An experiment involves subjecting the library to different selections and looking at how the frequencies of the variants changes by using sequencing to count the barcodes in each condition.</p>
<p>Here, we simulate an experiment by simulating variant counts for samples that have undergone various selections.</p>
<p>For these simulations, we first need to define a “phenotype” for each variant. The phenotype represents how much the frequency of the variant is expected to increase or decrease after selection.</p>
<div class="section" id="Define-phenotype-function">
<h3>Define phenotype function<a class="headerlink" href="#Define-phenotype-function" title="Permalink to this headline">¶</a></h3>
<p>First, we define a “phenotype” function. We will do this using a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator">SigmoidPhenotypeSimulator</a>, which follow the “global epistasis” concepts of <a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al</a> and <a class="reference external" href="https://doi.org/10.1534/genetics.116.195214">Sailer and Harms</a> to define the phenotype in two steps: an underlying latent phenotype that mutations affect additively,
and then an observed phenotype that is a non-linear function of the latent phenotype. The variants are then simulated according to their observed enrichments, which are the exponentials of the observed phenotypes.</p>
<p>First, we initialize the simulator:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">phenosimulator</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">SigmoidPhenotypeSimulator</span><span class="p">(</span>
                                            <span class="n">geneseq</span><span class="p">,</span>
                                            <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the simulated relationship of the latent phenotype with the observed enrichment and phenotype, with a dashed vertical line indicating the wildtype latent phenotype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotLatentVsObserved</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_50_0.png" src="_images/codonvariant_sim_data_50_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_50_1.png" src="_images/codonvariant_sim_data_50_1.png" />
</div>
</div>
<p>Plot the latent phenotype, observed phenotype, and observed enrichment of all single amino-acid mutants, with a dashed vertical line indicating the wildtype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;latentPhenotype&#39;</span><span class="p">,</span> <span class="s1">&#39;observedPhenotype&#39;</span><span class="p">,</span> <span class="s1">&#39;observedEnrichment&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotMutsHistogram</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_52_0.png" src="_images/codonvariant_sim_data_52_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_52_1.png" src="_images/codonvariant_sim_data_52_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_52_2.png" src="_images/codonvariant_sim_data_52_2.png" />
</div>
</div>
</div>
<div class="section" id="Simulate-variant-counts">
<h3>Simulate variant counts<a class="headerlink" href="#Simulate-variant-counts" title="Permalink to this headline">¶</a></h3>
<p>Now we use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulateSampleCounts">simulateSampleCounts</a> to simulate counts of variants when selection on each variant is proportional to its observed enrichment. In these simulations, we can fine-tune the simulations to reflect real experiments, such as by setting the error-rate in variant calling, bottlenecks when going from the pre- to post-selection samples, and non-uniformity in library composition.</p>
<p>Here we simulate using several bottlenecks in the library going from the pre- to post-selection samples, since in real experiments this seems to be the biggest source of noise / error:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">counts</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulateSampleCounts</span><span class="p">(</span>
        <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
        <span class="n">phenotype_func</span><span class="o">=</span><span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedEnrichment</span><span class="p">,</span>
        <span class="n">variant_error_rate</span><span class="o">=</span><span class="n">variant_error_rate</span><span class="p">,</span>
        <span class="n">pre_sample</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;total_count&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span>
                        <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">avgdepth_per_variant</span><span class="p">),</span>
                    <span class="s1">&#39;uniformity&#39;</span><span class="p">:</span> <span class="n">lib_uniformity</span><span class="p">},</span>
        <span class="n">pre_sample_name</span><span class="o">=</span><span class="s1">&#39;pre-selection&#39;</span><span class="p">,</span>
        <span class="n">post_samples</span><span class="o">=</span><span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;noise&#39;</span><span class="p">:</span> <span class="n">noise</span><span class="p">,</span>
                             <span class="s1">&#39;total_count&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span>
                                 <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">avgdepth_per_variant</span><span class="p">),</span>
                             <span class="s1">&#39;bottleneck&#39;</span><span class="p">:</span> <span class="n">bottle</span><span class="p">}</span>
                          <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">bottle</span> <span class="ow">in</span> <span class="n">bottlenecks</span><span class="o">.</span><span class="n">items</span><span class="p">()},</span>
        <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
        <span class="p">)</span>
</pre></div>
</div>
</div>
<p>First few lines of the data frame with the simulated counts:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>AAAAAAAACGTTTTGT</td>
      <td>pre-selection</td>
      <td>344</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>AAAAAAAGGCTTATAC</td>
      <td>pre-selection</td>
      <td>120</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>AAAAAAATCGTCCGTG</td>
      <td>pre-selection</td>
      <td>138</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>AAAAAAGACAACACCG</td>
      <td>pre-selection</td>
      <td>108</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>AAAAAAGATGTGGTGG</td>
      <td>pre-selection</td>
      <td>352</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Add-counts-to-variant-table">
<h3>Add counts to variant table<a class="headerlink" href="#Add-counts-to-variant-table" title="Permalink to this headline">¶</a></h3>
<p>Now add the simulated counts for each library / sample to the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">add_sample_counts_df</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Confirm that we have added the expected number of counts per library / sample:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>2760000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>2820000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>3105000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_2</td>
      <td>pre-selection</td>
      <td>2760000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_2</td>
      <td>loose_bottle</td>
      <td>2820000</td>
    </tr>
    <tr>
      <th>5</th>
      <td>lib_2</td>
      <td>tight_bottle</td>
      <td>3105000</td>
    </tr>
    <tr>
      <th>6</th>
      <td>all libraries</td>
      <td>pre-selection</td>
      <td>5520000</td>
    </tr>
    <tr>
      <th>7</th>
      <td>all libraries</td>
      <td>loose_bottle</td>
      <td>5640000</td>
    </tr>
    <tr>
      <th>8</th>
      <td>all libraries</td>
      <td>tight_bottle</td>
      <td>6210000</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
</div>
<div class="section" id="Analyze-sample-variant-counts">
<h2>Analyze sample variant counts<a class="headerlink" href="#Analyze-sample-variant-counts" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a> has methods for summarizing and plotting the variant counts for different samples. These methods are mostly the same as we used above to analyze the variant composition of the libraries themselves, but now we set <code class="docutils literal notranslate"><span class="pre">samples</span></code> to the samples that we want to analyze (typically <code class="docutils literal notranslate"><span class="pre">'all'</span></code>). Therefore, rather than each variant always counting once,
each variant is counted in each sample in proportion to how many counts it has.</p>
<p>In the rawest form, we can directly access a data frame giving the counts of each variant in each sample:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>barcode</th>
      <th>count</th>
      <th>variant_call_support</th>
      <th>codon_substitutions</th>
      <th>aa_substitutions</th>
      <th>n_codon_substitutions</th>
      <th>n_aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>AACCCGTTCACCACCA</td>
      <td>722</td>
      <td>2</td>
      <td>ATC17CGA CGG29TGC</td>
      <td>I17R R29C</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>CGCAAAAGTCTATTAC</td>
      <td>646</td>
      <td>1</td>
      <td>TCA11GCG</td>
      <td>S11A</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>TCCATTTGAATTGTTT</td>
      <td>640</td>
      <td>1</td>
      <td>CGG12CCA CGT15CCG</td>
      <td>R12P R15P</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>CCAGAAAACCCACCAA</td>
      <td>623</td>
      <td>7</td>
      <td>TTA18CTC</td>
      <td></td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>TTTGCGCGACTTTTAG</td>
      <td>610</td>
      <td>1</td>
      <td>AAC21CAT ATC28TGG</td>
      <td>N21H I28W</td>
      <td>2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>First, we get the <strong>average</strong> number of counts per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">avgCountsPerVariant</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>avg_counts_per_variant</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>184</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>188</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>207</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_2</td>
      <td>pre-selection</td>
      <td>184</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_2</td>
      <td>loose_bottle</td>
      <td>188</td>
    </tr>
    <tr>
      <th>5</th>
      <td>lib_2</td>
      <td>tight_bottle</td>
      <td>207</td>
    </tr>
    <tr>
      <th>6</th>
      <td>all libraries</td>
      <td>pre-selection</td>
      <td>184</td>
    </tr>
    <tr>
      <th>7</th>
      <td>all libraries</td>
      <td>loose_bottle</td>
      <td>188</td>
    </tr>
    <tr>
      <th>8</th>
      <td>all libraries</td>
      <td>tight_bottle</td>
      <td>207</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We also plot the average counts per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotAvgCountsPerVariant</span><span class="p">()</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_66_0.png" src="_images/codonvariant_sim_data_66_0.png" />
</div>
</div>
<p>Plot the number of counts for each variant in each sample. The horizontal dashed line shows the total number of variants. The plot shows that all variants are well-sampled in the pre-selection libraries, but that post- selection some variants are sampled more or less. This is expected since selection will decrease and increase the frequency of variants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulVariantCounts</span><span class="p">()</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_68_0.png" src="_images/codonvariant_sim_data_68_0.png" />
</div>
</div>
<p>An alternative way to see how many any given variant dominates the library. One way to do that is to determine fraction of counts in each library from each variant and then plot how these fractions are distributed among variants.</p>
<p>In the plot below, variants with no counts are shown on the log y-scale below the dotted gray line:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCountsPerVariant</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_70_0.png" src="_images/codonvariant_sim_data_70_0.png" />
</div>
</div>
<p>We can also make a similar plot showing the values for each class of variant separately, where we can see how post-selection some classes of variants drop in counts more than others:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCountsPerVariant</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span><span class="p">,</span>
                                  <span class="n">by_variant_class</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_72_0.png" src="_images/codonvariant_sim_data_72_0.png" />
</div>
</div>
<p>We can also show total counts rather than fraction of counts:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCountsPerVariant</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span><span class="p">,</span>
                                  <span class="n">ystat</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_74_0.png" src="_images/codonvariant_sim_data_74_0.png" />
</div>
</div>
<p>Or plot on a non-log y-scale (although this generally is less informative):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCountsPerVariant</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span><span class="p">,</span>
                                  <span class="n">logy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_76_0.png" src="_images/codonvariant_sim_data_76_0.png" />
</div>
</div>
<p>Distribution of the number of amino-acid mutations per variant in each sample. As expected, mutations go down after selection:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_78_0.png" src="_images/codonvariant_sim_data_78_0.png" />
</div>
</div>
<p>Average number of mutations of each type per variant among just single mutants (and wildtype) and among all variants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">variant_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="n">variant_type</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_80_0.png" src="_images/codonvariant_sim_data_80_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_80_1.png" src="_images/codonvariant_sim_data_80_1.png" />
</div>
</div>
<p>Here are the numerical data plotted above (just for the individual libraries and all mutations):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">variants</span><span class="o">.</span><span class="n">numCodonMutsByType</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>mutation_type</th>
      <th>num_muts_count</th>
      <th>count</th>
      <th>number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>nonsynonymous</td>
      <td>4934898</td>
      <td>2760000</td>
      <td>1.788</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>synonymous</td>
      <td>279387</td>
      <td>2760000</td>
      <td>0.101</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>stop</td>
      <td>260846</td>
      <td>2760000</td>
      <td>0.095</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>nonsynonymous</td>
      <td>2463111</td>
      <td>2820000</td>
      <td>0.873</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>synonymous</td>
      <td>282348</td>
      <td>2820000</td>
      <td>0.100</td>
    </tr>
    <tr>
      <th>5</th>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>stop</td>
      <td>900</td>
      <td>2820000</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>6</th>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>nonsynonymous</td>
      <td>2727707</td>
      <td>3105000</td>
      <td>0.878</td>
    </tr>
    <tr>
      <th>7</th>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>synonymous</td>
      <td>305938</td>
      <td>3105000</td>
      <td>0.099</td>
    </tr>
    <tr>
      <th>8</th>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>stop</td>
      <td>933</td>
      <td>3105000</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>9</th>
      <td>lib_2</td>
      <td>pre-selection</td>
      <td>nonsynonymous</td>
      <td>4964286</td>
      <td>2760000</td>
      <td>1.799</td>
    </tr>
    <tr>
      <th>10</th>
      <td>lib_2</td>
      <td>pre-selection</td>
      <td>synonymous</td>
      <td>279432</td>
      <td>2760000</td>
      <td>0.101</td>
    </tr>
    <tr>
      <th>11</th>
      <td>lib_2</td>
      <td>pre-selection</td>
      <td>stop</td>
      <td>255340</td>
      <td>2760000</td>
      <td>0.093</td>
    </tr>
    <tr>
      <th>12</th>
      <td>lib_2</td>
      <td>loose_bottle</td>
      <td>nonsynonymous</td>
      <td>2553924</td>
      <td>2820000</td>
      <td>0.906</td>
    </tr>
    <tr>
      <th>13</th>
      <td>lib_2</td>
      <td>loose_bottle</td>
      <td>synonymous</td>
      <td>288997</td>
      <td>2820000</td>
      <td>0.102</td>
    </tr>
    <tr>
      <th>14</th>
      <td>lib_2</td>
      <td>loose_bottle</td>
      <td>stop</td>
      <td>998</td>
      <td>2820000</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>15</th>
      <td>lib_2</td>
      <td>tight_bottle</td>
      <td>nonsynonymous</td>
      <td>2782835</td>
      <td>3105000</td>
      <td>0.896</td>
    </tr>
    <tr>
      <th>16</th>
      <td>lib_2</td>
      <td>tight_bottle</td>
      <td>synonymous</td>
      <td>328619</td>
      <td>3105000</td>
      <td>0.106</td>
    </tr>
    <tr>
      <th>17</th>
      <td>lib_2</td>
      <td>tight_bottle</td>
      <td>stop</td>
      <td>1418</td>
      <td>3105000</td>
      <td>0.000</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Here are mutation frequencies as a function of primary sequence among all variants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotMutFreqs</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
                          <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_84_0.png" src="_images/codonvariant_sim_data_84_0.png" />
</div>
</div>
<p>Plot how thoroughly amino-acid mutations are sampled, doing this separately among all variants and single-mutant variants. The plots below show that the stop mutations are sampled very poorly post-selection because they are eliminated during selection:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">variant_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulMutCoverage</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="n">variant_type</span><span class="p">,</span>
                                      <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_86_0.png" src="_images/codonvariant_sim_data_86_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_86_1.png" src="_images/codonvariant_sim_data_86_1.png" />
</div>
</div>
</div>
<div class="section" id="Functional-scores-for-variants">
<h2>Functional scores for variants<a class="headerlink" href="#Functional-scores-for-variants" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.func_scores">CodonVariantTable.func_scores</a> method calculates a functional score for each variant based on its change in frequency from pre- to post-selection.</p>
<p>To calculate these scores, we need to pair each post-selection sample with a pre-selection one. In this case, the pre-selection sample is named ‘pre-selection’ for all post-selection samples:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s1">&#39;pre-selection&#39;</span><span class="p">,</span>
                                   <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The resulting data frame has a functional score (and its variance) for each barcoded variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">func_scores</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>pre_sample</th>
      <th>post_sample</th>
      <th>barcode</th>
      <th>func_score</th>
      <th>func_score_var</th>
      <th>pre_count</th>
      <th>post_count</th>
      <th>pre_count_wt</th>
      <th>post_count_wt</th>
      <th>pseudocount</th>
      <th>codon_substitutions</th>
      <th>n_codon_substitutions</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>AACCCGTTCACCACCA</td>
      <td>-0.824</td>
      <td>0.005</td>
      <td>722</td>
      <td>1077</td>
      <td>388018</td>
      <td>1024412</td>
      <td>0.5</td>
      <td>ATC17CGA CGG29TGC</td>
      <td>2</td>
      <td>I17R R29C</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>CGCAAAAGTCTATTAC</td>
      <td>-0.120</td>
      <td>0.005</td>
      <td>646</td>
      <td>1570</td>
      <td>388018</td>
      <td>1024412</td>
      <td>0.5</td>
      <td>TCA11GCG</td>
      <td>1</td>
      <td>S11A</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>TCCATTTGAATTGTTT</td>
      <td>-2.141</td>
      <td>0.009</td>
      <td>640</td>
      <td>383</td>
      <td>388018</td>
      <td>1024412</td>
      <td>0.5</td>
      <td>CGG12CCA CGT15CCG</td>
      <td>2</td>
      <td>R12P R15P</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>CCAGAAAACCCACCAA</td>
      <td>0.224</td>
      <td>0.004</td>
      <td>623</td>
      <td>1922</td>
      <td>388018</td>
      <td>1024412</td>
      <td>0.5</td>
      <td>TTA18CTC</td>
      <td>1</td>
      <td></td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>TTTGCGCGACTTTTAG</td>
      <td>-6.610</td>
      <td>0.130</td>
      <td>610</td>
      <td>16</td>
      <td>388018</td>
      <td>1024412</td>
      <td>0.5</td>
      <td>AAC21CAT ATC28TGG</td>
      <td>2</td>
      <td>N21H I28W</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We can also calculate functional scores at the level of amino-acid or codon substitutions rather than at the level of variants. This calculation groups all variants with the same substitutions before calculating the functional score. Here are scores grouping by amino-acid substitutions; we also set <code class="docutils literal notranslate"><span class="pre">syn_as_wt=True</span></code> to include variants with only synonymous mutations in the counts of wild type in this case:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_func_scores</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s1">&#39;pre-selection&#39;</span><span class="p">,</span>
                                      <span class="n">by</span><span class="o">=</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
                                      <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                      <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">)</span>
<span class="n">aa_func_scores</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>pre_sample</th>
      <th>post_sample</th>
      <th>aa_substitutions</th>
      <th>func_score</th>
      <th>func_score_var</th>
      <th>pre_count</th>
      <th>post_count</th>
      <th>pre_count_wt</th>
      <th>post_count_wt</th>
      <th>pseudocount</th>
      <th>n_aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>I17R R29C</td>
      <td>-0.824</td>
      <td>0.005</td>
      <td>722</td>
      <td>1077</td>
      <td>425632</td>
      <td>1123517</td>
      <td>0.5</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>S11A</td>
      <td>-0.032</td>
      <td>0.002</td>
      <td>1822</td>
      <td>4704</td>
      <td>425632</td>
      <td>1123517</td>
      <td>0.5</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>R12P R15P</td>
      <td>-2.140</td>
      <td>0.009</td>
      <td>640</td>
      <td>383</td>
      <td>425632</td>
      <td>1123517</td>
      <td>0.5</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td></td>
      <td>0.000</td>
      <td>0.000</td>
      <td>425632</td>
      <td>1123517</td>
      <td>425632</td>
      <td>1123517</td>
      <td>0.5</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>loose_bottle</td>
      <td>N21H I28W</td>
      <td>-6.610</td>
      <td>0.130</td>
      <td>610</td>
      <td>16</td>
      <td>425632</td>
      <td>1123517</td>
      <td>0.5</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We can plot the distribution of the functional scores for variants. These plots are most informative if we classify variants by the “types” of mutations they have, which we do here using the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.classifyVariants">CodonVariantTable.classifyVariants</a> method, which adds a column called <code class="docutils literal notranslate"><span class="pre">variant_class</span></code> to the data frame:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">codonvarianttable</span><span class="o">.</span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">func_scores</span><span class="p">)</span>

<span class="p">(</span><span class="n">func_scores</span>
 <span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
 <span class="p">[[</span><span class="s1">&#39;codon_substitutions&#39;</span><span class="p">,</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_class&#39;</span><span class="p">]]</span>
 <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>codon_substitutions</th>
      <th>aa_substitutions</th>
      <th>variant_class</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ATC17CGA CGG29TGC</td>
      <td>I17R R29C</td>
      <td>&gt;1 nonsynonymous</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TCA11GCG</td>
      <td>S11A</td>
      <td>1 nonsynonymous</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CGG12CCA CGT15CCG</td>
      <td>R12P R15P</td>
      <td>&gt;1 nonsynonymous</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TTA18CTC</td>
      <td></td>
      <td>synonymous</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AAC21CAT ATC28TGG</td>
      <td>N21H I28W</td>
      <td>&gt;1 nonsynonymous</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we use <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a> to plot the distributions of scores in ggplot2-like syntax, coloring by the variant class. This plot shows the expected behavior for different variant classes; for instance, stop codon variants tend to have low scores and synonymous variants tend to have wildtype-like (near 0) scores. As expected, there is more noise with a tighter bottleneck:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">func_scores</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;variant_class&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score&#39;</span><span class="p">))</span> <span class="o">+</span>
    <span class="n">geom_violin</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s1">&#39;variant_class&#39;</span><span class="p">))</span> <span class="o">+</span>
    <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;functional score&#39;</span><span class="p">)</span> <span class="o">+</span>
    <span class="n">xlab</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="o">+</span>
    <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;post_sample ~ library&#39;</span><span class="p">)</span> <span class="o">+</span>
    <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.75</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">libs</span><span class="p">),</span>
                       <span class="mi">2</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">bottlenecks</span><span class="p">)),</span>
          <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
          <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
          <span class="p">)</span> <span class="o">+</span>
    <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span> <span class="p">:],</span> <span class="n">guide</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_96_0.png" src="_images/codonvariant_sim_data_96_0.png" />
</div>
</div>
</div>
<div class="section" id="Fit-global-epistasis-models">
<h2>Fit global epistasis models<a class="headerlink" href="#Fit-global-epistasis-models" title="Permalink to this headline">¶</a></h2>
<p>We now fit global epistasis models to the functional scores from each simulated experiment. For background on these models, see <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>. The models we fits are implemented in <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html">dms_variants.globalepistasis</a>, which has extensive documentation.</p>
<p>The primary model of interest is <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasisGaussianLikelihood">MonotonicSplineEpistasisGaussianLikelihood</a>, which assumes that the observed phenotype is a monotonic non-linear function of an underlying additive latent phenotype. As a control, we also fit a
<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasisGaussianLikelihood">NoEpistasisGaussianLikelihood</a> model, which assumes that mutations simply contribute additively to the observed phenotype. Note that for both models we are using the Gaussian-likelihood calculation method to determine model fit (see <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html">dms_variants.globalepistasis</a> for details)–as
described in a later example, you might want to use a Bottleneck- or Cauchy-likelihood if there is a lot of experimental noise.</p>
<p>For the fitting, we first convert the set of functional scores into a binary representation using a <a class="reference external" href="(https://jbloomlab.github.io/dms_variants/dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap)">BinaryMap</a>. Then we create the model, fit it, and store it:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">models</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># store models, keyed by `(modeltype, sample, lib)`</span>

<span class="k">for</span> <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">lib</span><span class="p">),</span> <span class="n">scores</span> <span class="ow">in</span> <span class="n">func_scores</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;post_sample&#39;</span><span class="p">,</span>
                                                  <span class="s1">&#39;library&#39;</span><span class="p">]):</span>

    <span class="n">bmap</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">binarymap</span><span class="o">.</span><span class="n">BinaryMap</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">modeltype</span><span class="p">,</span> <span class="n">EpistasisModel</span> <span class="ow">in</span> <span class="p">[</span>
            <span class="p">(</span><span class="s1">&#39;global epistasis&#39;</span><span class="p">,</span>
             <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">MonotonicSplineEpistasisGaussianLikelihood</span><span class="p">),</span>
            <span class="p">(</span><span class="s1">&#39;no epistasis&#39;</span><span class="p">,</span>
             <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">NoEpistasisGaussianLikelihood</span><span class="p">),</span>
            <span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fitting </span><span class="si">{modeltype}</span><span class="s2"> model to </span><span class="si">{sample}</span><span class="s2">, </span><span class="si">{lib}</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>

        <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">EpistasisModel</span><span class="p">(</span><span class="n">bmap</span><span class="p">)</span>
        <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fitting took {time.time() - start:.1f} sec.&quot;</span><span class="p">)</span>
        <span class="n">models</span><span class="p">[(</span><span class="n">modeltype</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">lib</span><span class="p">)]</span> <span class="o">=</span> <span class="n">model</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Fitting global epistasis model to loose_bottle, lib_1... fitting took 29.6 sec.
Fitting no epistasis model to loose_bottle, lib_1... fitting took 0.2 sec.
Fitting global epistasis model to loose_bottle, lib_2... fitting took 29.0 sec.
Fitting no epistasis model to loose_bottle, lib_2... fitting took 0.2 sec.
Fitting global epistasis model to tight_bottle, lib_1... fitting took 8.0 sec.
Fitting no epistasis model to tight_bottle, lib_1... fitting took 0.2 sec.
Fitting global epistasis model to tight_bottle, lib_2... fitting took 14.6 sec.
Fitting no epistasis model to tight_bottle, lib_2... fitting took 0.2 sec.
</pre></div></div>
</div>
<p>Now we want to see which model fits the data better. To do this, we get the log likelihood of each model along with the number of model parameters, as well as the <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>. Models with lower AIC are better, and below we see that the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a> global epistasis model always fits the data
much better:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">logliks_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
            <span class="p">[(</span><span class="n">modeltype</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">lib</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">nparams</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">loglik</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">aic</span><span class="p">)</span> <span class="k">for</span>
             <span class="p">(</span><span class="n">modeltype</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">lib</span><span class="p">),</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()],</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;model&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;n_parameters&#39;</span><span class="p">,</span> <span class="s1">&#39;log_likelihood&#39;</span><span class="p">,</span> <span class="s1">&#39;AIC&#39;</span><span class="p">]</span>
            <span class="p">)</span>
    <span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">])</span>
    <span class="p">)</span>

<span class="n">logliks_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>model</th>
      <th>n_parameters</th>
      <th>log_likelihood</th>
      <th>AIC</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>library</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="4" valign="top">loose_bottle</th>
      <th>lib_1</th>
      <td>global epistasis</td>
      <td>608</td>
      <td>-12028.1</td>
      <td>25272.3</td>
    </tr>
    <tr>
      <th>lib_1</th>
      <td>no epistasis</td>
      <td>602</td>
      <td>-30387.9</td>
      <td>61979.8</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>global epistasis</td>
      <td>608</td>
      <td>-12776.8</td>
      <td>26769.6</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>no epistasis</td>
      <td>602</td>
      <td>-30255.3</td>
      <td>61714.7</td>
    </tr>
    <tr>
      <th rowspan="4" valign="top">tight_bottle</th>
      <th>lib_1</th>
      <td>global epistasis</td>
      <td>608</td>
      <td>-21882.9</td>
      <td>44981.8</td>
    </tr>
    <tr>
      <th>lib_1</th>
      <td>no epistasis</td>
      <td>602</td>
      <td>-32341.2</td>
      <td>65886.5</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>global epistasis</td>
      <td>608</td>
      <td>-22463.9</td>
      <td>46143.9</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>no epistasis</td>
      <td>602</td>
      <td>-32360.1</td>
      <td>65924.1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Because the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a> global epistasis model fits so much better than the additive <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasis">NoEpistasis</a> model, below we are just going to analyze the results from the global epistasis model.</p>
<p>First, we will examine how the model looks on all the actual variants used to fit the model. We use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.phenotypes_df">AbstractEpistasis.phenotypes_df</a> to get a data frame of all the variants used to fit each global epistasis model along with their functional scores and the latent and observed phenotypes predicted by each model. We also compute the measured and
model-predicted enrichments for each variant (recall that the functional score / observed phenotype is the <span class="math notranslate nohighlight">\(\log_2\)</span> of the enrichment as described <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.func_scores">here</a>):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span><span class="n">model</span><span class="o">.</span><span class="n">phenotypes_df</span>
         <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">,</span>
                 <span class="n">library</span><span class="o">=</span><span class="n">lib</span><span class="p">,</span>
                 <span class="n">predicted_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">model</span><span class="o">.</span><span class="n">enrichments</span><span class="p">(</span>
                                                 <span class="n">x</span><span class="p">[</span><span class="s1">&#39;observed_phenotype&#39;</span><span class="p">]),</span>
                 <span class="n">measured_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span><span class="o">**</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;func_score&#39;</span><span class="p">],</span>
                 <span class="p">)</span>
         <span class="k">for</span> <span class="p">(</span><span class="n">modeltype</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">lib</span><span class="p">),</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
         <span class="k">if</span> <span class="n">modeltype</span> <span class="o">==</span> <span class="s1">&#39;global epistasis&#39;</span><span class="p">],</span>
        <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">variants_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>aa_substitutions</th>
      <th>func_score</th>
      <th>func_score_var</th>
      <th>latent_phenotype</th>
      <th>observed_phenotype</th>
      <th>sample</th>
      <th>library</th>
      <th>predicted_enrichment</th>
      <th>measured_enrichment</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>I17R R29C</td>
      <td>-0.82</td>
      <td>0.00</td>
      <td>-0.80</td>
      <td>-1.06</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>0.49</td>
      <td>0.56</td>
    </tr>
    <tr>
      <th>1</th>
      <td>S11A</td>
      <td>-0.12</td>
      <td>0.00</td>
      <td>-0.12</td>
      <td>-0.03</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>1.00</td>
      <td>0.92</td>
    </tr>
    <tr>
      <th>2</th>
      <td>R12P R15P</td>
      <td>-2.14</td>
      <td>0.01</td>
      <td>-0.99</td>
      <td>-2.24</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>0.22</td>
      <td>0.23</td>
    </tr>
    <tr>
      <th>3</th>
      <td></td>
      <td>0.22</td>
      <td>0.00</td>
      <td>0.00</td>
      <td>-0.03</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>1.00</td>
      <td>1.17</td>
    </tr>
    <tr>
      <th>4</th>
      <td>N21H I28W</td>
      <td>-6.61</td>
      <td>0.13</td>
      <td>-1.63</td>
      <td>-6.74</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>0.01</td>
      <td>0.01</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Below we plot the relationships among the latent phenotype from the model, the observed phenotype from the model, and the measured functional score for all variants used to fit the model:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">([</span><span class="s1">&#39;latent_phenotype&#39;</span><span class="p">,</span>
                                    <span class="s1">&#39;observed_phenotype&#39;</span><span class="p">,</span>
                                    <span class="s1">&#39;func_score&#39;</span><span class="p">],</span>
                                   <span class="mi">2</span><span class="p">):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">))</span> <span class="o">+</span>
        <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">+</span>
        <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;library ~ sample&#39;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s1">&#39;free_y&#39;</span><span class="p">)</span> <span class="o">+</span>
        <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
                           <span class="mi">2</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()),</span>
              <span class="p">)</span>
        <span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_105_0.png" src="_images/codonvariant_sim_data_105_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_105_1.png" src="_images/codonvariant_sim_data_105_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_105_2.png" src="_images/codonvariant_sim_data_105_2.png" />
</div>
</div>
</div>
<div class="section" id="Model-vs-measurements-vs-truth-for-all-variants">
<h2>Model vs measurements vs truth for all variants<a class="headerlink" href="#Model-vs-measurements-vs-truth-for-all-variants" title="Permalink to this headline">¶</a></h2>
<p>Because we simulated the variants, we can also use the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator">SigmoidPhenotypeSimulator</a> used in these simulations to get the <strong>true</strong> phenotype of each variant. (Obviously in real non-simulated experiments the true phenotypes are unknown.) We get the true latent phenotype, the true observed phenotype (which corresponds to the functional score), and true enrichment:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">variants_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">variants_df</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">true_latent_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">),</span>
        <span class="n">true_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedEnrichment</span><span class="p">),</span>
        <span class="n">true_observed_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

<span class="n">variants_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>aa_substitutions</th>
      <th>func_score</th>
      <th>func_score_var</th>
      <th>latent_phenotype</th>
      <th>observed_phenotype</th>
      <th>sample</th>
      <th>library</th>
      <th>predicted_enrichment</th>
      <th>measured_enrichment</th>
      <th>true_latent_phenotype</th>
      <th>true_enrichment</th>
      <th>true_observed_phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>I17R R29C</td>
      <td>-0.82</td>
      <td>0.00</td>
      <td>-0.80</td>
      <td>-1.06</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>0.49</td>
      <td>0.56</td>
      <td>-0.01</td>
      <td>0.50</td>
      <td>-0.99</td>
    </tr>
    <tr>
      <th>1</th>
      <td>S11A</td>
      <td>-0.12</td>
      <td>0.00</td>
      <td>-0.12</td>
      <td>-0.03</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>1.00</td>
      <td>0.92</td>
      <td>4.14</td>
      <td>0.99</td>
      <td>-0.01</td>
    </tr>
    <tr>
      <th>2</th>
      <td>R12P R15P</td>
      <td>-2.14</td>
      <td>0.01</td>
      <td>-0.99</td>
      <td>-2.24</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>0.22</td>
      <td>0.23</td>
      <td>-1.28</td>
      <td>0.22</td>
      <td>-2.18</td>
    </tr>
    <tr>
      <th>3</th>
      <td></td>
      <td>0.22</td>
      <td>0.00</td>
      <td>0.00</td>
      <td>-0.03</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>1.00</td>
      <td>1.17</td>
      <td>5.00</td>
      <td>1.00</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>N21H I28W</td>
      <td>-6.61</td>
      <td>0.13</td>
      <td>-1.63</td>
      <td>-6.74</td>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>0.01</td>
      <td>0.01</td>
      <td>-5.00</td>
      <td>0.01</td>
      <td>-7.02</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we calculate the correlations between the true enrichment (from the simulation) and the measured enrichment or model-predicted enrichment, calculating the correlations across all variants. We do this using enrichment rather than observed phenotype as we expect the observed phenotypes to be very noisy at the “low end” (since experiments lose the sensitivity to distinguish among bad and really-bad variants):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">enrichments_corr</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">variants_df</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;model prediction&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;measurement&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
          <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;model prediction&#39;</span><span class="p">,</span> <span class="s1">&#39;measurement&#39;</span><span class="p">],</span>
          <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
          <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
          <span class="p">)</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment_type&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;enrichment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">]))</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;correlation&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">],</span>
                 <span class="n">values</span><span class="o">=</span><span class="s1">&#39;correlation&#39;</span><span class="p">,</span>
                 <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">)</span>
    <span class="p">)</span>

<span class="n">enrichments_corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enrichment_type</th>
      <th>measurement</th>
      <th>model prediction</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>library</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="2" valign="top">loose_bottle</th>
      <th>lib_1</th>
      <td>0.98</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>0.98</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">tight_bottle</th>
      <th>lib_1</th>
      <td>0.83</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>0.82</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>And below are the correlation plots corresponding to the correlation coefficients tabulated above:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">variants_df</span>
            <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
                  <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">],</span>
                  <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
                  <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
                  <span class="p">),</span>
            <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment&#39;</span><span class="p">))</span> <span class="o">+</span>
     <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;enrichment_type ~ sample + library&#39;</span><span class="p">,</span>
                <span class="n">scales</span><span class="o">=</span><span class="s1">&#39;free_y&#39;</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.5</span> <span class="o">*</span> <span class="p">(</span><span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span> <span class="o">+</span>
                               <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()),</span>
                        <span class="mi">5</span><span class="p">)</span>
           <span class="p">)</span>
     <span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_113_0.png" src="_images/codonvariant_sim_data_113_0.png" />
</div>
</div>
<p>These results show that the model predictions are closer to the true values than the actual measurements, which is great! This could be because modeling the data helps extract more signal from multiple mutants–but it could also be for the trivial reason that the model effectively averages over multiple variants with the same mutations, which could be more accurate than the individual measurements. Therefore, in the next section we look at this issue…</p>
</div>
<div class="section" id="Model-vs-measurements-vs-truth-for-amino-acid-variants">
<h2>Model vs measurements vs truth for amino-acid variants<a class="headerlink" href="#Model-vs-measurements-vs-truth-for-amino-acid-variants" title="Permalink to this headline">¶</a></h2>
<p>We want to see if the model still outperforms the experiments if look at the level of <strong>amino-acid</strong> variants (meaning we pool the counts for all variants with the same amino-acid substitutions together before calculating the functional scores). Recall that above we already got such functional scores into the variable <code class="docutils literal notranslate"><span class="pre">aa_func_scores</span></code>:</p>
<p>First we get a data frame of simulated functional scores calculated at the level of amino-acid variants (pooling all barcoded variants with the same amino-acid mutations):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_variants_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_func_scores</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;post_sample&#39;</span><span class="p">:</span> <span class="s1">&#39;sample&#39;</span><span class="p">})</span>
    <span class="p">[[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
      <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score&#39;</span><span class="p">]]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">measured_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span><span class="o">**</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;func_score&#39;</span><span class="p">])</span>
    <span class="p">)</span>

<span class="n">aa_variants_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample</th>
      <th>library</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
      <th>func_score</th>
      <th>measured_enrichment</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>I17R R29C</td>
      <td>2</td>
      <td>-0.82</td>
      <td>0.56</td>
    </tr>
    <tr>
      <th>1</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>S11A</td>
      <td>1</td>
      <td>-0.03</td>
      <td>0.98</td>
    </tr>
    <tr>
      <th>2</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>R12P R15P</td>
      <td>2</td>
      <td>-2.14</td>
      <td>0.23</td>
    </tr>
    <tr>
      <th>3</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td></td>
      <td>0</td>
      <td>0.00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>N21H I28W</td>
      <td>2</td>
      <td>-6.61</td>
      <td>0.01</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we add the model-predicted latent phenotype, observed phenotype, and enrichment for each amino-acid variant using the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.add_phenotypes_to_df">AbstractEpistasis.add_phenotypes_to_df</a> method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">df_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">lib</span><span class="p">),</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">aa_variants_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">]):</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">models</span><span class="p">[(</span><span class="s1">&#39;global epistasis&#39;</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">lib</span><span class="p">)]</span>
    <span class="n">df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">add_phenotypes_to_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
                   <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">predicted_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                              <span class="n">model</span><span class="o">.</span><span class="n">enrichments</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;observed_phenotype&#39;</span><span class="p">])</span>
                           <span class="p">)</span>
                   <span class="p">)</span>
<span class="n">aa_variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_list</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">aa_variants_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample</th>
      <th>library</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
      <th>func_score</th>
      <th>measured_enrichment</th>
      <th>latent_phenotype</th>
      <th>observed_phenotype</th>
      <th>predicted_enrichment</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>I17R R29C</td>
      <td>2</td>
      <td>-0.82</td>
      <td>0.56</td>
      <td>-0.80</td>
      <td>-1.06</td>
      <td>0.49</td>
    </tr>
    <tr>
      <th>1</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>S11A</td>
      <td>1</td>
      <td>-0.03</td>
      <td>0.98</td>
      <td>-0.12</td>
      <td>-0.03</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>R12P R15P</td>
      <td>2</td>
      <td>-2.14</td>
      <td>0.23</td>
      <td>-0.99</td>
      <td>-2.24</td>
      <td>0.22</td>
    </tr>
    <tr>
      <th>3</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td></td>
      <td>0</td>
      <td>0.00</td>
      <td>1.00</td>
      <td>0.00</td>
      <td>-0.03</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>N21H I28W</td>
      <td>2</td>
      <td>-6.61</td>
      <td>0.01</td>
      <td>-1.63</td>
      <td>-6.74</td>
      <td>0.01</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we add the “true” values from the simulator:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">aa_variants_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_variants_df</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">true_latent_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">),</span>
        <span class="n">true_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedEnrichment</span><span class="p">),</span>
        <span class="n">true_observed_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

<span class="n">aa_variants_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample</th>
      <th>library</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
      <th>func_score</th>
      <th>measured_enrichment</th>
      <th>latent_phenotype</th>
      <th>observed_phenotype</th>
      <th>predicted_enrichment</th>
      <th>true_latent_phenotype</th>
      <th>true_enrichment</th>
      <th>true_observed_phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>I17R R29C</td>
      <td>2</td>
      <td>-0.82</td>
      <td>0.56</td>
      <td>-0.80</td>
      <td>-1.06</td>
      <td>0.49</td>
      <td>-0.01</td>
      <td>0.50</td>
      <td>-0.99</td>
    </tr>
    <tr>
      <th>1</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>S11A</td>
      <td>1</td>
      <td>-0.03</td>
      <td>0.98</td>
      <td>-0.12</td>
      <td>-0.03</td>
      <td>1.00</td>
      <td>4.14</td>
      <td>0.99</td>
      <td>-0.01</td>
    </tr>
    <tr>
      <th>2</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>R12P R15P</td>
      <td>2</td>
      <td>-2.14</td>
      <td>0.23</td>
      <td>-0.99</td>
      <td>-2.24</td>
      <td>0.22</td>
      <td>-1.28</td>
      <td>0.22</td>
      <td>-2.18</td>
    </tr>
    <tr>
      <th>3</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td></td>
      <td>0</td>
      <td>0.00</td>
      <td>1.00</td>
      <td>0.00</td>
      <td>-0.03</td>
      <td>1.00</td>
      <td>5.00</td>
      <td>1.00</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>N21H I28W</td>
      <td>2</td>
      <td>-6.61</td>
      <td>0.01</td>
      <td>-1.63</td>
      <td>-6.74</td>
      <td>0.01</td>
      <td>-5.00</td>
      <td>0.01</td>
      <td>-7.02</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we calculate and plot the correlations between the true enrichments from the simulations and the measured values and the ones predicted by the model:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">enrichments_corr</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_variants_df</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;model prediction&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;measurement&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
          <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;model prediction&#39;</span><span class="p">,</span> <span class="s1">&#39;measurement&#39;</span><span class="p">],</span>
          <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
          <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
          <span class="p">)</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment_type&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;enrichment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">]))</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;correlation&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">],</span>
                 <span class="n">values</span><span class="o">=</span><span class="s1">&#39;correlation&#39;</span><span class="p">,</span>
                 <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">)</span>
    <span class="p">)</span>

<span class="n">enrichments_corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enrichment_type</th>
      <th>measurement</th>
      <th>model prediction</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>library</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="2" valign="top">loose_bottle</th>
      <th>lib_1</th>
      <td>0.99</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>0.99</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">tight_bottle</th>
      <th>lib_1</th>
      <td>0.88</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>0.87</td>
      <td>0.99</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">aa_variants_df</span>
            <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
                  <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">],</span>
                  <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
                  <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
                  <span class="p">),</span>
            <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment&#39;</span><span class="p">))</span> <span class="o">+</span>
     <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;enrichment_type ~ sample + library&#39;</span><span class="p">,</span>
                <span class="n">scales</span><span class="o">=</span><span class="s1">&#39;free_y&#39;</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.5</span> <span class="o">*</span> <span class="p">(</span><span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span> <span class="o">+</span>
                               <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()),</span>
                        <span class="mi">5</span><span class="p">)</span>
           <span class="p">)</span>
     <span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_124_0.png" src="_images/codonvariant_sim_data_124_0.png" />
</div>
</div>
<p>The results above show that the model still outperforms the direct measurements even when the measurements are aggregated at the level of amino-acid variants.</p>
</div>
<div class="section" id="Model-vs-measurements-vs-truth-for-single-mutants">
<h2>Model vs measurements vs truth for single mutants<a class="headerlink" href="#Model-vs-measurements-vs-truth-for-single-mutants" title="Permalink to this headline">¶</a></h2>
<p>Now we repeat the above analysis for just <strong>single</strong> amino-acid mutants.</p>
<p>First, get the single amino-acid mutant variants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">single_aa_variants_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_variants_df</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_aa_substitutions == 1&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="p">)</span>

<span class="n">single_aa_variants_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample</th>
      <th>library</th>
      <th>aa_substitutions</th>
      <th>n_aa_substitutions</th>
      <th>func_score</th>
      <th>measured_enrichment</th>
      <th>latent_phenotype</th>
      <th>observed_phenotype</th>
      <th>predicted_enrichment</th>
      <th>true_latent_phenotype</th>
      <th>true_enrichment</th>
      <th>true_observed_phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>S11A</td>
      <td>1</td>
      <td>-0.03</td>
      <td>0.98</td>
      <td>-0.12</td>
      <td>-0.03</td>
      <td>1.00</td>
      <td>4.14</td>
      <td>0.99</td>
      <td>-0.01</td>
    </tr>
    <tr>
      <th>1</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>R1A</td>
      <td>1</td>
      <td>0.06</td>
      <td>1.04</td>
      <td>0.12</td>
      <td>-0.03</td>
      <td>1.00</td>
      <td>5.76</td>
      <td>1.00</td>
      <td>0.01</td>
    </tr>
    <tr>
      <th>2</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>I4P</td>
      <td>1</td>
      <td>-6.89</td>
      <td>0.01</td>
      <td>-1.63</td>
      <td>-6.75</td>
      <td>0.01</td>
      <td>-4.96</td>
      <td>0.01</td>
      <td>-6.97</td>
    </tr>
    <tr>
      <th>3</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>R26V</td>
      <td>1</td>
      <td>0.00</td>
      <td>1.00</td>
      <td>-0.32</td>
      <td>-0.03</td>
      <td>1.00</td>
      <td>3.15</td>
      <td>0.97</td>
      <td>-0.05</td>
    </tr>
    <tr>
      <th>4</th>
      <td>loose_bottle</td>
      <td>lib_1</td>
      <td>R6*</td>
      <td>1</td>
      <td>-10.47</td>
      <td>0.00</td>
      <td>-2.45</td>
      <td>-8.62</td>
      <td>0.00</td>
      <td>-10.00</td>
      <td>0.00</td>
      <td>-9.90</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now tabulate and plot how the model predictions and measurements compare to the true values:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">enrichments_corr</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">single_aa_variants_df</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;model prediction&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;measurement&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
          <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;model prediction&#39;</span><span class="p">,</span> <span class="s1">&#39;measurement&#39;</span><span class="p">],</span>
          <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
          <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
          <span class="p">)</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment_type&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;enrichment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">]))</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;correlation&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">],</span>
                 <span class="n">values</span><span class="o">=</span><span class="s1">&#39;correlation&#39;</span><span class="p">,</span>
                 <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">)</span>
    <span class="p">)</span>

<span class="n">enrichments_corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enrichment_type</th>
      <th>measurement</th>
      <th>model prediction</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>library</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="2" valign="top">loose_bottle</th>
      <th>lib_1</th>
      <td>0.99</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>0.99</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">tight_bottle</th>
      <th>lib_1</th>
      <td>0.93</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>lib_2</th>
      <td>0.94</td>
      <td>1.00</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">single_aa_variants_df</span>
            <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
                  <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">],</span>
                  <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
                  <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
                  <span class="p">),</span>
            <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment&#39;</span><span class="p">))</span> <span class="o">+</span>
     <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;enrichment_type ~ sample + library&#39;</span><span class="p">,</span>
                <span class="n">scales</span><span class="o">=</span><span class="s1">&#39;free_y&#39;</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.5</span> <span class="o">*</span> <span class="p">(</span><span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span> <span class="o">+</span>
                               <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()),</span>
                        <span class="mi">5</span><span class="p">)</span>
           <span class="p">)</span>
     <span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_132_0.png" src="_images/codonvariant_sim_data_132_0.png" />
</div>
</div>
<p>We see that even for single mutants, fitting the global epistasis model leads to more accurate measurements of the effects of mutations. Presumably this is because these models are able to effectively extract information from the multiple mutants.</p>
</div>
<div class="section" id="Effects-of-mutations">
<h2>Effects of mutations<a class="headerlink" href="#Effects-of-mutations" title="Permalink to this headline">¶</a></h2>
<p>It’s often useful to try to summarize the effects of individual mutations with a single value. These values can then be nicely displayed in logo plots using programs like:</p>
<ul class="simple">
<li><p><a class="reference external" href="http://jbloomlab.github.io/dms_tools2/dms2_logoplot.html">dms2_logoplot</a></p></li>
<li><p><a class="reference external" href="https://jbloomlab.github.io/dmslogo">dmslogo</a></p></li>
<li><p><a class="reference external" href="https://logomaker.readthedocs.io">logomaker</a></p></li>
</ul>
<p>Here we will make some plots of mutational effects of various types using <a class="reference external" href="https://jbloomlab.github.io/dmslogo">dmslogo</a>. Do keep in mind, however, that the whole point of the epistasis models is that it isn’t possible to summarize everything about a mutation with a single value–so use such plots as they are very informative ways to look at the data, but also recognize they are summarizing the impacts of mutations down to one value.</p>
<div class="section" id="Mutational-effects-on-phenotype">
<h3>Mutational effects on phenotype<a class="headerlink" href="#Mutational-effects-on-phenotype" title="Permalink to this headline">¶</a></h3>
<p>One useful measure of the effects of mutations is simply how the single mutation changes the latent or observed phenotype (note that while the effect on the latent phenotype should be “universal”, that on the observed phenotype depends on the wildtype sequence).</p>
<p>The epistasis models have a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.single_mut_effects">single_mut_effects</a> method to get these mutational effects. By default, this method scales the effects so that their mean absolute value is one.</p>
<p>Mutational effects on the latent phenotype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">latenteffects_df</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">single_mut_effects</span><span class="p">(</span><span class="s1">&#39;latent&#39;</span><span class="p">)</span>

<span class="n">latenteffects_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mutation</th>
      <th>wildtype</th>
      <th>site</th>
      <th>mutant</th>
      <th>effect</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>V13*</td>
      <td>V</td>
      <td>13</td>
      <td>*</td>
      <td>-5.41</td>
    </tr>
    <tr>
      <th>1</th>
      <td>L23*</td>
      <td>L</td>
      <td>23</td>
      <td>*</td>
      <td>-5.40</td>
    </tr>
    <tr>
      <th>2</th>
      <td>V3L</td>
      <td>V</td>
      <td>3</td>
      <td>L</td>
      <td>-5.18</td>
    </tr>
    <tr>
      <th>3</th>
      <td>R26*</td>
      <td>R</td>
      <td>26</td>
      <td>*</td>
      <td>-5.15</td>
    </tr>
    <tr>
      <th>4</th>
      <td>N27*</td>
      <td>N</td>
      <td>27</td>
      <td>*</td>
      <td>-5.10</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Mutational effects on the observed phenotype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">observedeffects_df</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">single_mut_effects</span><span class="p">(</span><span class="s1">&#39;observed&#39;</span><span class="p">)</span>

<span class="n">observedeffects_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mutation</th>
      <th>wildtype</th>
      <th>site</th>
      <th>mutant</th>
      <th>effect</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>V13*</td>
      <td>V</td>
      <td>13</td>
      <td>*</td>
      <td>-3.41</td>
    </tr>
    <tr>
      <th>1</th>
      <td>L23*</td>
      <td>L</td>
      <td>23</td>
      <td>*</td>
      <td>-3.41</td>
    </tr>
    <tr>
      <th>2</th>
      <td>V3L</td>
      <td>V</td>
      <td>3</td>
      <td>L</td>
      <td>-3.41</td>
    </tr>
    <tr>
      <th>3</th>
      <td>R26*</td>
      <td>R</td>
      <td>26</td>
      <td>*</td>
      <td>-3.41</td>
    </tr>
    <tr>
      <th>4</th>
      <td>N27*</td>
      <td>N</td>
      <td>27</td>
      <td>*</td>
      <td>-3.41</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Phenotypic-mutational-effects-to-“preferences”">
<h3>Phenotypic mutational effects to “preferences”<a class="headerlink" href="#Phenotypic-mutational-effects-to-“preferences”" title="Permalink to this headline">¶</a></h3>
<p>It is sometimes useful to convert mutational effects to “preferences.” Essentially, the preferences are the exponential of the mutational effects re-scaled to sum to one at each site. In other words, the preference <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span> of site <span class="math notranslate nohighlight">\(r\)</span> for character (e.g., amino acid) <span class="math notranslate nohighlight">\(a\)</span> is simply:</p>
<div class="math notranslate nohighlight">
\[\pi_{r,a} = \frac{b^{p_{r,a}}}{\sum_{a'} b^{p_{r,a'}}}\]</div>
<p>where <span class="math notranslate nohighlight">\(p_{r,a}\)</span> is the phenotype of the variant with the single mutation of site <span class="math notranslate nohighlight">\(r\)</span> to <span class="math notranslate nohighlight">\(a\)</span>, and <span class="math notranslate nohighlight">\(b\)</span> is the logarithm base.</p>
<p>Amino-acid preferences are useful as they can be displayed in logo plots where the height of the letter is proportional to the preference for that amino acid, and can be used to fit <a class="reference external" href="https://jbloomlab.github.io/phydms/ExpCM.html">experimentally informed phylogenetic substitution models</a> using programs such as <a class="reference external" href="https://jbloomlab.github.io/phydms">phydms</a> to directly compare the deep mutational scanning experiments to natural evolution.</p>
<p>The epistasis models have a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.preferences">preferences</a> method for getting the amino-acid preferences from the effects of single mutations on either the latent or observed phenotype (see also the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.utils.html#dms_variants.utils.scores_to_prefs">scores_to_prefs</a> function if you need to directly convert functional scores
not within an epistasis model).</p>
<p>A <strong>very important point</strong> is that preferences are estimated for all characters (i.e., amino acids) at a site, even if there is not an effect estimated for that mutation! For missing scores, the preferences for missing characters are set to the average effects of all mutations (or optionally ammutations at that site); see docs for <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.preferences">preferences</a> for more
details. This is necessary since the preferences are normalized per site. If you are missing just a few measurements this extrapolation of missing values may be fine, but it probably is not a good idea if you are missing lots of measurements.</p>
<p>For instance, the next cell converts the latent effects to a data frame of preferences using an exponent base of 10:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">prefs_base10</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">preferences</span><span class="p">(</span><span class="n">phenotype</span><span class="o">=</span><span class="s1">&#39;latent&#39;</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="n">prefs_base10</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>site</th>
      <th>A</th>
      <th>C</th>
      <th>D</th>
      <th>E</th>
      <th>F</th>
      <th>G</th>
      <th>H</th>
      <th>I</th>
      <th>K</th>
      <th>...</th>
      <th>M</th>
      <th>N</th>
      <th>P</th>
      <th>Q</th>
      <th>R</th>
      <th>S</th>
      <th>T</th>
      <th>V</th>
      <th>W</th>
      <th>Y</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>0.094</td>
      <td>0.028</td>
      <td>0.001</td>
      <td>0.011</td>
      <td>0.196</td>
      <td>0.009</td>
      <td>0.025</td>
      <td>0.006</td>
      <td>0.173</td>
      <td>...</td>
      <td>0.018</td>
      <td>0.040</td>
      <td>0.071</td>
      <td>0.072</td>
      <td>0.076</td>
      <td>0.002</td>
      <td>0.059</td>
      <td>0.035</td>
      <td>0.044</td>
      <td>0.001</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>0.096</td>
      <td>0.018</td>
      <td>0.096</td>
      <td>0.006</td>
      <td>0.000</td>
      <td>0.007</td>
      <td>0.003</td>
      <td>0.027</td>
      <td>0.003</td>
      <td>...</td>
      <td>0.095</td>
      <td>0.048</td>
      <td>0.006</td>
      <td>0.065</td>
      <td>0.020</td>
      <td>0.156</td>
      <td>0.004</td>
      <td>0.154</td>
      <td>0.066</td>
      <td>0.000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>0.125</td>
      <td>0.033</td>
      <td>0.000</td>
      <td>0.146</td>
      <td>0.044</td>
      <td>0.005</td>
      <td>0.025</td>
      <td>0.069</td>
      <td>0.029</td>
      <td>...</td>
      <td>0.002</td>
      <td>0.006</td>
      <td>0.147</td>
      <td>0.053</td>
      <td>0.002</td>
      <td>0.100</td>
      <td>0.002</td>
      <td>0.119</td>
      <td>0.002</td>
      <td>0.090</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>0.012</td>
      <td>0.012</td>
      <td>0.001</td>
      <td>0.063</td>
      <td>0.029</td>
      <td>0.160</td>
      <td>0.067</td>
      <td>0.085</td>
      <td>0.055</td>
      <td>...</td>
      <td>0.043</td>
      <td>0.062</td>
      <td>0.002</td>
      <td>0.008</td>
      <td>0.044</td>
      <td>0.090</td>
      <td>0.112</td>
      <td>0.046</td>
      <td>0.007</td>
      <td>0.056</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>0.072</td>
      <td>0.004</td>
      <td>0.003</td>
      <td>0.007</td>
      <td>0.003</td>
      <td>0.008</td>
      <td>0.062</td>
      <td>0.005</td>
      <td>0.059</td>
      <td>...</td>
      <td>0.225</td>
      <td>0.113</td>
      <td>0.004</td>
      <td>0.004</td>
      <td>0.113</td>
      <td>0.084</td>
      <td>0.015</td>
      <td>0.027</td>
      <td>0.077</td>
      <td>0.002</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21 columns</p>
</div></div>
</div>
<p>We can also get the preferences in tidy format:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">prefs_base10_tidy</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">preferences</span><span class="p">(</span><span class="n">phenotype</span><span class="o">=</span><span class="s1">&#39;latent&#39;</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">returnformat</span><span class="o">=</span><span class="s1">&#39;tidy&#39;</span><span class="p">)</span>

<span class="n">prefs_base10_tidy</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>wildtype</th>
      <th>site</th>
      <th>mutant</th>
      <th>preference</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>595</th>
      <td>P</td>
      <td>24</td>
      <td>Y</td>
      <td>0.232</td>
    </tr>
    <tr>
      <th>596</th>
      <td>S</td>
      <td>11</td>
      <td>C</td>
      <td>0.232</td>
    </tr>
    <tr>
      <th>597</th>
      <td>I</td>
      <td>25</td>
      <td>Y</td>
      <td>0.234</td>
    </tr>
    <tr>
      <th>598</th>
      <td>N</td>
      <td>10</td>
      <td>D</td>
      <td>0.238</td>
    </tr>
    <tr>
      <th>599</th>
      <td>R</td>
      <td>29</td>
      <td>I</td>
      <td>0.250</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Here we will plot these preferences using <a class="reference external" href="https://jbloomlab.github.io/dmslogo">dmslogo</a>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[75]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">_</span> <span class="o">=</span> <span class="n">dmslogo</span><span class="o">.</span><span class="n">draw_logo</span><span class="p">(</span>
            <span class="n">data</span><span class="o">=</span><span class="n">prefs_base10_tidy</span><span class="p">,</span>
            <span class="n">x_col</span><span class="o">=</span><span class="s1">&#39;site&#39;</span><span class="p">,</span>
            <span class="n">letter_col</span><span class="o">=</span><span class="s1">&#39;mutant&#39;</span><span class="p">,</span>
            <span class="n">letter_height_col</span><span class="o">=</span><span class="s1">&#39;preference&#39;</span><span class="p">,</span>
            <span class="n">title</span><span class="o">=</span><span class="s1">&#39;exponent base 10&#39;</span>
            <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_143_0.png" src="_images/codonvariant_sim_data_143_0.png" />
</div>
</div>
<p>Note how the preferences plotted above look rather “flat”, with no amino acid appearing to be vastly more preferred than another. If we re-plot the data using a larger exponent base, the preferences are now less flat and more peaked:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[76]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">prefs_base100_tidy</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">preferences</span><span class="p">(</span><span class="n">phenotype</span><span class="o">=</span><span class="s1">&#39;latent&#39;</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">returnformat</span><span class="o">=</span><span class="s1">&#39;tidy&#39;</span><span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">dmslogo</span><span class="o">.</span><span class="n">draw_logo</span><span class="p">(</span>
            <span class="n">data</span><span class="o">=</span><span class="n">prefs_base100_tidy</span><span class="p">,</span>
            <span class="n">x_col</span><span class="o">=</span><span class="s1">&#39;site&#39;</span><span class="p">,</span>
            <span class="n">letter_col</span><span class="o">=</span><span class="s1">&#39;mutant&#39;</span><span class="p">,</span>
            <span class="n">letter_height_col</span><span class="o">=</span><span class="s1">&#39;preference&#39;</span><span class="p">,</span>
            <span class="n">title</span><span class="o">=</span><span class="s1">&#39;exponent base 100&#39;</span>
            <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_145_0.png" src="_images/codonvariant_sim_data_145_0.png" />
</div>
</div>
<p>So what is the “best” exponent base to use? It turns out that different exponent bases simply correspond to different linear scalings of the phenotypic effects of mutations (you can see this by working through the math in the equation above to convert the phenotypes to preferences). So it’s not obvious what is the “best” exponent base when converting phenotypic effects to preferences, particularly for the latent effects of mutations which are themselves arbitrarily scaled.</p>
<p>We therefore recommend just choosing a base and then using <a class="reference external" href="https://jbloomlab.github.io/phydms">phydms</a> to fit a stringency parameter to scale the preferences so that they best match the strength of selection observed in natural evolution (see <a class="reference external" href="https://peerj.com/articles/3657">here</a>). It turns out that scaling by a stringency parameter in this way is equivalent to choosing a different exponent base.</p>
<p>An <strong>important caveat</strong> in doing this is that for numerical reasons, <a class="reference external" href="https://jbloomlab.github.io/phydms">phydms</a> can only fit stringency parameters up to a maximum bound (currently 10). So if you are gettting that upper bound stringency parameter, you should use a larger exponent base when initially converting the phenotypic effects into preferences.</p>
<p>After you’ve gotten a stringency parameter, you can use it to re-scale the preferences you obtain from an epistasis model. For instance:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[77]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">stringency_param</span> <span class="o">=</span> <span class="mi">2</span>

<span class="n">prefs_base100_tidy_stringency</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">preferences</span><span class="p">(</span><span class="n">phenotype</span><span class="o">=</span><span class="s1">&#39;latent&#39;</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                                                  <span class="n">returnformat</span><span class="o">=</span><span class="s1">&#39;tidy&#39;</span><span class="p">,</span>
                                                  <span class="n">stringency_param</span><span class="o">=</span><span class="n">stringency_param</span><span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">dmslogo</span><span class="o">.</span><span class="n">draw_logo</span><span class="p">(</span>
            <span class="n">data</span><span class="o">=</span><span class="n">prefs_base100_tidy_stringency</span><span class="p">,</span>
            <span class="n">x_col</span><span class="o">=</span><span class="s1">&#39;site&#39;</span><span class="p">,</span>
            <span class="n">letter_col</span><span class="o">=</span><span class="s1">&#39;mutant&#39;</span><span class="p">,</span>
            <span class="n">letter_height_col</span><span class="o">=</span><span class="s1">&#39;preference&#39;</span><span class="p">,</span>
            <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;exponent base 100, stringency param </span><span class="si">{stringency_param}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/codonvariant_sim_data_147_0.png" src="_images/codonvariant_sim_data_147_0.png" />
</div>
</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_variants">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="examples.html">Examples</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Analyze simulated library of codon variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data_multi_targets.html">Libraries with secondary target genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="narrow_bottleneck.html">Fitting gobal epistasis models with experimental “noise”</a></li>
<li class="toctree-l2"><a class="reference internal" href="predict_variants.html">Predict unseen variants using global epistasis models</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_plot_formatting.html">Formatting <code class="docutils literal notranslate"><span class="pre">CodonVariantTable</span></code> plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="parsebarcodes_sim_data.html">Parse barcodes</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dms_variants.html">dms_variants package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="examples.html">Examples</a><ul>
      <li>Previous: <a href="examples.html" title="previous chapter">Examples</a></li>
      <li>Next: <a href="codonvariant_sim_data_multi_targets.html" title="next chapter">Libraries with secondary target genes</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2020.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/codonvariant_sim_data.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>