<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>codonvarianttable &#8212; dms_variants 1.5.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=420b4293" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css?v=7f9a90b1" />
    <script src="_static/documentation_options.js?v=e0a75244"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="constants" href="dms_variants.constants.html" />
    <link rel="prev" title="bottlenecks" href="dms_variants.bottlenecks.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="codonvarianttable">
<span id="module-dms_variants.codonvarianttable"></span><h1>codonvarianttable<a class="headerlink" href="#codonvarianttable" title="Link to this heading">¶</a></h1>
<p>Defines <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> objects for storing and
handling codon variants of a gene.</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dms_variants.codonvarianttable.</span></span><span class="sig-name descname"><span class="pre">CodonVariantTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_variant_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geneseq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allowgaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substitutions_are_codon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substitutions_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'substitutions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store and handle counts of barcoded codon variants of a gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>barcode_variant_file</strong> (<em>str</em>) – CSV file giving barcodes and variants. Must have columns “library”,
“barcode”, “substitutions” (nucleotide mutations in 1, …, numbering
in a format like “G301A A302T G856C”), and “variant_call_support”
(sequences supporting barcode-variant call). Additional columns are
removed unless they are specified <cite>extra_cols</cite>.</p></li>
<li><p><strong>geneseq</strong> (<em>str</em>) – Sequence of wildtype protein-coding gene.</p></li>
<li><p><strong>substitutions_are_codon</strong> (<em>bool</em>) – If <cite>True</cite>, then “substitutions” column in <cite>barcode_variant_file</cite> gives
substitutions as codon rather than nucleotide mutations (e.g.,
“ATG1ATA GTA5CCC” for substitutions at codons 1 and 5.</p></li>
<li><p><strong>allowgaps</strong> (<em>bool</em>) – Allow in-frame codon-length gaps (<code class="docutils literal notranslate"><span class="pre">---</span></code>) as valid codon substitution.</p></li>
<li><p><strong>extra_cols</strong> (<em>list</em>) – Additional columns in <cite>barcode_variant_file</cite> to retain when creating
<cite>barcode_variant_df</cite> and <cite>variant_count_df</cite> attributes.</p></li>
<li><p><strong>substitutions_col</strong> (<em>str</em>) – Name of substitutions column in <cite>barcode_variant_file</cite> (use if you
want it to be something other than “substitutions”).</p></li>
<li><p><strong>primary_target</strong> (<em>str</em><em> or </em><em>None</em>) – Use this option if you have additional targets beyond the main gene
for which we are analyzing variants. This might be the case if you
have spiked other genes into the library. If this option is set to
something other than <cite>None</cite>, then there must be a column in
<cite>barcode_variant_file</cite> named “target” and one of these targets must
be equal to ‘primary_target’. If there are other targets, they should
<strong>not</strong> have any substitutions as we don’t parse substitutions in
non-primary targets. Instead, <cite>substitutions_col</cite> for secondary targets
should be empty or just have the name of the secondary target.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.geneseq">
<span class="sig-name descname"><span class="pre">geneseq</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.geneseq" title="Link to this definition">¶</a></dt>
<dd><p>Wild-type sequence passed at initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.sites">
<span class="sig-name descname"><span class="pre">sites</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.sites" title="Link to this definition">¶</a></dt>
<dd><p>List of all codon sites in 1, 2, … numbering.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.codons">
<span class="sig-name descname"><span class="pre">codons</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.codons" title="Link to this definition">¶</a></dt>
<dd><p><cite>codons[r]</cite> is wildtype codon at site <cite>r</cite>, ordered by sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>collections.OrderedDict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.aas">
<span class="sig-name descname"><span class="pre">aas</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.aas" title="Link to this definition">¶</a></dt>
<dd><p><cite>aas[r]</cite> is wildtype amino acid at site <cite>r</cite>, ordered by sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>collections.OrderedDict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.libraries">
<span class="sig-name descname"><span class="pre">libraries</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.libraries" title="Link to this definition">¶</a></dt>
<dd><p>List of libraries in <cite>barcode_variant_file</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.barcode_variant_df">
<span class="sig-name descname"><span class="pre">barcode_variant_df</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.barcode_variant_df" title="Link to this definition">¶</a></dt>
<dd><p>Info about codon mutations parsed from <cite>barcode_variant_file</cite>.
For non-primary targets, the mutations column just hold the
target name.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.variant_count_df">
<span class="sig-name descname"><span class="pre">variant_count_df</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.variant_count_df" title="Link to this definition">¶</a></dt>
<dd><p>Initially <cite>None</cite>, but after data added with
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts" title="dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.addSampleCounts</span></code></a>, holds counts of
each variant for each sample. Differs from <cite>barcode_variant_df</cite>
in that the former just holds barcode-variant definitions,
whereas <cite>variant_count_df</cite> has counts for each sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.primary_target">
<span class="sig-name descname"><span class="pre">primary_target</span></span><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.primary_target" title="Link to this definition">¶</a></dt>
<dd><p>If multiple targets, name of the main target for which we are
calling variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">addMergedLibraries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_lib</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all</span> <span class="pre">libraries'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.addMergedLibraries"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries" title="Link to this definition">¶</a></dt>
<dd><p>Add data to <cite>df</cite> for all libraries merged.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame that includes columns named ‘library’ and ‘barcode’.</p></li>
<li><p><strong>all_lib</strong> (<em>str</em>) – Name given to library that is merge of all other libraries.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>df</cite> only has data for one library, just returns <cite>df</cite>. Otherwise
returns copy of <cite>df</cite> with new library with name given by <cite>all_lib</cite>
that contains data for all individual libraries with the ‘barcode’
column giving library name followed by a hyphen and barcode.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts">
<span class="sig-name descname"><span class="pre">addSampleCounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">library</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcodecounts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.addSampleCounts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts" title="Link to this definition">¶</a></dt>
<dd><p>Add variant counts for a sample to <cite>variant_count_df</cite>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you have many samples to add at once, it is faster to use
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df" title="dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.add_sample_counts_df()</span></code></a> rather than
repeatedly calling this method.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>library</strong> (<em>str</em>) – Valid <cite>library</cite> for the <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</p></li>
<li><p><strong>sample</strong> (<em>str</em>) – Sample name, must <strong>not</strong> already be in
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.samples" title="dms_variants.codonvarianttable.CodonVariantTable.samples"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.samples</span></code></a> for <cite>library</cite>.</p></li>
<li><p><strong>barcodecounts</strong> (<em>pandas.DataFrame</em>) – Counts for each variant by barcode. Must have columns “barcode”
and “count”. The “barcode” column must have all barcodes in
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes" title="dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.valid_barcodes</span></code></a> for <cite>library</cite>. Such
data frames are returned by
<a class="reference internal" href="dms_variants.illuminabarcodeparser.html#dms_variants.illuminabarcodeparser.IlluminaBarcodeParser" title="dms_variants.illuminabarcodeparser.IlluminaBarcodeParser"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dms_variants.illuminabarcodeparser.IlluminaBarcodeParser</span></code></a>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.add_frac_counts">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">add_frac_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_count_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.add_frac_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.add_frac_counts" title="Link to this definition">¶</a></dt>
<dd><p>Add fraction of counts from each variant in library/sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>variant_count_df</strong> (<em>pandas.DataFrame</em>) – Same format as <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-attr docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of <cite>variant_count_df</cite> with added column ‘frac_counts’
that gives fraction of all counts in that library / sample for
that variant.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
<span class="gp">... </span>       <span class="p">[(</span><span class="s1">&#39;lib1&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;lib1&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;lib1&#39;</span><span class="p">,</span> <span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;GG&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;lib1&#39;</span><span class="p">,</span> <span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;GA&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;lib2&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;CC&#39;</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;lib2&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;GA&#39;</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
<span class="gp">... </span>        <span class="p">],</span>
<span class="gp">... </span>       <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">add_frac_counts</span><span class="p">(</span><span class="n">variant_count_df</span><span class="p">)</span>
<span class="go">  library sample barcode  count  frac_counts</span>
<span class="go">0    lib1     s1      AA      1         0.25</span>
<span class="go">1    lib1     s1      AT      3         0.75</span>
<span class="go">2    lib1     s2      GG      0         0.00</span>
<span class="go">3    lib1     s2      GA     10         1.00</span>
<span class="go">4    lib2     s1      CC      5         0.50</span>
<span class="go">5    lib2     s1      GA      5         0.50</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.add_full_seqs">
<span class="sig-name descname"><span class="pre">add_full_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_seq_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aa_sequence'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon_seq_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'codon_sequence'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.add_full_seqs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.add_full_seqs" title="Link to this definition">¶</a></dt>
<dd><p>Add full sequences to data frame, for primary target only.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame that specifies substitutions. Might be the
<cite>barcode_variant_df</cite> or <cite>variant_count_df</cite> attributes,
or the result of calling <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.func_scores" title="dms_variants.codonvarianttable.CodonVariantTable.func_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.func_scores()</span></code></a>.</p></li>
<li><p><strong>aa_seq_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with amino-acid sequences. In order to add,
<cite>df</cite> must have column named ‘aa_substitutions’.</p></li>
<li><p><strong>codon_seq_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with codon sequences. In order to add,
<cite>df</cite> must have column named ‘codon_substitutions’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Copy of <cite>df</cite> with columns <cite>aa_seq_col</cite> and/or <cite>codon_seq_col</cite>,
and only the primary target retained if there is a ‘target’
column in <cite>df</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df">
<span class="sig-name descname"><span class="pre">add_sample_counts_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">counts_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.add_sample_counts_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df" title="Link to this definition">¶</a></dt>
<dd><p>Add variant counts for several samples to <cite>variant_count_df</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>counts_df</strong> (<em>pandas.DataFrame</em>) – Must have columns ‘library’, ‘sample’, ‘barcode’, and ‘count’.
The sample must <strong>not</strong> already be in <cite>CodonVariantTable.samples</cite>
for that library. The barcode columns must have all barcodes for
that library including zero-count ones.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.avgCountsPerVariant">
<span class="sig-name descname"><span class="pre">avgCountsPerVariant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.avgCountsPerVariant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.avgCountsPerVariant" title="Link to this definition">¶</a></dt>
<dd><p>Get average counts per variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Include all libraries including a merge, only a merge of all
libraries, or a list of libraries.</p></li>
<li><p><strong>samples</strong> (<em>{'all'</em><em>, </em><em>list}</em>) – Include all samples or just samples in list.</p></li>
<li><p><strong>min_support</strong> (<em>int</em>) – Only include variants with variant call support &gt;= this.</p></li>
<li><p><strong>sample_rename</strong> (<em>dict</em><em> or </em><em>None</em>) – Rename samples by specifying original name as key and new name
as value.</p></li>
<li><p><strong>by_target</strong> (<em>bool</em>) – If <cite>True</cite>, also group counts by target if multiple targets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Average counts per variant for each library and sample, and
possibly target.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.classifyVariants">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">classifyVariants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_class_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variant_class'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">syn_as_wt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_primary_target_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'secondary</span> <span class="pre">target'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">class_as_categorical</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.classifyVariants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.classifyVariants" title="Link to this definition">¶</a></dt>
<dd><p>Classifies codon variants in <cite>df</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Must have columns named ‘aa_substitutions’, ‘n_aa_substitutions’,
and ‘n_codon_substitutions’. Such a data frame can be obtained via
<cite>variant_count_df</cite> or <cite>barcode_variant_df</cite> attributes of a
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> or via
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.func_scores" title="dms_variants.codonvarianttable.CodonVariantTable.func_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.func_scores()</span></code></a>.</p></li>
<li><p><strong>variant_class_col</strong> (<em>str</em>) – Name of column added to <cite>df</cite> that contains variant classification.
Overwritten if already exists.</p></li>
<li><p><strong>max_aa</strong> (<em>int</em>) – When classifying, group all with &gt;= this many amino-acid mutations.</p></li>
<li><p><strong>syn_as_wt</strong> (<em>bool</em>) – Do not have a category of ‘synonymous’ and instead classify
synonymous variants as ‘wildtype’. If using this option, <cite>df</cite>
does not need column ‘n_codon_substitutions’.</p></li>
<li><p><strong>primary_target</strong> (<em>None</em><em> or </em><em>str</em>) – If <cite>df</cite> has a column named ‘target’, then this must specify
primary target (e.g., <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.primary_target" title="dms_variants.codonvarianttable.CodonVariantTable.primary_target"><code class="xref py py-attr docutils literal notranslate"><span class="pre">CodonVariantTable.primary_target</span></code></a>).
Variants not from primary target are classified as
<cite>non_primary_target_class</cite>.</p></li>
<li><p><strong>non_primary_target_class</strong> (<em>str</em>) – Classification used for non-primary targets.</p></li>
<li><p><strong>class_as_categorical</strong> (<em>bool</em>) – Return <cite>variant_class</cite> as a categorical variable with a
reasonable ordering.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Copy of <cite>df</cite> with column specified by <cite>variant_class_col</cite> as:</dt><dd><ul class="simple">
<li><p>’wildtype’: no codon mutations</p></li>
<li><p>’synonymous’: only synonymous codon mutations</p></li>
<li><p>’stop’: at least one stop-codon mutation</p></li>
<li><p>’{n_aa} nonsynonymous’ where <cite>n_aa</cite> is number of amino-acid
mutations, or is ‘&gt;{max_aa}’ if more than <cite>max_aa</cite>.</p></li>
<li><p>potentially <cite>non_primary_target_class</cite>.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[(</span><span class="s1">&#39;AAA&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;AAG&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;ATA&#39;</span><span class="p">,</span> <span class="s1">&#39;M1* G5K&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;GAA&#39;</span><span class="p">,</span> <span class="s1">&#39;G5H&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CTT&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C G5C&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CTT&#39;</span><span class="p">,</span> <span class="s1">&#39;M1A L3T G5C&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;AAG&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">],</span>
<span class="gp">... </span>        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                 <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;n_codon_substitutions&#39;</span><span class="p">]</span>
<span class="gp">... </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_classify</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df_classify</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="s1">&#39;n_codon_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="s1">&#39;variant_class&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_classify</span><span class="p">[[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_class&#39;</span><span class="p">]]</span>
<span class="go">  barcode     variant_class</span>
<span class="go">0     AAA          wildtype</span>
<span class="go">1     AAG        synonymous</span>
<span class="go">2     ATA              stop</span>
<span class="go">3     GAA   1 nonsynonymous</span>
<span class="go">4     CTT  &gt;1 nonsynonymous</span>
<span class="go">5     CTT  &gt;1 nonsynonymous</span>
<span class="go">6     AAG        synonymous</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_syn_as_wt</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">df</span><span class="p">,</span>
<span class="gp">... </span>                                                  <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_syn_as_wt</span><span class="p">[[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_class&#39;</span><span class="p">]]</span>
<span class="go">  barcode     variant_class</span>
<span class="go">0     AAA          wildtype</span>
<span class="go">1     AAG          wildtype</span>
<span class="go">2     ATA              stop</span>
<span class="go">3     GAA   1 nonsynonymous</span>
<span class="go">4     CTT  &gt;1 nonsynonymous</span>
<span class="go">5     CTT  &gt;1 nonsynonymous</span>
<span class="go">6     AAG          wildtype</span>
</pre></div>
</div>
<p>Now show how we need to specify how to handle when multiple targets:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;target&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;secondary&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;primary&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="gp">... </span> <span class="p">[[</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_class&#39;</span><span class="p">]])</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">`df` has &quot;target&quot; so give `primary_target`</span>
</pre></div>
</div>
<p>We need to specify how to handle targets:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span>
<span class="gp">... </span>                        <span class="n">df</span><span class="p">,</span>
<span class="gp">... </span>                        <span class="n">primary_target</span><span class="o">=</span><span class="s1">&#39;primary&#39;</span><span class="p">,</span>
<span class="gp">... </span>                        <span class="n">non_primary_target_class</span><span class="o">=</span><span class="s1">&#39;homolog&#39;</span><span class="p">)</span>
<span class="gp">... </span> <span class="p">[[</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_class&#39;</span><span class="p">]])</span>
<span class="go">      target barcode     variant_class</span>
<span class="go">0  secondary     AAA           homolog</span>
<span class="go">1    primary     AAG        synonymous</span>
<span class="go">2    primary     ATA              stop</span>
<span class="go">3    primary     GAA   1 nonsynonymous</span>
<span class="go">4    primary     CTT  &gt;1 nonsynonymous</span>
<span class="go">5    primary     CTT  &gt;1 nonsynonymous</span>
<span class="go">6    primary     AAG        synonymous</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.codonToAAMuts">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">codonToAAMuts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_mut_str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allowgaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.codonToAAMuts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.codonToAAMuts" title="Link to this definition">¶</a></dt>
<dd><p>Convert string of codon mutations to amino-acid mutations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>codon_mut_str</strong> (<em>str</em>) – Codon mutations, delimited by a space and in 1, … numbering.</p></li>
<li><p><strong>allowgaps</strong> (<em>bool</em>) – Allow gap codon (<code class="docutils literal notranslate"><span class="pre">---</span></code>) translated to gap amino acid (<code class="docutils literal notranslate"><span class="pre">--</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Amino acid mutations in 1, … numbering.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">codonToAAMuts</span><span class="p">(</span><span class="s1">&#39;ATG1GTG GGA2GGC TGA3AGA&#39;</span><span class="p">)</span>
<span class="go">&#39;M1V *3R&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.escape_scores">
<span class="sig-name descname"><span class="pre">escape_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'barcode'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logbase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">floor_B</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">floor_E</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ceil_B</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ceil_E</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.escape_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.escape_scores" title="Link to this definition">¶</a></dt>
<dd><p>Compute a score that represents escape from binding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Here we couch the explanation in terms of variant escape from
antibody binding.</p>
<p>Let <span class="math notranslate nohighlight">\(v\)</span> be a variant, let <span class="math notranslate nohighlight">\(B_v\)</span> be the fraction of this
variant that is bound by antibody, and let <span class="math notranslate nohighlight">\(E_v = 1 - B_v\)</span> be
the fraction that escapes binding. A variant that completely escapes
binding has <span class="math notranslate nohighlight">\(B_v = 0\)</span> and <span class="math notranslate nohighlight">\(E_v = 1\)</span>; a variant that is
completely bound has <span class="math notranslate nohighlight">\(B_v = 1\)</span> and <span class="math notranslate nohighlight">\(E_v = 0\)</span>.</p>
<p>We define the escape score <span class="math notranslate nohighlight">\(s_v\)</span> in one of three ways depending
on the value of <cite>score_type</cite> parameter:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>As the escape fraction, so <span class="math notranslate nohighlight">\(s_v = E_v\)</span>.</p></li>
<li><p>As the log of the escape fraction, so
<span class="math notranslate nohighlight">\(s_v = \log_b E_v\)</span> where <span class="math notranslate nohighlight">\(b\)</span> is the logarithm base.</p></li>
<li><p>As minus the log of the binding fraction, so
<span class="math notranslate nohighlight">\(s_v = -\log_b B_v\)</span>.</p></li>
</ol>
</div></blockquote>
<p>In all cases, larger values of <span class="math notranslate nohighlight">\(s_v\)</span> indicate more escape.</p>
<p>We calculate <span class="math notranslate nohighlight">\(E_v\)</span> as follows. Let <span class="math notranslate nohighlight">\(f_v^{\rm{pre}}\)</span> be
the frequency of <span class="math notranslate nohighlight">\(v\)</span> prior  to selection for binding (so
<span class="math notranslate nohighlight">\(\sum_v f_v^{\rm{pre}} = 1\)</span>). Then the fraction of the library
that is <span class="math notranslate nohighlight">\(v\)</span> <strong>after</strong> selecting for unbound variants is</p>
<div class="math notranslate nohighlight">
\[f_v^{\rm{post}} =
\frac{f_v^{\rm{pre}} \times E_v}
     {\sum_{v'} f_{v'}^{\rm{pre}} \times E_{v'}}.\]</div>
<p>Note that the denominator of the above equation,
<span class="math notranslate nohighlight">\(F = \sum_v f_v^{\rm{pre}} \times E_v\)</span>,
represents the overall fraction of the library that escapes binding,
<strong>which we assume is directly measured experimentally</strong>.</p>
<p>We can easily solve for <span class="math notranslate nohighlight">\(E_v\)</span> as</p>
<div class="math notranslate nohighlight">
\[E_v = \frac{F \times f_v^{\rm{post}}}{f_v^{\rm{pre}}},\]</div>
<p>and can similarly obtain <span class="math notranslate nohighlight">\(B_v\)</span> from <span class="math notranslate nohighlight">\(B_v = 1 - E_v\)</span>.</p>
<p>We calculate <span class="math notranslate nohighlight">\(E_v\)</span> directly from the actual counts of the
variants pre- and post-selection. Let <span class="math notranslate nohighlight">\(n_v^{\rm{pre}}\)</span> and
<span class="math notranslate nohighlight">\(n_v^{\rm{post}}\)</span> be the counts of variant <span class="math notranslate nohighlight">\(v\)</span> pre-
and post-selection <strong>after</strong> adding a pseudocount of <span class="math notranslate nohighlight">\(P \ge 0\)</span>,
and let <span class="math notranslate nohighlight">\(N^{\rm{pre}} = \sum_v n_v^{\rm{pre}}\)</span> and
<span class="math notranslate nohighlight">\(N^{\rm{post}} = \sum_v n_v^{\rm{post}}\)</span> be the total counts of
all variants pre- and post-selection. Then:</p>
<div class="math notranslate nohighlight">
\[E_v
=
F \times \frac{n_v^{\rm{post}} N^{\rm{pre}}}
              {n_v^{\rm{pre}} N^{\rm{post}}}\]</div>
<p>A complication is that <span class="math notranslate nohighlight">\(s_v\)</span> can be undefined for certain values
of <span class="math notranslate nohighlight">\(E_v\)</span> or <span class="math notranslate nohighlight">\(B_v\)</span>. Specifically, when using the log escape
fraction definition (<span class="math notranslate nohighlight">\(s_v = \log_b E_v\)</span>) then <span class="math notranslate nohighlight">\(s_v\)</span> is
undefined if <span class="math notranslate nohighlight">\(E_v = 0\)</span>, so we put a floor on <span class="math notranslate nohighlight">\(E_v\)</span> by
defining <span class="math notranslate nohighlight">\(s_v = \log_b \max\left(E_v, E_{\rm{floor}}\right)\)</span>.
When using the minus log binding fraction definition
(<span class="math notranslate nohighlight">\(s_v = -\log_b B_v\)</span>), then <span class="math notranslate nohighlight">\(s_v\)</span> is undefined if
<span class="math notranslate nohighlight">\(B_v \le 0\)</span>, so we put a floor on <span class="math notranslate nohighlight">\(B_v\)</span> by defining
defining <span class="math notranslate nohighlight">\(s_v = -\log_b \max\left(B_v, B_{\rm{floor}}\right)\)</span>.
We similarly define ceilings on <span class="math notranslate nohighlight">\(E_v\)</span> and <span class="math notranslate nohighlight">\(B_v\)</span>, although
these ceilings are optional.</p>
<p>We also estimate the variance <span class="math notranslate nohighlight">\(\sigma_{s_v}^2\)</span> on
<span class="math notranslate nohighlight">\(s_v\)</span> from the variances on the counts, which we assume
are <span class="math notranslate nohighlight">\(\sigma_{n_v^{\rm{pre}}}^2 = n_v^{\rm{pre}}\)</span> and
<span class="math notranslate nohighlight">\(\sigma_{n_v^{\rm{post}}}^2 = n_v^{\rm{post}}\)</span> from
Poisson counting statistics. To do this, we propagate the errors:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\sigma_{s_v}^2
&amp;=&amp;
\left(\frac{\partial s_v}{\partial n_v^{\rm{pre}}}\right)^2
\sigma_{n_v^{\rm{pre}}}^2 +
\left(\frac{\partial s_v}{\partial n_v^{\rm{post}}}\right)^2
\sigma_{n_v^{\rm{post}}}^2 \\
&amp;=&amp;
\left(\frac{\partial s_v}{\partial n_v^{\rm{pre}}}\right)^2
n_v^{\rm{pre}} +
\left(\frac{\partial s_v}{\partial n_v^{\rm{post}}}\right)^2
n_v^{\rm{post}}.\end{split}\]</div>
<p>We calculate the derivatives of <span class="math notranslate nohighlight">\(s_v\)</span> with respect to the counts
numerically with a step size of one rather than analytically, since
analytical calculations are confounded by the floors on <span class="math notranslate nohighlight">\(E_v\)</span>
or <span class="math notranslate nohighlight">\(B_v\)</span>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_df</strong> (<em>pandas.DataFrame</em>) – Comparisons we use to compute the functional scores. Should have
these columns: ‘pre_sample’ (pre-selection sample), ‘post_sample’
(post-selection sample), ‘library’, ‘name’ (name for output),
‘frac_escape’ (the overall fraction escaping <span class="math notranslate nohighlight">\(F\)</span>).</p></li>
<li><p><strong>score_type</strong> (<em>{'frac_escape'</em><em>, </em><em>'log_escape'</em><em>, </em><em>'minus_log_bind'}</em>) – How to define escape score: <span class="math notranslate nohighlight">\(E_v\)</span> if ‘frac_escape’;
<span class="math notranslate nohighlight">\(\log_b E_v\)</span> if ‘log_escape’; <span class="math notranslate nohighlight">\(-\log_b B_v\)</span> if
‘minus_log_bind’.</p></li>
<li><p><strong>pseudocount</strong> (<em>float</em>) – Pseudocount added to each count.</p></li>
<li><p><strong>by</strong> (<em>{'barcode'</em><em>, </em><em>'aa_substitutions'</em><em>, </em><em>'codon_substitutions'}</em>) – Compute effects for each barcode”, set of amino-acid substitutions,
or set of codon substitutions. In the last two cases, all barcodes
with each set of substitutions are combined.</p></li>
<li><p><strong>logbase</strong> (<em>float</em>) – Base for logarithm when calculating functional score.</p></li>
<li><p><strong>floor_B</strong> (<em>float</em>) – Floor assigned to <span class="math notranslate nohighlight">\(B_v\)</span>, <span class="math notranslate nohighlight">\(B_{\rm{floor}}\)</span>
if <cite>score_type</cite> is ‘minus_log_bind’.</p></li>
<li><p><strong>floor_E</strong> (<em>float</em>) – Floor assigned to <span class="math notranslate nohighlight">\(E_v\)</span>, <span class="math notranslate nohighlight">\(E_{\rm{floor}}\)</span>
if <cite>score_type</cite> is ‘frac_escape’ or ‘log_escape’.</p></li>
<li><p><strong>ceil_B</strong> (<em>float</em><em> or </em><em>None</em>) – Ceiling assigned to <span class="math notranslate nohighlight">\(B_v\)</span>, or <cite>None</cite> if no ceiling.</p></li>
<li><p><strong>ceil_E</strong> (<em>float</em><em> or </em><em>None</em>) – Ceiling assigned to <span class="math notranslate nohighlight">\(E_v\)</span>, or <cite>None</cite> if no ceiling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Has the following columns:</dt><dd><ul class="simple">
<li><p>’name’: specified in <cite>sample_df</cite></p></li>
<li><p>’library’: the library</p></li>
<li><p>’pre_sample’: specified in <cite>sample_df</cite></p></li>
<li><p>’post_sample’: specified in <cite>sample_df</cite></p></li>
<li><p>the grouping used to compute scores (the value of <cite>by</cite>)</p></li>
<li><p>’score’: <span class="math notranslate nohighlight">\(s_v\)</span></p></li>
<li><p>’score_var’: <span class="math notranslate nohighlight">\(\sigma_{s_v}^2\)</span></p></li>
<li><p>’pre_count’: <span class="math notranslate nohighlight">\(n_v^{\rm{pre}}\)</span> (without pseudocount)</p></li>
<li><p>’post_count’: <span class="math notranslate nohighlight">\(n_v^{\rm{post}}\)</span> (without pseudocount)</p></li>
<li><p>as many of ‘aa_substitutions’, ‘n_aa_substitutions’,
‘codon_substitutions’, and ‘n_codon_substitutions’ as
makes sense to retain given value of <cite>by</cite>.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.from_variant_count_df">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_variant_count_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_count_df_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geneseq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_all_libs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allowgaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.from_variant_count_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.from_variant_count_df" title="Link to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> from CSV of <cite>variant_count_df</cite>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Use this method when you have written a CSV file of <cite>variant_count_df</cite>
attribute of a <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>, and wish to re-initialize.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_count_df_file</strong> (<em>str</em>) – Name of CSV file containing the <cite>variant_count_df</cite>. Must have
following columns: “barcode”, “library”, “variant_call_support”,
“codon_substitutions”, “sample”, and “count”.</p></li>
<li><p><strong>geneseq</strong> (<em>str</em>) – Sequence of wildtype protein-coding gene.</p></li>
<li><p><strong>drop_all_libs</strong> (<em>bool</em>) – If there is a library named “all libraries”, drop it as it probably
added by <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries" title="dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.addMergedLibraries()</span></code></a> and
duplicates information for the individual libraries.</p></li>
<li><p><strong>allowgaps</strong> (<em>bool</em>) – Meaning described in main <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> doc string.</p></li>
<li><p><strong>primary_target</strong> (<em>None</em><em> or </em><em>str</em>) – Meaning described in main <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> doc string.</p></li>
<li><p><strong>extra_cols</strong> (<em>list</em>) – Meaning described in main <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> doc string.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.func_scores">
<span class="sig-name descname"><span class="pre">func_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">preselection</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'barcode'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">syn_as_wt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logbase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permit_zero_wt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permit_self_comp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.func_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.func_scores" title="Link to this definition">¶</a></dt>
<dd><p>Get data frame with functional scores for variants.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The functional score is calculated from the change in counts
for a variant pre- and post-selection using the formula in
<a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al (2018)</a>.</p>
<p>Specifically, let <span class="math notranslate nohighlight">\(n^v_{pre}\)</span> and <span class="math notranslate nohighlight">\(n^v_{post}\)</span> be
the counts of variant <span class="math notranslate nohighlight">\(v\)</span> pre- and post-selection, and
let <span class="math notranslate nohighlight">\(n^{wt}_{pre}\)</span> and <span class="math notranslate nohighlight">\(n^{wt}_{post}\)</span> be
the summed counts of <strong>all</strong> wildtype variants pre- and post-
selection.</p>
<p>Then the functional score of the variant is:</p>
<div class="math notranslate nohighlight">
\[f_v = \log_b\left(\frac{n^v_{post} /
n^{wt}_{post}}{n^v_{pre} / n^{wt}_{pre}}\right).\]</div>
<p>The variance due to Poisson sampling statistics is:</p>
<div class="math notranslate nohighlight">
\[\sigma^2_v = \frac{1}{\left(\ln b\right)^2}
\left(\frac{1}{n^v_{post}} + \frac{1}{n^{wt}_{post}} +
\frac{1}{n^v_{pre}} + \frac{1}{n^{wt}_{pre}}\right)\]</div>
<p>where <span class="math notranslate nohighlight">\(b\)</span> is logarithm base (see <cite>logbase</cite> parameter).</p>
<p>For both calculations, a pseudocount (see <cite>pseudocount</cite> parameter)
is added to each count first. The wildtype counts are computed
across all <strong>fully wildtype</strong> variants (see <cite>syn_as_wt</cite> for
how this is defined).</p>
<p>If there are multiple targets, the functional scores for all
targets are relative to the wildtype of the primary target.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>preselection</strong> (<em>str</em><em> or </em><em>dict</em>) – Pre-selection sample. If the same for all post-selection then
provide the name as str. If it differs among post-selection
samples, then provide a dict keyed by each post-selection
sample with the pre-selection sample being the value.</p></li>
<li><p><strong>pseudocount</strong> (<em>float</em>) – Pseudocount added to each count.</p></li>
<li><p><strong>by</strong> (<em>{'barcode'</em><em>, </em><em>'aa_substitutions'</em><em>, </em><em>'codon_substitutions'}</em>) – Compute effects for each barcode”, set of amino-acid substitutions,
or set of codon substitutions. In the last two cases, all barcodes
with each set of substitutions are combined. If you use
“aa_substitutions” then it may be more sensible to set
<cite>syn_as_wt</cite> to <cite>True</cite>.</p></li>
<li><p><strong>syn_as_wt</strong> (<em>bool</em>) – In formula for functional scores, consider variants with only
synonymous mutations when determining wildtype counts? If <cite>False</cite>,
only variants with <strong>no</strong> mutations of any type contribute.</p></li>
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Perform calculation for all libraries including a merge
(named “all libraries”), only for the merge of all libraries,
or for the libraries in the list. If <cite>by</cite> is ‘barcode’, then
the barcodes for the merge have the library name pre-pended.</p></li>
<li><p><strong>logbase</strong> (<em>float</em>) – Base for logarithm when calculating functional score.</p></li>
<li><p><strong>permit_zero_wt</strong> (<em>bool</em>) – If the wildtype counts are zero for any sample, raise an error
or permit the calculation to proceed just using pseudocount?</p></li>
<li><p><strong>permit_self_comp</strong> (<em>bool</em>) – Permit comparisons for sample pre- and post-selection?</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Has the following columns:</dt><dd><ul class="simple">
<li><p>”library”: the library</p></li>
<li><p>”pre_sample”: the pre-selection sample</p></li>
<li><p>”post_sample”: the post-selection sample</p></li>
<li><p>value corresponding to grouping used to compute effects (<cite>by</cite>)</p></li>
<li><p>”func_score”: the functional score</p></li>
<li><p>”func_score_var”: variance on the functional score</p></li>
<li><p>”pre_count”: pre-selection counts</p></li>
<li><p>”post_count: post-selection counts</p></li>
<li><p>”pre_count_wt”: pre-selection counts for all wildtype</p></li>
<li><p>”post_count_wt”: post-selection counts for all wildtype</p></li>
<li><p>”pseudocount”: the pseudocount value</p></li>
<li><p>as many of “aa_substitutions”, “n_aa_substitutions”,
“codon_substitutions”, and “n_codon_substitutions”
as can be retained given value of <cite>by</cite>.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.mutCounts">
<span class="sig-name descname"><span class="pre">mutCounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.mutCounts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.mutCounts" title="Link to this definition">¶</a></dt>
<dd><p>Get counts of each individual mutations (only in primary target).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include just single mutants, or all mutants?</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tidy data frame with columns named “library”, “sample”,
“mutation”, “count”, “mutation_type”, and “site”. If there
are multiple targets, only returns counts for the primary
target.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.n_variants_df">
<span class="sig-name descname"><span class="pre">n_variants_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.n_variants_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.n_variants_df" title="Link to this definition">¶</a></dt>
<dd><p>Get number variants per library / sample (and target if specified).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;= 1 <cite>mut_type</cite> mutation.</p></li>
<li><p><strong>mut_type</strong> (<em>{'aa'</em><em>, </em><em>'codon'</em><em>, </em><em>None}</em>) – If <cite>variant_type</cite> is ‘single’, indicate what type of single
mutants we are filtering for.</p></li>
<li><p><strong>primary_target_only</strong> (<em>bool</em>) – Only return counts for the primary target.</p></li>
<li><p><strong>args</strong> (<em>All other</em>) – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.numCodonMutsByType">
<span class="sig-name descname"><span class="pre">numCodonMutsByType</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.numCodonMutsByType"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.numCodonMutsByType" title="Link to this definition">¶</a></dt>
<dd><p>Get average nonsynonymous, synonymous, stop mutations per variant.</p>
<p>These statistics are only for the primary target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>all_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumCodonMutsByType()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Data frame with average mutations of each type per variant.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotAvgCountsPerVariant">
<span class="sig-name descname"><span class="pre">plotAvgCountsPerVariant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotAvgCountsPerVariant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotAvgCountsPerVariant" title="Link to this definition">¶</a></dt>
<dd><p>Plot average counts per variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Include all libraries including a merge, only a merge of all
libraries, or a list of libraries.</p></li>
<li><p><strong>samples</strong> (<em>{'all'</em><em>, </em><em>list}</em>) – Include all samples or just samples in list.</p></li>
<li><p><strong>by_target</strong> (<em>bool</em>) – If <cite>True</cite>, also group counts by target if multiple targets.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotCountsPerVariant">
<span class="sig-name descname"><span class="pre">plotCountsPerVariant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ystat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'frac_counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_variant_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classifyVariants_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aa'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotCountsPerVariant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCountsPerVariant" title="Link to this definition">¶</a></dt>
<dd><p>Plot variant index versus counts (or frac counts).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ystat</strong> (<em>{'frac_counts'</em><em>, </em><em>'count'}</em>) – Is y-axis counts from variant, or fraction of counts in
library / sample from variant?</p></li>
<li><p><strong>logy</strong> (<em>bool</em>) – Show the y-axis on a log scale. If so, all values of 0 are
set to half the minimum observed value &gt; 0, and dashed line
is drawn to indicate that points below it are not observed.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulVariantCounts" title="dms_variants.codonvarianttable.CodonVariantTable.plotCumulVariantCounts"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotCumulVariantCounts()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage">
<span class="sig-name descname"><span class="pre">plotCumulMutCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotCumulMutCoverage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="Link to this definition">¶</a></dt>
<dd><p>Frac mutations seen &lt;= some number of times (primary target only).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;=1 <cite>mut_type</cite> mutation.</p></li>
<li><p><strong>max_count</strong> (<em>None</em><em> or </em><em>int</em>) – Plot cumulative fraction plot out to this number of observations.
If <cite>None</cite>, a reasonable value is automatically determined.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotCumulVariantCounts">
<span class="sig-name descname"><span class="pre">plotCumulVariantCounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aa'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tot_variants_hline</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotCumulVariantCounts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulVariantCounts" title="Link to this definition">¶</a></dt>
<dd><p>Plot number variants with &gt;= that each number of counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;=1 <cite>mut_type</cite> mutation.</p></li>
<li><p><strong>tot_variants_hline</strong> (<em>bool</em>) – Include dotted horizontal line indicating total number of variants.</p></li>
<li><p><strong>primary_target_only</strong> (<em>bool</em>) – Only show counts for the primary target.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotMutFreqs">
<span class="sig-name descname"><span class="pre">plotMutFreqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotMutFreqs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotMutFreqs" title="Link to this definition">¶</a></dt>
<dd><p>Mutation frequency along length of gene (primary target only).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>parameters</strong> (<em>All</em>) – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotCumulMutCoverage()</span></code></a>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotMutHeatmap">
<span class="sig-name descname"><span class="pre">plotMutHeatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_or_frequency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'frequency'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotMutHeatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotMutHeatmap" title="Link to this definition">¶</a></dt>
<dd><p>Heatmap of mutation counts or frequencies (for primary target only).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>count_or_frequency</strong> (<em>{'count'</em><em>, </em><em>'frequency'}</em>) – Plot mutation counts or frequencies?</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotCumulMutCoverage()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType">
<span class="sig-name descname"><span class="pre">plotNumCodonMutsByType</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotNumCodonMutsByType"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType" title="Link to this definition">¶</a></dt>
<dd><p>Plot average nonsynonymous, synonymous, stop mutations per variant.</p>
<p>These statistics are only for the primary target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;=1 codon mutation.</p></li>
<li><p><strong>ylabel</strong> (<em>None</em><em> or </em><em>str</em>) – If not <cite>None</cite>, specify y-axis label (otherwise it is autoset).</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram">
<span class="sig-name descname"><span class="pre">plotNumMutsHistogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_muts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotNumMutsHistogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="Link to this definition">¶</a></dt>
<dd><p>Plot histogram of num mutations per variant (primary target only).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mut_type</strong> (<em>{'codon'</em><em> or </em><em>'aa'}</em>) – Mutation type.</p></li>
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Include all libraries including a merge, only a merge of all
libraries, or a list of libraries.</p></li>
<li><p><strong>samples</strong> (<em>{'all'</em><em>, </em><em>None</em><em>, </em><em>list}</em>) – Include all samples, a list of samples, or <cite>None</cite> to just count
each barcoded variant once.</p></li>
<li><p><strong>plotfile</strong> (<em>None</em><em> or </em><em>str</em>) – Name of file to which to save plot.</p></li>
<li><p><strong>orientation</strong> (<em>{'h'</em><em>, </em><em>'v'}</em>) – Facet libraries horizontally or vertically?</p></li>
<li><p><strong>widthscale</strong> (<em>float</em>) – Expand width of plot by this factor.</p></li>
<li><p><strong>heightscale</strong> (<em>float</em>) – Expand height of plot by this factor.</p></li>
<li><p><strong>min_support</strong> (<em>int</em>) – Only include variants with variant call support &gt;= this.</p></li>
<li><p><strong>max_muts</strong> (<em>int</em><em> or </em><em>None</em>) – Group together all variants with &gt;= this many mutations.</p></li>
<li><p><strong>sample_rename</strong> (<em>dict</em><em> or </em><em>None</em>) – Rename samples by specifying original name as key and new name
as value.</p></li>
<li><p><strong>one_lib_facet</strong> (<em>bool</em>) – Plot library facet title even if just one library.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.plotVariantSupportHistogram">
<span class="sig-name descname"><span class="pre">plotVariantSupportHistogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">libraries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plotfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widthscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">heightscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_support</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_lib_facet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotVariantSupportHistogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotVariantSupportHistogram" title="Link to this definition">¶</a></dt>
<dd><p>Plot histogram of variant call support for variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>max_support</strong> (<em>int</em><em> or </em><em>None</em>) – Group together all variants with &gt;= this support.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
<li><p><strong>primary_target_only</strong> (<em>bool</em>) – Only include variants that are of the primary target.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.prob_escape">
<span class="sig-name descname"><span class="pre">prob_escape</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selections_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neut_standard_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'neut_standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'barcode'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_neut_standard_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_neut_standard_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ceil_n_aa_substitutions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_neut_standard_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_target_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.prob_escape"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.prob_escape" title="Link to this definition">¶</a></dt>
<dd><p>Compute probability of escape relative to a neutralization standard.</p>
<p>Assumes one of the targets is the neutralization standard and is
unaffected by antibody selection. Then computes escape probability
of each variant as
<span class="math notranslate nohighlight">\(p_v = \left(n_v^{sel}/S^{sel}\right) / \left(n_v^0/S^0\right)\)</span>
where <span class="math notranslate nohighlight">\(n_v^{sel}\)</span> is the counts of variant <span class="math notranslate nohighlight">\(v\)</span> in the
selected condition, <span class="math notranslate nohighlight">\(n_v^0\)</span> is the counts of variant <span class="math notranslate nohighlight">\(v\)</span>
in the no-antibody contition, <span class="math notranslate nohighlight">\(S^{sel}\)</span> is the total counts
of the neutralization standard in the selected condition, and
<span class="math notranslate nohighlight">\(S^0\)</span> is the total counts of the neutralization standard in
the no-antibody condition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selections_df</strong> (<em>pandas.DataFrame</em>) – How to pair samples. Should have columns ‘library’,
‘antibody_sample’, and ‘no-antibody_sample’ for each row.
The ‘library’ and ‘antibody_sample’ columns must be unique
for each row.</p></li>
<li><p><strong>neut_standard_target</strong> (<em>str</em>) – Name of target that is neutralization standard.</p></li>
<li><p><strong>by</strong> (<em>{'barcode'</em><em>, </em><em>'aa_substitutions'</em><em>, </em><em>'codon_substitutions'}</em>) – Compute for each barcode”, set of amino-acid substitutions,
or set of codon substitutions. In the last two cases, all barcodes
with each set of substitutions are combined.</p></li>
<li><p><strong>min_neut_standard_frac</strong> (<em>float</em>) – Raise error if neutralization standard not at least
this fraction of the total counts for each library / sample.</p></li>
<li><p><strong>min_neut_standard_count</strong> (<em>int</em>) – Raise error if not at least this many counts for neutralization
standard for each library / sample.</p></li>
<li><p><strong>ceil_n_aa_substitutions</strong> (<em>int</em>) – When computing the returned <cite>neutralization</cite> data frame, group variants
with &gt;= this many amino-acid substitutions.</p></li>
<li><p><strong>drop_neut_standard_target</strong> (<em>bool</em>) – Drop the neutralization standard variant-level results from the
returned data frames.</p></li>
<li><p><strong>primary_target_only</strong> (<em>bool</em>) – Drop everything except the primary target and neutralization standard
target before beginning calculations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>This tuple consists of three pandas data frames:</p>
<ul>
<li><p><cite>prob_escape</cite>: gives the escape probabilities, and has the
following columns:</p>
<blockquote>
<div><ul class="simple">
<li><p>’library’</p></li>
<li><p>’antibody_sample’</p></li>
<li><p>’no-antibody_sample’</p></li>
<li><p>as many of ‘target’, ‘aa_substitutions’, ‘n_aa_substitutions’,
‘codon_substitutions’, and ‘n_codon_substitutions’ as
makes sense to retain given value of <cite>by</cite>.</p></li>
<li><p>’prob_escape’: censored to be between 0 and 1</p></li>
<li><p>’prob_escape_uncensored’: not censored to be between 0 and 1</p></li>
<li><p>’antibody_count’: counts for variant in antibody condition</p></li>
<li><p>’no-antibody_count’: counts for variant no-antibody condition</p></li>
<li><p>’antibody_neut_standard_count’: counts for neut standard</p></li>
<li><p>’no-antibody_neut_standard_count’: counts for neut standard</p></li>
</ul>
</div></blockquote>
</li>
<li><p><cite>neut_standard_fracs</cite> : a version of <cite>selections_df</cite> that also
has columns <cite>antibody_sample_frac</cite>, <cite>no-antibody_sample_frac</cite>,
<cite>antibody_sample_count</cite>, <cite>no-antibody_sample_count</cite> giving
total counts and fraction of neutralization standard for each
row in <cite>selections_df</cite>.</p></li>
<li><p><cite>neutralization</cite> for each target and number
of amino-acid (up to <cite>ceil_n_aa_substitutions</cite>), gives the overall
probability escape in a column <cite>prob_escape</cite>.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(prob_escape, neut_standard_fracs, neutralization)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.samples">
<span class="sig-name descname"><span class="pre">samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">library</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.samples" title="Link to this definition">¶</a></dt>
<dd><p>List of all samples for <cite>library</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>library</strong> (<em>str</em>) – Valid <cite>library</cite> for the <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All samples for which barcode counts have been added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.subs_to_seq">
<span class="sig-name descname"><span class="pre">subs_to_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subs_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'codon'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.subs_to_seq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.subs_to_seq" title="Link to this definition">¶</a></dt>
<dd><p>Convert substitutions to full sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>subs</strong> (<em>str</em>) – Space delimited substitutions in 1, … numbering.</p></li>
<li><p><strong>subs_type</strong> (<em>{'codon'</em><em>, </em><em>'aa'}</em>) – Are substitutions codon or amino acid?</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence created by these mutations, either codon or
amino acid depending on value of <cite>subs_type</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">geneseq</span> <span class="o">=</span> <span class="s1">&#39;ATGGGATGA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;library,barcode,substitutions,&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;variant_call_support&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">variants</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="p">(</span>
<span class="gp">... </span>                <span class="n">barcode_variant_file</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
<span class="gp">... </span>                <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span>
<span class="gp">... </span>                <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">subs_to_seq</span><span class="p">(</span><span class="s1">&#39;GGA2CGT ATG1GTG&#39;</span><span class="p">)</span>
<span class="go">&#39;GTGCGTTGA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">subs_to_seq</span><span class="p">(</span><span class="s1">&#39;*3W M1C&#39;</span><span class="p">,</span> <span class="n">subs_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">)</span>
<span class="go">&#39;CGW&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes">
<span class="sig-name descname"><span class="pre">valid_barcodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">library</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.valid_barcodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes" title="Link to this definition">¶</a></dt>
<dd><p>Set of valid barcodes for <cite>library</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>library</strong> (<em>str</em>) – Name of a valid library.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>set</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.CodonVariantTable.writeCodonCounts">
<span class="sig-name descname"><span class="pre">writeCodonCounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">single_or_all</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_all_libs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.writeCodonCounts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.writeCodonCounts" title="Link to this definition">¶</a></dt>
<dd><p>Write codon counts files for all libraries and samples.</p>
<p>Only writes the counts for the primary target.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Useful if you want to analyze individual mutations using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/">dms_tools2</a>. File format:
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html#counts-file">https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html#counts-file</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>single_or_all</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – If ‘single’, then counts just from single-codon mutants and
wildtype, and count all wildtype codons and just mutated codons
for single-codon mutants. If ‘all’, count all codons for all
variants at all sites. Appropriate if enrichment of each mutation
is supposed to represent its effect for ‘single’, and if enrichment
of mutation is supposed to represent its average effect across
genetic backgrounds in the library for ‘all’ provided mutations are
Poisson distributed.</p></li>
<li><p><strong>outdir</strong> (<em>None</em><em> or </em><em>str</em>) – Name of directory to write counts, created if it does not exist.
Use <cite>None</cite> to write to current directory.</p></li>
<li><p><strong>include_all_libs</strong> (<em>bool</em>) – Include data for a library (named “all-libraries”) that has
summmed data for all individual libraries if multiple libraries.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Gives names of created files. Has columns “library”, “sample”,
and “countfile”. The “countfile” columns gives name of the
created CSV file, <code class="docutils literal notranslate"><span class="pre">&lt;library&gt;_&lt;sample&gt;_codoncounts.csv</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dms_variants.codonvarianttable.filter_by_subs_observed">
<span class="sig-prename descclassname"><span class="pre">dms_variants.codonvarianttable.</span></span><span class="sig-name descname"><span class="pre">filter_by_subs_observed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_single_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_any_counts</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">and_vs_or</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'and'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subs_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aa_substitutions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_subs_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'n_aa_substitutions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#filter_by_subs_observed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dms_variants.codonvarianttable.filter_by_subs_observed" title="Link to this definition">¶</a></dt>
<dd><p>Filter for variants by observations substitutions in entire data frame.</p>
<p>Filter data frames of the type returned by <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> to
get just variants that have substitutions that are observed in some other
number of variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame containing variants as rows.</p></li>
<li><p><strong>min_single_counts</strong> (<em>int</em>) – Only keep variants with substitutions that are observed as single-
substitution variants in at least this many variants.</p></li>
<li><p><strong>min_any_counts</strong> (<em>int</em>) – Only keep variants with substitutions that are observed in at
least this many variants with any number of substitutions.</p></li>
<li><p><strong>and_vs_or</strong> (<em>{'and'</em><em>, </em><em>'or'}</em>) – Require variants to pass the <cite>min_single_counts</cite> <strong>and</strong> the
‘min_any_counts’ filters, or just require to pass one <strong>or</strong> the
other.</p></li>
<li><p><strong>subs_col</strong> (<em>'aa_substitutions'</em>) – Column in <cite>df</cite> with the substitutions as space-delimited strings.</p></li>
<li><p><strong>n_subs_col</strong> (<em>'n_aa_substitutions'</em>) – Column in <cite>df</cite> with the number of substitutions per variant.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of <cite>df</cite> that only contains the variants that pass the filters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Create a data frame of variants to filter:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">([</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;var1&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;var2&#39;</span><span class="p">,</span> <span class="s1">&#39;M1A&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;var3&#39;</span><span class="p">,</span> <span class="s1">&#39;M1A G2A&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;var4&#39;</span><span class="p">,</span> <span class="s1">&#39;M1A G2C&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;var5&#39;</span><span class="p">,</span> <span class="s1">&#39;G2A&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;var6&#39;</span><span class="p">,</span> <span class="s1">&#39;M1A&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;var7&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">],</span>
<span class="gp">... </span>         <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>Calling with <cite>min_single_counts=1</cite> and <cite>min_any_counts=1</cite> gets only
variants with substitutions that are observed in single-substitution
variants:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filter_by_subs_observed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">  barcode aa_substitutions  n_aa_substitutions</span>
<span class="go">0    var1                                    0</span>
<span class="go">1    var2              M1A                   1</span>
<span class="go">2    var3          M1A G2A                   2</span>
<span class="go">3    var5              G2A                   1</span>
<span class="go">4    var6              M1A                   1</span>
<span class="go">5    var7              M1C                   1</span>
</pre></div>
</div>
<p>We can also require substitutions to be seen multiple times as
single variants:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filter_by_subs_observed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">  barcode aa_substitutions  n_aa_substitutions</span>
<span class="go">0    var1                                    0</span>
<span class="go">1    var2              M1A                   1</span>
<span class="go">2    var6              M1A                   1</span>
</pre></div>
</div>
<p>Or that substitutions be seen a specified number of times
in both single- and multi-substitution contexts:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filter_by_subs_observed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">  barcode aa_substitutions  n_aa_substitutions</span>
<span class="go">0    var1                                    0</span>
<span class="go">1    var2              M1A                   1</span>
<span class="go">2    var3          M1A G2A                   2</span>
<span class="go">3    var5              G2A                   1</span>
<span class="go">4    var6              M1A                   1</span>
</pre></div>
</div>
<p>We can also make the requirement <strong>or</strong> rather than <strong>and</strong>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filter_by_subs_observed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">and_vs_or</span><span class="o">=</span><span class="s1">&#39;or&#39;</span><span class="p">)</span>
<span class="go">  barcode aa_substitutions  n_aa_substitutions</span>
<span class="go">0    var1                                    0</span>
<span class="go">1    var2              M1A                   1</span>
<span class="go">2    var3          M1A G2A                   2</span>
<span class="go">3    var5              G2A                   1</span>
<span class="go">4    var6              M1A                   1</span>
<span class="go">5    var7              M1C                   1</span>
</pre></div>
</div>
<p>Do not filter any variants:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filter_by_subs_observed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">  barcode aa_substitutions  n_aa_substitutions</span>
<span class="go">0    var1                                    0</span>
<span class="go">1    var2              M1A                   1</span>
<span class="go">2    var3          M1A G2A                   2</span>
<span class="go">3    var4          M1A G2C                   2</span>
<span class="go">4    var5              G2A                   1</span>
<span class="go">5    var6              M1A                   1</span>
<span class="go">6    var7              M1C                   1</span>
</pre></div>
</div>
</dd></dl>



          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo" />
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dms_variants.html">dms_variants package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dms_variants.html#dms-variants">dms_variants</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="dms_variants.html#submodules">Submodules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="dms_variants.html">dms_variants package</a><ul>
      <li>Previous: <a href="dms_variants.bottlenecks.html" title="previous chapter">bottlenecks</a></li>
      <li>Next: <a href="dms_variants.constants.html" title="next chapter">constants</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2019--2024.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/dms_variants.codonvarianttable.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://github.blog/wp-content/uploads/2008/12/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>