
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>globalepistasis &#8212; dms_variants 0.8.2 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="illuminabarcodeparser" href="dms_variants.illuminabarcodeparser.html" />
    <link rel="prev" title="fastq" href="dms_variants.fastq.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_variants.globalepistasis"></span><div class="section" id="globalepistasis">
<h1>globalepistasis<a class="headerlink" href="#globalepistasis" title="Permalink to this headline">¶</a></h1>
<p>Implements global epistasis models that are based on (but extend in some
ways) those described in <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>.
See also <a class="reference external" href="https://www.genetics.org/content/205/3/1079">Sailer and Harms (2017)</a> and <a class="reference external" href="https://doi.org/10.1093/molbev/msy141">Otwinoski (2018)</a>.</p>
<div class="contents local topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#global-epistasis-function" id="id12">Global epistasis function</a></p>
<ul>
<li><p><a class="reference internal" href="#no-epistasis-function" id="id13">No epistasis function</a></p></li>
<li><p><a class="reference internal" href="#monotonic-spline-epistasis-function" id="id14">Monotonic spline epistasis function</a></p></li>
<li><p><a class="reference internal" href="#multiple-latent-phenotypes" id="id15">Multiple latent phenotypes</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#likelihood-calculation" id="id16">Likelihood calculation</a></p>
<ul>
<li><p><a class="reference internal" href="#gaussian-likelihood" id="id17">Gaussian likelihood</a></p></li>
<li><p><a class="reference internal" href="#cauchy-likelihood" id="id18">Cauchy likelihood</a></p></li>
<li><p><a class="reference internal" href="#bottleneck-likelihood" id="id19">Bottleneck likelihood</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#the-model-classes" id="id20">The model classes</a></p></li>
<li><p><a class="reference internal" href="#details-of-fitting" id="id21">Details of fitting</a></p>
<ul>
<li><p><a class="reference internal" href="#fitting-workflow" id="id22">Fitting workflow</a></p></li>
<li><p><a class="reference internal" href="#fitting-multiple-latent-phenotypes" id="id23">Fitting multiple latent phenotypes</a></p></li>
<li><p><a class="reference internal" href="#conveniently-fitting-and-comparing-several-models" id="id24">Conveniently fitting and comparing several models</a></p></li>
<li><p><a class="reference internal" href="#vector-representation-of-beta-rm-wt" id="id25">Vector representation of <span class="math notranslate nohighlight">\(\beta_{\rm{wt}}\)</span></a></p></li>
<li><p><a class="reference internal" href="#optimization" id="id26">Optimization</a></p></li>
<li><p><a class="reference internal" href="#gradients-used-in-optimization" id="id27">Gradients used in optimization</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#detailed-documentation-of-models" id="id28">Detailed documentation of models</a></p></li>
</ul>
</div>
<div class="section" id="global-epistasis-function">
<span id="id1"></span><h2><a class="toc-backref" href="#id12">Global epistasis function</a><a class="headerlink" href="#global-epistasis-function" title="Permalink to this headline">¶</a></h2>
<p>The global epistasis function is defined as follows.
Let <span class="math notranslate nohighlight">\(v\)</span> be a variant. We convert
<span class="math notranslate nohighlight">\(v\)</span> into a binary representation with respect to some wildtype
sequence. This representation is a vector <span class="math notranslate nohighlight">\(\mathbf{b}\left(v\right)\)</span>
with element <span class="math notranslate nohighlight">\(b\left(v\right)_m\)</span> equal to 1 if the variant has mutation
<span class="math notranslate nohighlight">\(m\)</span> and 0 otherwise, and <span class="math notranslate nohighlight">\(m\)</span> ranging over all <span class="math notranslate nohighlight">\(M\)</span> mutations
observed in the overall set of variants (so <span class="math notranslate nohighlight">\(\mathbf{b}\left(v\right)\)</span>
is of length <span class="math notranslate nohighlight">\(M\)</span>). Variants can be converted into this binary form
using <a class="reference internal" href="dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap" title="dms_variants.binarymap.BinaryMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.binarymap.BinaryMap</span></code></a>.</p>
<p>We define a <em>latent effect</em> for each mutation <span class="math notranslate nohighlight">\(m\)</span>, which we denote as
<span class="math notranslate nohighlight">\(\beta_m\)</span>. The latent effects of mutations contribute additively to the
<em>latent phenotype</em>, and the latent phenotype of the wildtype sequence is
<span class="math notranslate nohighlight">\(\beta_{\rm{wt}}\)</span>. So the <em>latent phenotype</em> of variant <span class="math notranslate nohighlight">\(v\)</span> is:</p>
<div class="math notranslate nohighlight" id="equation-latent-phenotype">
<span class="eqno">(1)<a class="headerlink" href="#equation-latent-phenotype" title="Permalink to this equation">¶</a></span>\[\phi\left(v\right) = \beta_{\rm{wt}} +
                     \sum_{m=1}^M \beta_m b\left(v\right)_m.\]</div>
<p>The predicted <em>observed phenotype</em> <span class="math notranslate nohighlight">\(p\left(v\right)\)</span> is a function of the
latent phenotype:</p>
<div class="math notranslate nohighlight" id="equation-observed-phenotype">
<span class="eqno">(2)<a class="headerlink" href="#equation-observed-phenotype" title="Permalink to this equation">¶</a></span>\[p\left(v\right) = g\left(\phi\left(v\right)\right)\]</div>
<p>where <span class="math notranslate nohighlight">\(g\)</span> is the <em>global epistasis function</em>.</p>
<p>We define the following global epistasis functions:</p>
<div class="section" id="no-epistasis-function">
<span id="id2"></span><h3><a class="toc-backref" href="#id13">No epistasis function</a><a class="headerlink" href="#no-epistasis-function" title="Permalink to this headline">¶</a></h3>
<p>No epistasis, so the observed phenotype is just the latent phenotype:</p>
<div class="math notranslate nohighlight" id="equation-noepistasis">
<span class="eqno">(3)<a class="headerlink" href="#equation-noepistasis" title="Permalink to this equation">¶</a></span>\[g\left(x\right) = x.\]</div>
<p>This function is implemented as <a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasis</span></code></a>.</p>
</div>
<div class="section" id="monotonic-spline-epistasis-function">
<span id="id3"></span><h3><a class="toc-backref" href="#id14">Monotonic spline epistasis function</a><a class="headerlink" href="#monotonic-spline-epistasis-function" title="Permalink to this headline">¶</a></h3>
<p>This is the function used in <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>. It transforms
the latent phenotype to the observed phenotype using monotonic I-splines with
linear extrapolation outside the spline boundaries:</p>
<div class="math notranslate nohighlight" id="equation-monotonicspline">
<span class="eqno">(4)<a class="headerlink" href="#equation-monotonicspline" title="Permalink to this equation">¶</a></span>\[\begin{split}g\left(x\right)
=
\begin{cases}
c_{\alpha} + \sum_{m=1}^M \alpha_{m} I_m\left(x\right)
  &amp; \rm{if\;} L \le x \le U, \\
c_{\alpha} + \sum_{m=1}^M \alpha_m
  \left[I_m\left(L\right) + \left(x - L\right)
        \left.\frac{\partial I_m\left(y\right)}
                   {\partial y}\right\rvert_{y=L}
  \right]
  &amp; \rm{if\;} x &lt; L, \\
c_{\alpha} + \sum_{m=1}^M \alpha_m
  \left[I_m\left(U\right) + \left(x - U\right)
        \left.\frac{\partial I_m\left(y\right)}
                   {\partial y}\right\rvert_{y=U}
  \right]
  &amp; \rm{if\;} x &gt; U,
\end{cases}\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(c_{\alpha}\)</span> is an arbitrary number giving the <em>minimum</em>
observed phenotype, the <span class="math notranslate nohighlight">\(\alpha_m\)</span> coefficients are all <span class="math notranslate nohighlight">\(\ge 0\)</span>,
<span class="math notranslate nohighlight">\(I_m\)</span> indicates a family of I-splines defined via
<a class="reference internal" href="dms_variants.ispline.html#dms_variants.ispline.Isplines_total" title="dms_variants.ispline.Isplines_total"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.ispline.Isplines_total</span></code></a>, and <span class="math notranslate nohighlight">\(L\)</span> and <span class="math notranslate nohighlight">\(U\)</span> are
the lower and upper bounds on the regions over which the I-splines are defined.
Note how when <span class="math notranslate nohighlight">\(x\)</span> is outside the range of the I-splines, we linearly
extrapolate <span class="math notranslate nohighlight">\(g\)</span> from its range boundaries to calculate.</p>
<p>This function is implemented as <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a>. By default,
the I-splines are of order 3 and are defined on a mesh of four evenly spaced
points such that the total number of I-splines is <span class="math notranslate nohighlight">\(M=5\)</span> (although these
options can be adjusted when initializing a <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a>
model).</p>
<p>The latent effects are scaled so that their mean absolute value is one,
and the latent phenotype of the wildtype is set to zero.</p>
</div>
<div class="section" id="multiple-latent-phenotypes">
<span id="multi-latent"></span><h3><a class="toc-backref" href="#id15">Multiple latent phenotypes</a><a class="headerlink" href="#multiple-latent-phenotypes" title="Permalink to this headline">¶</a></h3>
<p>Equations <a class="reference internal" href="#equation-latent-phenotype">(1)</a> and <a class="reference internal" href="#equation-observed-phenotype">(2)</a> can be
generalized to the case where multiple latent phenotypes contribute
to the observed phenotype. Specifically, let there be
<span class="math notranslate nohighlight">\(k = 1, \ldots, K\)</span> different latent phenotypes, and let
<span class="math notranslate nohighlight">\(\beta_m^k\)</span> denote the effect of mutation <span class="math notranslate nohighlight">\(m\)</span> on latent phenotype
<span class="math notranslate nohighlight">\(k\)</span>. Then we generalize Equation <a class="reference internal" href="#equation-latent-phenotype">(1)</a> to</p>
<div class="math notranslate nohighlight" id="equation-latent-phenotype-multi">
<span class="eqno">(5)<a class="headerlink" href="#equation-latent-phenotype-multi" title="Permalink to this equation">¶</a></span>\[\phi_k\left(v\right) = \beta_{\rm{wt}}^k +
                       \sum_{m=1}^M \beta_m^k b\left(v\right)_m,\]</div>
<p>and Equation <a class="reference internal" href="#equation-observed-phenotype">(2)</a> to</p>
<div class="math notranslate nohighlight" id="equation-observed-phenotype-multi">
<span class="eqno">(6)<a class="headerlink" href="#equation-observed-phenotype-multi" title="Permalink to this equation">¶</a></span>\[p\left(v\right) = \sum_{k=1}^K g_k\left(\phi_k\left(v\right)\right),\]</div>
<p>where <span class="math notranslate nohighlight">\(\phi_k\left(v\right)\)</span> is the <span class="math notranslate nohighlight">\(k\)</span>-th latent phenotype of
variant <span class="math notranslate nohighlight">\(v\)</span>, and <span class="math notranslate nohighlight">\(g_k\)</span> is the <span class="math notranslate nohighlight">\(k\)</span>-th global epistasis
function.</p>
<p>Note that it does <strong>not</strong> make sense to fit multiple latent phenotypes
to a non-epistatic model as a linear combination of linear effects
reduces to a simple linear model (in other words, a multi-latent
phenotype non-epistatic model is no differen than a one-latent
phenotype non-epistatic model).</p>
</div>
</div>
<div class="section" id="likelihood-calculation">
<span id="id4"></span><h2><a class="toc-backref" href="#id16">Likelihood calculation</a><a class="headerlink" href="#likelihood-calculation" title="Permalink to this headline">¶</a></h2>
<p>We defined a <em>likelihood</em> capturing how well the model describes the
actual data, and then fit the models by finding the parameters that
maximize this likelihood. This means that different epistasis functions
(as described in <a class="reference internal" href="#id1">Global epistasis function</a>) can be compared via
their likelihoods after correcting for the number of parameters
(e.g. by <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>).</p>
<p>We consider several different forms for calculating the likelihood.
Note that epistasis functions can only be compared within the same form of
the likelihood on the same dataset–you <strong>cannot</strong> compare likelihoods
calculated using different methods, or on different datasets.</p>
<div class="section" id="gaussian-likelihood">
<span id="id5"></span><h3><a class="toc-backref" href="#id17">Gaussian likelihood</a><a class="headerlink" href="#gaussian-likelihood" title="Permalink to this headline">¶</a></h3>
<p>This is the form of the likelihood used in <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>. It is
most appropriate when we have functional scores for variants along with
good estimates of normally distributed errors on these functional scores.</p>
<p>For each variant <span class="math notranslate nohighlight">\(v\)</span>, we have an experimentally measured functional
score <span class="math notranslate nohighlight">\(y_v\)</span> and optionally an estimate of the error (variance)
<span class="math notranslate nohighlight">\(\sigma^2_{y_v}\)</span> in this functional score measurement. If no error
estimates are available, then we set <span class="math notranslate nohighlight">\(\sigma^2_{y_v} = 0\)</span>.</p>
<p>The goal of the fitting is to parameterize the model so the observed phenotype
<span class="math notranslate nohighlight">\(p\left(v\right)\)</span> predicted by the model is as close as possible to the
measured functional score <span class="math notranslate nohighlight">\(y_v\)</span>. Following <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>,
we assume the likelihood of measuring a functional score <span class="math notranslate nohighlight">\(y_v\)</span> is
normally distributed around the model prediction <span class="math notranslate nohighlight">\(p\left(v\right)\)</span>
with variance <span class="math notranslate nohighlight">\(\sigma^2_{y_v} + \sigma^2_{\rm{HOC}}\)</span>, where
<span class="math notranslate nohighlight">\(\sigma^2_{\rm{HOC}}\)</span> is the un-modeled <em>house-of-cards epistasis</em>
(although in practice it could also represent experimental noise not
capture in the variance estimates). So the overall log likelihood of
the model is</p>
<div class="math notranslate nohighlight" id="equation-loglik-gaussian">
<span class="eqno">(7)<a class="headerlink" href="#equation-loglik-gaussian" title="Permalink to this equation">¶</a></span>\[\mathcal{L} = \sum_{v=1}^V \ln\left[N\left(y_v \mid p\left(v\right),
              \sigma^2_{y_v} + \sigma^2_{\rm{HOC}}\right)\right]\]</div>
<p>where <span class="math notranslate nohighlight">\(V\)</span> is the number of variants and <span class="math notranslate nohighlight">\(N\)</span> is the normal
distribution defined by</p>
<div class="math notranslate nohighlight" id="equation-normaldist">
<span class="eqno">(8)<a class="headerlink" href="#equation-normaldist" title="Permalink to this equation">¶</a></span>\[N\left(y \mid \mu, \sigma^2\right) = \frac{1}{\sqrt{2 \pi \sigma^2}} \exp
                 \left(-\frac{\left(y - \mu\right)^2}{2 \sigma^2}\right).\]</div>
<p>This likelihood calculation is implemented as <a class="reference internal" href="#dms_variants.globalepistasis.GaussianLikelihood" title="dms_variants.globalepistasis.GaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">GaussianLikelihood</span></code></a>.</p>
</div>
<div class="section" id="cauchy-likelihood">
<span id="id6"></span><h3><a class="toc-backref" href="#id18">Cauchy likelihood</a><a class="headerlink" href="#cauchy-likelihood" title="Permalink to this headline">¶</a></h3>
<p>This form of the likelihood assumes that the difference between the
measured functional scores and the model’s observed phenotypes follows a
<a class="reference external" href="https://en.wikipedia.org/wiki/Cauchy_distribution">Cauchy distribution</a>.</p>
<p>A potential advantage over the <a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a> is the
<a class="reference external" href="https://en.wikipedia.org/wiki/Fat-tailed_distribution">fatter tails</a>
of the Cauchy distribution relative to the Gaussian distribution.
This could be advantageous if some measurements are very large outliers
(either due to un-modeled epistasis or experimental factors such as
mis-calling of variant sequences) in ways that are not
capture in the functional score variance estimates <span class="math notranslate nohighlight">\(\sigma^2_{y_v}\)</span>.
Such outlier measurements will have less influence on the overall model
fit for the Cauchy likelihood relative to the <a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a>.</p>
<p>Specifically, we compute the overall log likelihood as</p>
<div class="math notranslate nohighlight" id="equation-loglik-cauchy">
<span class="eqno">(9)<a class="headerlink" href="#equation-loglik-cauchy" title="Permalink to this equation">¶</a></span>\[\mathcal{L} = -\sum_{v=1}^V
              \ln\left[\pi \sqrt{\gamma^2 + \sigma^2_{y_v}}
                       \left(1 + \frac{\left[y_v - p\left(v\right)\right]^2}
                                      {\gamma^2 + \sigma^2_{y_v}}
                       \right)
                  \right]\]</div>
<p>where <span class="math notranslate nohighlight">\(\gamma\)</span> is the scale parameters, and the functional score
variance estimates <span class="math notranslate nohighlight">\(\sigma^2_{y_v}\)</span> are incorporated in heuristic way
(with no real theoretical basis)
that qualitatively captures the fact that larger variance estimates give a
broader distribution. If variance estimates are not available then
<span class="math notranslate nohighlight">\(\sigma^2_{y_v}\)</span> is set to zero.</p>
<p>This likelihood calculation is implemented as <a class="reference internal" href="#dms_variants.globalepistasis.CauchyLikelihood" title="dms_variants.globalepistasis.CauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">CauchyLikelihood</span></code></a>.</p>
</div>
<div class="section" id="bottleneck-likelihood">
<span id="id7"></span><h3><a class="toc-backref" href="#id19">Bottleneck likelihood</a><a class="headerlink" href="#bottleneck-likelihood" title="Permalink to this headline">¶</a></h3>
<p>This form of the likelihood is appropriate when most noise in the experiment
comes from a bottleneck when passaging the library from the pre-selection to
post-selection condition. This will be the case when the total pre- and post-
selection sequencing depths greatly exceed the number of variants
that were physically passaged from the pre-selection library to the post-
selection condition. At least in Bloom lab viral deep mutational scanning
experiments, this situation is quite common.</p>
<p>A full derivation of the log likelihood in this situation is given in
<a class="reference internal" href="bottleneck_likelihood.html"><span class="doc">Bottleneck likelihood for global epistasis models</span></a>. As explained in those calculations, the
likelihood is computed from the experimental observables
<span class="math notranslate nohighlight">\(f_v^{\text{pre}}\)</span> and <span class="math notranslate nohighlight">\(f_v^{\text{post}}\)</span>, which represent
the pre- and post-selection frequencies of variant <span class="math notranslate nohighlight">\(v\)</span>. They are
computed from the pre- and post-selection counts <span class="math notranslate nohighlight">\(n_v^{\text{pre}}\)</span>
and <span class="math notranslate nohighlight">\(n_v^{\text{post}}\)</span> of the variants as</p>
<div class="math notranslate nohighlight" id="equation-f-v-pre-post">
<span class="eqno">(10)<a class="headerlink" href="#equation-f-v-pre-post" title="Permalink to this equation">¶</a></span>\[\begin{split}f_v^{\text{pre}}
&amp;=&amp; \frac{n_v^{\text{pre}} + C}
         {\sum_{v'=1}^V \left(n_{v'}^{\text{pre}} + C \right)} \\
f_v^{\text{post}}
&amp;=&amp; \frac{n_v^{\text{post}} + C}
         {\sum_{v'=1}^V \left(n_{v'}^{\text{post}} + C \right)}\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(C\)</span> is a pseudocount which by default is 0.5.</p>
<p>Using the bottleneck likelihood also requires an estimation of the experimental
bottleneck <span class="math notranslate nohighlight">\(N_{\rm{bottle}}\)</span> when passaging the library from the pre-
to post-selection conditions. A smaller bottleneck will correspond to more
“noise” in the experiment, since random bottlenecking changes the frequencies
of variants unpredictably. You can either estimate <span class="math notranslate nohighlight">\(N_{\rm{bottle}}\)</span>
experimentally or from fluctuations in the relative frequencies of
wildtype or synonymous variants, such as via
<a class="reference internal" href="dms_variants.bottlenecks.html#dms_variants.bottlenecks.estimateBottleneck" title="dms_variants.bottlenecks.estimateBottleneck"><code class="xref py py-func docutils literal notranslate"><span class="pre">dms_variants.bottlenecks.estimateBottleneck()</span></code></a>.</p>
<p>Given these experimentally measured parameters, the overall log likelihood is:</p>
<div class="math notranslate nohighlight" id="equation-loglik-bottleneck">
<span class="eqno">(11)<a class="headerlink" href="#equation-loglik-bottleneck" title="Permalink to this equation">¶</a></span>\[\mathcal{L}
=
\sum_{v=1}^V
\left[ n_v^{\rm{bottle}} \ln\left(N_{\rm{bottle}} f_v^{\text{pre}}\right)
       - \ln \Gamma \left(n_v^{\rm{bottle}} + 1\right)
     \right] - N_{\rm{bottle}}\]</div>
<p>where <span class="math notranslate nohighlight">\(\Gamma\)</span> is the
<a class="reference external" href="https://en.wikipedia.org/wiki/Gamma_function">gamma function</a>
and <span class="math notranslate nohighlight">\(n_v^{\rm{bottle}}\)</span> is the estimated number of copies of variant
<span class="math notranslate nohighlight">\(v\)</span> that made it through the bottleneck, which is defined in terms
of the phenotype <span class="math notranslate nohighlight">\(p\left(v\right)\)</span> as</p>
<div class="math notranslate nohighlight" id="equation-n-v-bottle">
<span class="eqno">(12)<a class="headerlink" href="#equation-n-v-bottle" title="Permalink to this equation">¶</a></span>\[n_v^{\rm{bottle}}
=
\frac{f_v^{\rm{post}} N_{\rm{bottle}}
      \sum_{v'=1}^V f_{v'}^{\text{pre}} 2^{p\left(v'\right)}}
     {2^{p\left(v\right)}}.\]</div>
<p>The free parameters are therefore the <span class="math notranslate nohighlight">\(p\left(v\right)\)</span> values
(the <span class="math notranslate nohighlight">\(n_v^{\rm{bottle}}\)</span> values are hidden variables that are
not explicitly estimated). Note that Eq. <a class="reference internal" href="#equation-n-v-bottle">(12)</a>
uses an exponent base of 2, but it can be set to arbitrary positive values
in the actual implementation.</p>
<p>After fitting the observed phenotypes, the parameters are re-scaled
so that the observed phenotype of wildtype is zero
(i.e., <span class="math notranslate nohighlight">\(p\left(\rm{wt}\right) = 0\)</span>).</p>
<p>This likelihood calculation is implemented as <a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood" title="dms_variants.globalepistasis.BottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">BottleneckLikelihood</span></code></a>.</p>
</div>
</div>
<div class="section" id="the-model-classes">
<h2><a class="toc-backref" href="#id20">The model classes</a><a class="headerlink" href="#the-model-classes" title="Permalink to this headline">¶</a></h2>
<p>The epistasis models are defined in a set of classes. All these classes
inherit their main functionality from the <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a>
abstract base class.</p>
<p>There are subclasses of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> that implement the
global epistasis functions and likelihood calculation methods. Specifically,
the following classes implement a <a class="reference internal" href="#global-epistasis-function"><span class="std std-ref">Global epistasis function</span></a>:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasis</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a></p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>and the following classes each implement a <a class="reference internal" href="#likelihood-calculation"><span class="std std-ref">Likelihood calculation</span></a>:</dt><dd><ul class="simple">
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.GaussianLikelihood" title="dms_variants.globalepistasis.GaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">GaussianLikelihood</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.CauchyLikelihood" title="dms_variants.globalepistasis.CauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">CauchyLikelihood</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood" title="dms_variants.globalepistasis.BottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">BottleneckLikelihood</span></code></a></p></li>
</ul>
</dd>
</dl>
<p>However, those classes can still not be directly instantianted, as a fully
concrete model subclass must have <strong>both</strong> a global epistasis function and
a likelihood calculation method.
The following classes implement both, and so can be directly instantiated
for use in analyses:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasisGaussianLikelihood" title="dms_variants.globalepistasis.NoEpistasisGaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasisGaussianLikelihood</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasisCauchyLikelihood" title="dms_variants.globalepistasis.NoEpistasisCauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasisCauchyLikelihood</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasisBottleneckLikelihood" title="dms_variants.globalepistasis.NoEpistasisBottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasisBottleneckLikelihood</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasisGaussianLikelihood" title="dms_variants.globalepistasis.MonotonicSplineEpistasisGaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasisGaussianLikelihood</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasisCauchyLikelihood" title="dms_variants.globalepistasis.MonotonicSplineEpistasisCauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasisCauchyLikelihood</span></code></a></p></li>
<li><p><a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasisBottleneckLikelihood" title="dms_variants.globalepistasis.MonotonicSplineEpistasisBottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasisBottleneckLikelihood</span></code></a></p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="details-of-fitting">
<h2><a class="toc-backref" href="#id21">Details of fitting</a><a class="headerlink" href="#details-of-fitting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="fitting-workflow">
<span id="id8"></span><h3><a class="toc-backref" href="#id22">Fitting workflow</a><a class="headerlink" href="#fitting-workflow" title="Permalink to this headline">¶</a></h3>
<p>The fitting workflow for a single latent phenotype is similar to that described
in <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>The latent effects are fit under an additive (non-epistatic) model
using least squares. The residuals from this fit are then used to
estimate <span class="math notranslate nohighlight">\(\sigma^2_{\rm{HOC}}\)</span> for <a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a>,
or <span class="math notranslate nohighlight">\(\gamma^2\)</span> for <a class="reference internal" href="#cauchy-likelihood"><span class="std std-ref">Cauchy likelihood</span></a>.</p></li>
<li><p>If there are any parameters in the epistasis function, they are set
to reasonable initial values. For <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a>
this involves setting the mesh to go from 0 to 1, scaling the latent
effects so that the latent phenotypes range from 0 to 1, setting
<span class="math notranslate nohighlight">\(c_{\alpha}\)</span> to the minimum functional score and setting the
weights <span class="math notranslate nohighlight">\(\alpha_m\)</span> to equal values such that the max of the
epistasis function is the same as the max functional score.</p></li>
<li><p>The overall model is fit by maximum likelihood.</p></li>
<li><p>For <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a>, the latent effects and wildtype
latent phenotype are rescaled so that the mean absolute value latent
effect is one and the wildtype latent phenotype is zero.</p></li>
</ol>
</div></blockquote>
</div>
<div class="section" id="fitting-multiple-latent-phenotypes">
<span id="fitting-multi-latent"></span><h3><a class="toc-backref" href="#id23">Fitting multiple latent phenotypes</a><a class="headerlink" href="#fitting-multiple-latent-phenotypes" title="Permalink to this headline">¶</a></h3>
<p>When there are multiple latent phenotypes (see <a class="reference internal" href="#multi-latent"><span class="std std-ref">Multiple latent phenotypes</span></a>), the
fitting workflow changes. To fit a model with <span class="math notranslate nohighlight">\(K &gt; 1\)</span> latent phenotypes,
first fit a model of the same type to the same data with <span class="math notranslate nohighlight">\(K - 1\)</span> latent
phenotype. The values from that fit for the first <span class="math notranslate nohighlight">\(K - 1\)</span> latent
phenotypes are used to initialize all parameters relevant to those first
<span class="math notranslate nohighlight">\(K - 1\)</span> latent phenotypes and the associated global epistasis
functions and likelihood calculations.</p>
<p>Then parameters relevant to latent phenotype <span class="math notranslate nohighlight">\(K\)</span> are set so the initial
contribution of this phenotype to the overall observed phenotype is zero.
Specifically the latent effects <span class="math notranslate nohighlight">\(\beta_m^K\)</span> for phenotype <span class="math notranslate nohighlight">\(K\)</span> are
all set to zero and <span class="math notranslate nohighlight">\(\beta_{\rm{wt}}^K\)</span> is chosen so that
<span class="math notranslate nohighlight">\(0 = g_K \left(\beta_{\rm{wt}}^K\right)\)</span>. For a
<a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a> model, initial parameters for <span class="math notranslate nohighlight">\(g_K\)</span> are
chosen so that over the mesh, <span class="math notranslate nohighlight">\(g_K\)</span> spans +/- the absolute value of
the largest residual of the model with <span class="math notranslate nohighlight">\(K - 1\)</span> latent phenotypes.</p>
<p>After initializing the paramters in this way, the entire model (all latent
phenotypes) is fit by maximum likelihood.</p>
</div>
<div class="section" id="conveniently-fitting-and-comparing-several-models">
<h3><a class="toc-backref" href="#id24">Conveniently fitting and comparing several models</a><a class="headerlink" href="#conveniently-fitting-and-comparing-several-models" title="Permalink to this headline">¶</a></h3>
<p>To conveniently fit and compare several models, use <a class="reference internal" href="#dms_variants.globalepistasis.fit_models" title="dms_variants.globalepistasis.fit_models"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_models()</span></code></a>.
This is especially useful when you are including models with multiple
latent phenotypes.</p>
</div>
<div class="section" id="vector-representation-of-beta-rm-wt">
<h3><a class="toc-backref" href="#id25">Vector representation of <span class="math notranslate nohighlight">\(\beta_{\rm{wt}}\)</span></a><a class="headerlink" href="#vector-representation-of-beta-rm-wt" title="Permalink to this headline">¶</a></h3>
<p>For the purposes of the optimization (and in the equations below), we change
how <span class="math notranslate nohighlight">\(\beta_{\rm{wt}}\)</span> is represented to simplify the calculations.
Specifically, to the binary encoding <span class="math notranslate nohighlight">\(\mathbf{b}\left(v\right)\)</span> of
each variant, we append a 1 so that the encodings are now of length
<span class="math notranslate nohighlight">\(M + 1\)</span>. We then define <span class="math notranslate nohighlight">\(\beta_{M + 1} = \beta_{\rm{wt}}\)</span>.
Then Eq. <a class="reference internal" href="#equation-latent-phenotype">(1)</a> can be rewritten as</p>
<div class="math notranslate nohighlight" id="equation-latent-phenotype-wt-vec">
<span class="eqno">(13)<a class="headerlink" href="#equation-latent-phenotype-wt-vec" title="Permalink to this equation">¶</a></span>\[\phi\left(v\right) = \sum_{m=1}^{M+1} \beta_m b\left(v\right)_m\]</div>
<p>enabling <span class="math notranslate nohighlight">\(\beta_{\rm{wt}}\)</span> to just be handled like the other
<span class="math notranslate nohighlight">\(\beta_m\)</span> parameters.</p>
</div>
<div class="section" id="optimization">
<h3><a class="toc-backref" href="#id26">Optimization</a><a class="headerlink" href="#optimization" title="Permalink to this headline">¶</a></h3>
<p>The optimization is performed by <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.fit" title="dms_variants.globalepistasis.AbstractEpistasis.fit"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AbstractEpistasis.fit()</span></code></a>.
There are several options to that method about how to do the optimization;
by default it uses a L-BFGS-B algorithm with exact gradients
calculated as below.</p>
</div>
<div class="section" id="gradients-used-in-optimization">
<h3><a class="toc-backref" href="#id27">Gradients used in optimization</a><a class="headerlink" href="#gradients-used-in-optimization" title="Permalink to this headline">¶</a></h3>
<p>For the optimization, we use the following gradients:</p>
<p>Gradient of latent phenotype with respect to latent effects:</p>
<div class="math notranslate nohighlight" id="equation-dlatent-phenotype-dlatent-effect">
<span class="eqno">(14)<a class="headerlink" href="#equation-dlatent-phenotype-dlatent-effect" title="Permalink to this equation">¶</a></span>\[\begin{split}\frac{\partial \phi_j\left(v\right)}{\partial \beta_m^k} =
\begin{cases}
b\left(v_m\right) &amp; \rm{if\;} j = k, \\
0 &amp; \rm{otherwise.} \\
\end{cases}\end{split}\]</div>
<p>Gradient of observed phenotype with respect to latent phenotypes:</p>
<div class="math notranslate nohighlight" id="equation-dobserved-phenotype-dlatent-effect">
<span class="eqno">(15)<a class="headerlink" href="#equation-dobserved-phenotype-dlatent-effect" title="Permalink to this equation">¶</a></span>\[\begin{split}\frac{\partial p\left(v\right)}{\partial \beta_m^k}
&amp;=&amp; \left.\frac{\partial g_k\left(x\right)}{\partial x}
    \right\rvert_{x = \phi_k\left(v\right)} \times
    \frac{\partial \phi_k\left(v\right)}{\partial \beta_m^k} \\
&amp;=&amp; \left.\frac{\partial g_k\left(x\right)}{\partial x}
    \right\rvert_{x = \phi_k\left(v\right)} \times b\left(v_m\right)\end{split}\]</div>
<p>Derivative of the likelihood with respect to latent effects:</p>
<div class="math notranslate nohighlight" id="equation-dloglik-dlatent-effect">
<span class="eqno">(16)<a class="headerlink" href="#equation-dloglik-dlatent-effect" title="Permalink to this equation">¶</a></span>\[\frac{\partial \mathcal{L}}{\partial \beta_m^k}
= \sum_{v=1}^V \frac{\mathcal{L}}
                    {\partial p\left(v\right)} \times
               \frac{\partial p\left(v\right)}{\partial \beta_m^k}.\]</div>
<p>Derivative of <a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a> (Eq. <a class="reference internal" href="#equation-loglik-gaussian">(7)</a>) with
respect to observed phenotype:</p>
<div class="math notranslate nohighlight" id="equation-dloglik-gaussian-dobserved-phenotype">
<span class="eqno">(17)<a class="headerlink" href="#equation-dloglik-gaussian-dobserved-phenotype" title="Permalink to this equation">¶</a></span>\[\frac{\partial \mathcal{L}}{\partial p\left(v\right)}
= \frac{y_v - p\left(v\right)}
       {\sigma_{y_v}^2 + \sigma^2_{\rm{HOC}}}.\]</div>
<p>Derivative of <a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a> (Eq. <a class="reference internal" href="#equation-loglik-gaussian">(7)</a>) with
respect to house-of-cards epistasis:</p>
<div class="math notranslate nohighlight" id="equation-dloglik-gaussian-depistasis-hoc">
<span class="eqno">(18)<a class="headerlink" href="#equation-dloglik-gaussian-depistasis-hoc" title="Permalink to this equation">¶</a></span>\[\frac{\partial \mathcal{L}}{\partial \sigma^2_{\rm{HOC}}} = \sum_{v=1}^V
\frac{1}{2} \left[\left(\frac{\partial \mathcal{L}}
                             {\partial p\left(v\right)}\right)^2
                  - \frac{1}{\sigma_{y_v}^2 + \sigma_{\rm{HOC}}^2} \right].\]</div>
<p>Derivative of <a class="reference internal" href="#cauchy-likelihood"><span class="std std-ref">Cauchy likelihood</span></a> (Eq. <a class="reference internal" href="#equation-loglik-cauchy">(9)</a>) with
respect to observed phenotype:</p>
<div class="math notranslate nohighlight" id="equation-dloglik-cauchy-dobserved-phenotype">
<span class="eqno">(19)<a class="headerlink" href="#equation-dloglik-cauchy-dobserved-phenotype" title="Permalink to this equation">¶</a></span>\[\frac{\partial \mathcal{L}}{\partial p\left(v\right)}
=
\frac{2\left[y_v - p\left(v\right)\right]}
     {\gamma^2 + \sigma^2_{y_v} +
      \left[y_v - p\left(v\right)\right]^2}.\]</div>
<p>Derivative of <a class="reference internal" href="#cauchy-likelihood"><span class="std std-ref">Cauchy likelihood</span></a> (Eq. <a class="reference internal" href="#equation-loglik-cauchy">(9)</a>) with
respect to scale parameter:</p>
<div class="math notranslate nohighlight" id="equation-dloglik-cauchy-dscale-parameter">
<span class="eqno">(20)<a class="headerlink" href="#equation-dloglik-cauchy-dscale-parameter" title="Permalink to this equation">¶</a></span>\[\frac{\partial \mathcal{L}}{\partial \gamma}
=
\sum_{v=1}^{V} \frac{\gamma\left(\left[y_v - p\left(v\right)\right]^2 -
                                 \gamma^2 - \sigma^2_{y_v}
                           \right)}
                    {\left(\gamma^2 + \sigma^2_{y_v}\right)
                     \left(\gamma^2 + \sigma^2_{y_v} +
                           \left[y_v - p\left(v\right)\right]^2\right)
                     }\]</div>
<p>Derivative of <a class="reference internal" href="#bottleneck-likelihood"><span class="std std-ref">Bottleneck likelihood</span></a> with respect to
<span class="math notranslate nohighlight">\(p\left(v\right)\)</span>:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\frac{\partial n_v^{\text{bottle}}}
     {\partial p\left(v'\right)}
&amp;=&amp;
\begin{cases}
\frac{\left(\ln 2\right) f_v^{\text{post}} N_{\text{bottle}}
      f_{v}^{\text{pre}} 2^{p\left(v\right)}}
     {2^{p\left(v\right)}}
- \left(\ln 2\right) n_v^{\text{bottle}}
&amp; \rm{if\;} v = v', \\
\frac{\left(\ln 2\right) f_v^{\text{post}} N_{\text{bottle}}
      f_{v'}^{\text{pre}} 2^{p\left(v'\right)}}
     {2^{p\left(v\right)}}
&amp; \text{otherwise}
\end{cases} \\
&amp;=&amp;
\frac{\left(\ln 2\right) f_v^{\text{post}} N_{\text{bottle}}
      f_{v'}^{\text{pre}} 2^{p\left(v'\right)}}
     {2^{p\left(v\right)}}
- \delta_{v,v'} \left(\ln 2\right) n_{v'}^{\text{bottle}}\end{split}\]</div>
<div class="math notranslate nohighlight" id="equation-dloglik-bottleneck-dobserved-phenotype">
<span class="eqno">(21)<a class="headerlink" href="#equation-dloglik-bottleneck-dobserved-phenotype" title="Permalink to this equation">¶</a></span>\[\begin{split}\frac{\partial \mathcal{L}}
     {\partial p\left(v'\right)}
&amp;=&amp;
\sum_{v=1}^V
\frac{\partial n_v^{\text{bottle}}}{\partial p\left(v'\right)}
\ln\left( N_{\text{bottle}} f_v^{\text{pre}} \right)
-
\frac{\partial n_v^{\text{bottle}}}{\partial p\left(v'\right)}
\psi_0\left(n_v^{\text{bottle}} + 1\right)
\\
&amp;=&amp;
\left(\ln 2\right) f_{v'}^{\text{pre}} 2^{p\left(v'\right)}
N_{\text{bottle}}
\left(\sum_{v=1}^V
      \frac{f_v^{\text{post}}}
           {2^{p\left(v\right)}}
      \left[\ln\left(N_{\text{bottle}} f_v^{\text{pre}}\right) -
            \psi_0\left(n_v^{\text{bottle}} + 1\right)
            \right]
      \right)
- \left(\ln 2\right) n_{v'}^{\text{bottle}}
  \left[\ln\left(N_{\text{bottle}} f_{v'}^{\text{pre}}\right) -
        \psi_0\left(n_{v'}^{\text{bottle}} + 1\right)
        \right]\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(\psi_0\)</span> is the
<a class="reference external" href="https://en.wikipedia.org/wiki/Digamma_function">digamma function</a>.</p>
<p>Derivative of <a class="reference internal" href="#monotonic-spline-epistasis-function"><span class="std std-ref">Monotonic spline epistasis function</span></a> with respect to its
parameters:</p>
<div class="math notranslate nohighlight" id="equation-dspline-epistasis-dcalpha">
<span class="eqno">(22)<a class="headerlink" href="#equation-dspline-epistasis-dcalpha" title="Permalink to this equation">¶</a></span>\[\frac{\partial g_k\left(x\right)}{\partial c_{\alpha}^k} = 1\]</div>
<div class="math notranslate nohighlight" id="equation-dspline-epistasis-dalpham">
<span class="eqno">(23)<a class="headerlink" href="#equation-dspline-epistasis-dalpham" title="Permalink to this equation">¶</a></span>\[\frac{\partial g_k\left(x\right)}{\partial \alpha^k_m}
= I^k_m\left(x\right)\]</div>
</div>
</div>
<div class="section" id="detailed-documentation-of-models">
<h2><a class="toc-backref" href="#id28">Detailed documentation of models</a><a class="headerlink" href="#detailed-documentation-of-models" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dms_variants.globalepistasis.AbstractEpistasis">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">AbstractEpistasis</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">abc.ABC</span></code></p>
<p>Abstract base class for epistasis models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binarymap</strong> (<a class="reference internal" href="dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap" title="dms_variants.binarymap.BinaryMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.binarymap.BinaryMap</span></code></a>) – Contains the variants, their functional scores, and score variances.</p></li>
<li><p><strong>n_latent_phenotypes</strong> (<em>int</em>) – Number of distinct latent phenotypes. See <a class="reference internal" href="#multi-latent"><span class="std std-ref">Multiple latent phenotypes</span></a>.</p></li>
<li><p><strong>model_one_less_latent</strong> (None or <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a>) – If <cite>n_latent_phenotypes</cite> &gt; 1, should be a fit model of the same
type fit the same <cite>binarymap</cite> for one less latent phenotype. This
is used to initialize the parameters. See <a class="reference internal" href="#fitting-multi-latent"><span class="std std-ref">Fitting multiple latent phenotypes</span></a>.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is an abstract base class. It implements most of the epistasis model
functionality, but requires subclasses to define the actual
<a class="reference internal" href="#global-epistasis-function"><span class="std std-ref">Global epistasis function</span></a> and <a class="reference internal" href="#likelihood-calculation"><span class="std std-ref">Likelihood calculation</span></a>.</p>
</div>
<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.add_phenotypes_to_df">
<code class="sig-name descname">add_phenotypes_to_df</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">*</em>, <em class="sig-param">substitutions_col=None</em>, <em class="sig-param">latent_phenotype_col='latent_phenotype'</em>, <em class="sig-param">observed_phenotype_col='observed_phenotype'</em>, <em class="sig-param">phenotype_col_overwrite=False</em>, <em class="sig-param">unknown_as_nan=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.add_phenotypes_to_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.add_phenotypes_to_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Add predicted phenotypes to data frame of variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame containing variants.</p></li>
<li><p><strong>substitutions_col</strong> (<em>str</em><em> or </em><em>None</em>) – Column in <cite>df</cite> giving variants as substitution strings in format
that can be processed by <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.binarymap" title="dms_variants.globalepistasis.AbstractEpistasis.binarymap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code></a>.
If <cite>None</cite>, defaults to the <cite>substitutions_col</cite> attribute of
that binary map.</p></li>
<li><p><strong>latent_phenotype_col</strong> (<em>str</em>) – Column(s) added to <cite>df</cite> containing predicted latent phenotypes.
If there are multiple latent phenotypes, this string is suffixed
with the latent phenotype number (i.e., ‘latent_phenotype_1’).</p></li>
<li><p><strong>observed_phenotype_col</strong> (<em>str</em>) – Column added to <cite>df</cite> containing predicted observed phenotypes.</p></li>
<li><p><strong>phenotype_col_overwrite</strong> (<em>bool</em>) – If the specified latent or observed phenotype column already
exist in <cite>df</cite>, overwrite it? If <cite>False</cite>, raise an error.</p></li>
<li><p><strong>unknown_as_nan</strong> (<em>bool</em>) – If some of the substitutions in a variant are not present in
the model (not in <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.binarymap" title="dms_variants.globalepistasis.AbstractEpistasis.binarymap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code></a>) set the
phenotypes to <cite>nan</cite> (not a number)? If <cite>False</cite>, raise an error.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A copy of <cite>df</cite> with the phenotypes added. Phenotypes are predicted
based on the current state of the model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.aic">
<em class="property">property </em><code class="sig-name descname">aic</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.aic" title="Permalink to this definition">¶</a></dt>
<dd><p>Aikake Information Criterion given current log likelihood.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.binarymap">
<em class="property">property </em><code class="sig-name descname">binarymap</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.binarymap" title="Permalink to this definition">¶</a></dt>
<dd><p>Variants to model.</p>
<p>The binary map is set during initialization of the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap" title="dms_variants.binarymap.BinaryMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.binarymap.BinaryMap</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.enrichments">
<code class="sig-name descname">enrichments</code><span class="sig-paren">(</span><em class="sig-param">observed_phenotypes</em>, <em class="sig-param">base=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.enrichments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.enrichments" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculated enrichment ratios from observed phenotypes.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In many cases, the functional scores used to fit the model are the
logarithm (most commonly base 2) of experimentally observed enrichments
For example, this is how functional scores are calculated by
<a class="reference internal" href="dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.func_scores" title="dms_variants.codonvarianttable.CodonVariantTable.func_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dms_variants.codonvarianttable.CodonVariantTable.func_scores()</span></code></a>.
In that case, the predicted enrichment value <span class="math notranslate nohighlight">\(E\left(v\right)\)</span>
for each variant <span class="math notranslate nohighlight">\(v\)</span> can be computed from the observed phenotype
<span class="math notranslate nohighlight">\(p\left(v\right)\)</span> as:</p>
<div class="math notranslate nohighlight">
\[E\left(v\right) = B^{p\left(v\right) - p\left(\rm{wt}\right)}\]</div>
<p>where <span class="math notranslate nohighlight">\(p\left(\rm{wt}\right)\)</span> is the observed phenotype
of wildtype, and <span class="math notranslate nohighlight">\(B\)</span> is the base for the exponent (by default
<span class="math notranslate nohighlight">\(B = 2\)</span>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>observed_phenotypes</strong> (<em>float</em><em> or </em><em>numpy.ndarray</em>) – The observed phenotypes.</p></li>
<li><p><strong>base</strong> (<em>float</em>) – The base for the exponent used to convert observed phenotypes
to enrichments.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The enrichments.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float or numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.epistasis_func">
<em class="property">abstract </em><code class="sig-name descname">epistasis_func</code><span class="sig-paren">(</span><em class="sig-param">latent_phenotype</em>, <em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.epistasis_func"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.epistasis_func" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#global-epistasis-function"><span class="std std-ref">Global epistasis function</span></a> <span class="math notranslate nohighlight">\(g\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>latent_phenotype</strong> (<em>numpy.ndarray</em>) – Latent phenotype(s) of one or more variants.</p></li>
<li><p><strong>k</strong> (<em>int</em><em> or </em><em>None</em>) – Latent phenotype number (1 &lt;= <cite>k</cite> &lt;= <cite>n_latent_phenotypes</cite>),
or can be <cite>None</cite> if just one latent phenotype. See
Eq. <a class="reference internal" href="#multi-latent"><span class="std std-ref">Multiple latent phenotypes</span></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Result of applying global epistasis function <span class="math notranslate nohighlight">\(g_k\)</span> to latent
phenotypes.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.epistasis_func_params_dict">
<em class="property">property </em><code class="sig-name descname">epistasis_func_params_dict</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.epistasis_func_params_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Parameters for the <a class="reference internal" href="#global-epistasis-function"><span class="std std-ref">Global epistasis function</span></a>.</p>
<p>Maps names of parameters defining the global epistasis function to
their current values.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">use_grad=True</em>, <em class="sig-param">optimize_method='L-BFGS-B'</em>, <em class="sig-param">ftol=1e-07</em>, <em class="sig-param">clearcache=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit all model params to maximum likelihood values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>use_grad</strong> (<em>bool</em>) – Use analytical gradients to help with fitting.</p></li>
<li><p><strong>optimize_method</strong> (<em>{'L-BFGS-B'</em><em>, </em><em>'TNC'}</em>) – Optimization method used by <cite>scipy.optimize.minimize</cite>.</p></li>
<li><p><strong>ftol</strong> (<em>float</em>) – Function convergence tolerance for optimization, used by
<cite>scipy.optimize.minimize</cite>.</p></li>
<li><p><strong>clearcache</strong> (<em>bool</em>) – Clear the cache after model fitting? This slightly increases
the time needed to compute properties after fitting, but
greatly saves memory usage.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The results of optimizing the full model.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>scipy.optimize.OptimizeResult</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.latent_effects_df">
<em class="property">property </em><code class="sig-name descname">latent_effects_df</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.latent_effects_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Latent effects of mutations.</p>
<p>For each single mutation in <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.binarymap" title="dms_variants.globalepistasis.AbstractEpistasis.binarymap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code></a>, gives
current predicted latent effect of that mutation. If there are multiple
latent phenotypes (<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.n_latent_phenotypes</span></code></a> &gt; 1),
also indicates the phenotype number of the latent effect.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.latent_phenotype_wt">
<code class="sig-name descname">latent_phenotype_wt</code><span class="sig-paren">(</span><em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.latent_phenotype_wt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.latent_phenotype_wt" title="Permalink to this definition">¶</a></dt>
<dd><p>Latent phenotype of wildtype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>k</strong> (<em>int</em><em> or </em><em>None</em>) – Which latent phenotype to get (1 &lt;= k &lt;=
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.n_latent_phenotypes</span></code></a>). If there
is just one latent phenotype, can also be <cite>None</cite>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Wildtype latent phenotype, which is <span class="math notranslate nohighlight">\(\beta_{\rm{wt}}\)</span> in
Eq. <a class="reference internal" href="#equation-latent-phenotype">(1)</a> or <span class="math notranslate nohighlight">\(\beta_{\rm{wt}}^k\)</span> in
Eq. <a class="reference internal" href="#equation-latent-phenotype-multi">(5)</a>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.likelihood_calc_params_dict">
<em class="property">property </em><code class="sig-name descname">likelihood_calc_params_dict</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.likelihood_calc_params_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Parameters for the <a class="reference internal" href="#likelihood-calculation"><span class="std std-ref">Likelihood calculation</span></a>.</p>
<p>Maps names of parameters defining the likelihood calculation to
their current values.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.loglik">
<em class="property">abstract property </em><code class="sig-name descname">loglik</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.loglik" title="Permalink to this definition">¶</a></dt>
<dd><p>Current log likelihood of model.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes">
<em class="property">property </em><code class="sig-name descname">n_latent_phenotypes</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>number of latent phenotypes, see <a class="reference internal" href="#multi-latent"><span class="std std-ref">Multiple latent phenotypes</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.nparams">
<em class="property">property </em><code class="sig-name descname">nparams</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.nparams" title="Permalink to this definition">¶</a></dt>
<dd><p>Total number of parameters in model.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.phenotypes_df">
<em class="property">property </em><code class="sig-name descname">phenotypes_df</code><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.phenotypes_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Phenotypes of variants used to fit model.</p>
<p>For each variant in <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.binarymap" title="dms_variants.globalepistasis.AbstractEpistasis.binarymap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code></a>, gives
the current predicted latent and observed phenotypes as well
as the functional score and its variance.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.phenotypes_frombinary">
<code class="sig-name descname">phenotypes_frombinary</code><span class="sig-paren">(</span><em class="sig-param">binary_variants</em>, <em class="sig-param">phenotype</em>, <em class="sig-param">*</em>, <em class="sig-param">wt_col=False</em>, <em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.phenotypes_frombinary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.phenotypes_frombinary" title="Permalink to this definition">¶</a></dt>
<dd><p>Phenotypes from binary variant representations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binary_variants</strong> (<em>scipy.sparse.csr.csr_matrix</em><em> or </em><em>numpy.ndarray</em>) – Binary variants in form used by
<a class="reference internal" href="dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap" title="dms_variants.binarymap.BinaryMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.binarymap.BinaryMap</span></code></a>.</p></li>
<li><p><strong>phenotype</strong> (<em>{'latent'</em><em>, </em><em>'observed'}</em>) – Calculate the latent or observed phenotype.</p></li>
<li><p><strong>wt_col</strong> (<em>bool</em>) – Set to <cite>True</cite> if <cite>binary_variants</cite> contains a terminal
column of ones to enable calculations in the form given
by Eq. <a class="reference internal" href="#equation-latent-phenotype-wt-vec">(13)</a>.</p></li>
<li><p><strong>k</strong> (<em>int</em><em> or </em><em>None</em>) – If <cite>phenotype</cite> is ‘latent’, which latent phenotype to use (1 &lt;= k
(&lt;= <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.n_latent_phenotypes</span></code></a>). If there
is just one latent phenotype, can also be <cite>None</cite>. Has no meaning
if <cite>phenotype</cite> is ‘observed’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Latent phenotypes calculated using Eq. <a class="reference internal" href="#equation-latent-phenotype">(1)</a> or
observed phenotypes calculated using Eq. <a class="reference internal" href="#equation-observed-phenotype">(2)</a>
(or Eqs. <a class="reference internal" href="#equation-latent-phenotype-multi">(5)</a> or
<a class="reference internal" href="#equation-observed-phenotype-multi">(6)</a>).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.preferences">
<code class="sig-name descname">preferences</code><span class="sig-paren">(</span><em class="sig-param">phenotype</em>, <em class="sig-param">base</em>, <em class="sig-param">*</em>, <em class="sig-param">missing='average'</em>, <em class="sig-param">exclude_chars=('*'</em>, <em class="sig-param">)</em>, <em class="sig-param">returnformat='wide'</em>, <em class="sig-param">stringency_param=1</em>, <em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.preferences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.preferences" title="Permalink to this definition">¶</a></dt>
<dd><p>Get preference of each site for each character.</p>
<p>Use the latent or observed phenotype to estimate the preference
<span class="math notranslate nohighlight">\(\pi_{r,a}\)</span> of each site <span class="math notranslate nohighlight">\(r\)</span> for each character (e.g.,
amino acid) <span class="math notranslate nohighlight">\(a\)</span>. These preferences can be displayed in logo plots
or used as input to <a class="reference external" href="https://jbloomlab.github.io/phydms/">phydms</a>
in experimentally informed substitution models.</p>
<p>The preferences are calculated from the phenotypes as follows. Let
<span class="math notranslate nohighlight">\(p_{r,a}\)</span> be the phenotype of the variant with the single
mutation of site <span class="math notranslate nohighlight">\(r\)</span> to <span class="math notranslate nohighlight">\(a\)</span> (when <span class="math notranslate nohighlight">\(a\)</span> is the wildtype
character, then <span class="math notranslate nohighlight">\(p_{r,a}\)</span> is the phenotype of the wildtype
sequence). Then the preference <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span> is defined as</p>
<div class="math notranslate nohighlight">
\[\pi_{r,a} = \frac{b^{p_{r,a}}}{\sum_{a'} b^{p_{r,a'}}}\]</div>
<p>where <span class="math notranslate nohighlight">\(b\)</span> is the base for the exponent. This definition
ensures that the preferences sum to one at each site.</p>
<p>The alphabet from which the characters are drawn and the site
numbers are extracted from <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.binarymap" title="dms_variants.globalepistasis.AbstractEpistasis.binarymap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The “flatness” of the preferences is determined by the exponent base.
A smaller <cite>base</cite> yields flatter preferences. There is no obvious “best”
<cite>base</cite> as different values correspond to different linear scalings of
of the phenotype. A recommended approach is simply to choose a value of
<cite>base</cite> and then re-scale the preferences by using
<a class="reference external" href="https://jbloomlab.github.io/phydms/">phydms</a> to optimize a
stringency parameter (<a class="reference external" href="https://peerj.com/articles/3657">see here</a>).
The stringency parameter and the <cite>base</cite> chosen here both apply the
same transformation to the data: linear scaling of the phenotypes.
But note that <a class="reference external" href="https://jbloomlab.github.io/phydms/">phydms</a>
has an upper bound on the largest stringency parameter it can fit,
so if you are hitting this upper bound then pre-scale the preferences
to be less flat by using a larger value of <cite>base</cite>. In particular,
the latent phenotypes from many of the epistasis models are scaled
during fitting to have a relatively tight range, so you may need a
large value of <cite>base</cite> such as 50.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>phenotype</strong> (<em>{'observed'</em><em>, </em><em>'latent'}</em>) – Calculate the preferences from observed or latent phenotypes?
Note that if there are multiple latent phenotypes, you must
also set <cite>k</cite>.</p></li>
<li><p><strong>base</strong> (<em>float</em>) – Base to which the exponent is taken in computing the preferences.</p></li>
<li><p><strong>missing</strong> (<em>{'average'</em><em>, </em><em>'site_average'</em><em>, </em><em>'error'}</em>) – What to do when there is no estimate of the phenotype for one of
the single mutants? Estimate the phenotype as the average of
all single mutants, as the average of all single mutants at that
site, or raise an error.</p></li>
<li><p><strong>exclude_chars</strong> (<em>tuple</em><em> or </em><em>list</em>) – Characters to exclude when calculating preferences (and when
averaging values for missing mutants). For instance, you might
want to exclude stop codons.</p></li>
<li><p><strong>returnformat</strong> (<em>{'tidy'</em><em>, </em><em>'wide'}</em>) – Return preferences in tidy or wide format data frame.</p></li>
<li><p><strong>stringency_param</strong> (<em>float</em>) – Re-scale preferences by this stringency parameter. This
involves raising each preference to the power of
<cite>stringency_param</cite>, and then re-normalizes. A similar
effect can be achieved by changing <cite>base</cite>.</p></li>
<li><p><strong>k</strong> (<em>int</em><em> or </em><em>None</em>) – Which latent phenotype to use (1 &lt;= k &lt;=
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.n_latent_phenotypes</span></code></a>). If there
is just one latent phenotype, can also be <cite>None</cite>. Has no
meaning if <cite>phenotype</cite> is ‘observed’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Data frame where first column is named ‘site’, other columns are
named for each character, and rows give preferences for each site.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.AbstractEpistasis.single_mut_effects">
<code class="sig-name descname">single_mut_effects</code><span class="sig-paren">(</span><em class="sig-param">phenotype</em>, <em class="sig-param">*</em>, <em class="sig-param">include_wildtype=True</em>, <em class="sig-param">standardize_range=True</em>, <em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#AbstractEpistasis.single_mut_effects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.AbstractEpistasis.single_mut_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Effects of single mutations on latent or observed phenotype.</p>
<p>For the effects on observed phenotype, this is how much the mutation
changes the observed phenotype relative to wildtype. Effects are
reported only for mutations present in <cite>AbstractEpistasis.binarymap</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>phenotype</strong> (<em>{'latent'</em><em>, </em><em>'observed'}</em>) – Get effect on this phenotype. If there are multiple latent
phenotypes, you must also set <cite>k</cite>.</p></li>
<li><p><strong>include_wildtype</strong> (<em>bool</em>) – Include the effect of “mutating” to wildtype identity at a site
(always zero).</p></li>
<li><p><strong>standardize_range</strong> (<em>bool</em>) – Scale effects so that the mean absolute value effect is one
(scaling is done before including wildtype).</p></li>
<li><p><strong>k</strong> (<em>int</em><em> or </em><em>None</em>) – Which latent phenotype to use (1 &lt;= k &lt;=
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.n_latent_phenotypes</span></code></a>). If there
is just one latent phenotype, can also be <cite>None</cite>. Has no
meaning if <cite>phenotype</cite> is ‘observed’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>The effects of all single mutations. Columns are:</p>
<blockquote>
<div><ul class="simple">
<li><p>’mutation’: mutation as str</p></li>
<li><p>’wildtype’: wildtype identity at site</p></li>
<li><p>’site’: site number</p></li>
<li><p>’mutant’: mutant identity at site</p></li>
<li><p>’effect’: effect of mutation on latent or observed phenotype</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.BottleneckLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">BottleneckLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">bottleneck</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em>, <em class="sig-param">pseudocount=0.5</em>, <em class="sig-param">base=2.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#BottleneckLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.BottleneckLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.AbstractEpistasis</span></code></a></p>
<p>Bottleneck likelihood calculation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Subclass of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> that implements the
<a class="reference internal" href="#bottleneck-likelihood"><span class="std std-ref">Bottleneck likelihood</span></a>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.binarymap" title="dms_variants.globalepistasis.AbstractEpistasis.binarymap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code></a> must have non-<cite>None</cite>
counts in <a class="reference internal" href="dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap.n_pre" title="dms_variants.binarymap.BinaryMap.n_pre"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dms_variants.binarymap.BinaryMap.n_pre</span></code></a> and
<a class="reference internal" href="dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap.n_post" title="dms_variants.binarymap.BinaryMap.n_post"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dms_variants.binarymap.BinaryMap.n_post</span></code></a>, since
as described in <a class="reference internal" href="#bottleneck-likelihood"><span class="std std-ref">Bottleneck likelihood</span></a>, the model is
actually fit to the <a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood.f_pre" title="dms_variants.globalepistasis.BottleneckLikelihood.f_pre"><code class="xref py py-attr docutils literal notranslate"><span class="pre">BottleneckLikelihood.f_pre</span></code></a> and
<a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood.f_post" title="dms_variants.globalepistasis.BottleneckLikelihood.f_post"><code class="xref py py-attr docutils literal notranslate"><span class="pre">BottleneckLikelihood.f_post</span></code></a> values calculated from
these counts.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bottleneck</strong> (<em>float</em>) – The estimated size of the experimental bottleneck between the
pre- and post-selection conditions. This is the <span class="math notranslate nohighlight">\(N_{\rm{bottle}}\)</span>
parameter described in <a class="reference internal" href="#bottleneck-likelihood"><span class="std std-ref">Bottleneck likelihood</span></a>.</p></li>
<li><p><strong>pseudocount</strong> (<em>float</em>) – The pseudocount used when converting the counts to frequencies
vi Eq. <a class="reference internal" href="#equation-f-v-pre-post">(10)</a>.</p></li>
<li><p><strong>base</strong> (<em>float</em>) – The exponent base in Eq. <a class="reference internal" href="#equation-n-v-bottle">(12)</a> used when exponentiating
the observed phenotypes <span class="math notranslate nohighlight">\(p\left(v\right)\)</span>. It is written
as 2 in Eq. <a class="reference internal" href="#equation-n-v-bottle">(12)</a>.</p></li>
</ul>
</dd>
</dl>
<dl class="method">
<dt id="dms_variants.globalepistasis.BottleneckLikelihood.bottleneck">
<em class="property">property </em><code class="sig-name descname">bottleneck</code><a class="headerlink" href="#dms_variants.globalepistasis.BottleneckLikelihood.bottleneck" title="Permalink to this definition">¶</a></dt>
<dd><p>Bottleneck pre- to post-selection, <span class="math notranslate nohighlight">\(N_{\rm{bottle}]\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.BottleneckLikelihood.f_post">
<em class="property">property </em><code class="sig-name descname">f_post</code><a class="headerlink" href="#dms_variants.globalepistasis.BottleneckLikelihood.f_post" title="Permalink to this definition">¶</a></dt>
<dd><p>Post-selection frequency of each variant.</p>
<p>The <span class="math notranslate nohighlight">\(f_v^{\rm{post}}\)</span> values in Eq. <a class="reference internal" href="#equation-f-v-pre-post">(10)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.BottleneckLikelihood.f_pre">
<em class="property">property </em><code class="sig-name descname">f_pre</code><a class="headerlink" href="#dms_variants.globalepistasis.BottleneckLikelihood.f_pre" title="Permalink to this definition">¶</a></dt>
<dd><p>Pre-selection frequency of each variant.</p>
<p>The <span class="math notranslate nohighlight">\(f_v^{\rm{pre}}\)</span> values in Eq. <a class="reference internal" href="#equation-f-v-pre-post">(10)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.BottleneckLikelihood.loglik">
<em class="property">property </em><code class="sig-name descname">loglik</code><a class="headerlink" href="#dms_variants.globalepistasis.BottleneckLikelihood.loglik" title="Permalink to this definition">¶</a></dt>
<dd><p>Current log likelihood from Eq. <a class="reference internal" href="#equation-loglik-bottleneck">(11)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.CauchyLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">CauchyLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#CauchyLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.CauchyLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.AbstractEpistasis</span></code></a></p>
<p>Cauchy likelihood calculation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Subclass of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> that implements the
<a class="reference internal" href="#cauchy-likelihood"><span class="std std-ref">Cauchy likelihood</span></a>.</p>
</div>
<dl class="method">
<dt id="dms_variants.globalepistasis.CauchyLikelihood.loglik">
<em class="property">property </em><code class="sig-name descname">loglik</code><a class="headerlink" href="#dms_variants.globalepistasis.CauchyLikelihood.loglik" title="Permalink to this definition">¶</a></dt>
<dd><p>Current log likelihood from Eq. <a class="reference internal" href="#equation-loglik-cauchy">(9)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="dms_variants.globalepistasis.EpistasisFittingError">
<em class="property">exception </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">EpistasisFittingError</code><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#EpistasisFittingError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.EpistasisFittingError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Error fitting an epistasis model.</p>
</dd></dl>

<dl class="exception">
<dt id="dms_variants.globalepistasis.EpistasisFittingWarning">
<em class="property">exception </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">EpistasisFittingWarning</code><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#EpistasisFittingWarning"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.EpistasisFittingWarning" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Warning</span></code></p>
<p>Warning when fitting epistasis model.</p>
</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.GaussianLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">GaussianLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#GaussianLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.GaussianLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.AbstractEpistasis</span></code></a></p>
<p>Gaussian likelihood calculation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Subclass of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> that implements the
<a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a>.</p>
</div>
<dl class="method">
<dt id="dms_variants.globalepistasis.GaussianLikelihood.loglik">
<em class="property">property </em><code class="sig-name descname">loglik</code><a class="headerlink" href="#dms_variants.globalepistasis.GaussianLikelihood.loglik" title="Permalink to this definition">¶</a></dt>
<dd><p>Current log likelihood from Eq. <a class="reference internal" href="#equation-loglik-gaussian">(7)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.MonotonicSplineEpistasis">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">MonotonicSplineEpistasis</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em>, <em class="sig-param">spline_order=3</em>, <em class="sig-param">meshpoints=4</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#MonotonicSplineEpistasis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.AbstractEpistasis</span></code></a></p>
<p>Monotonic spline global epistasis model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Subclass of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> that implements the
<a class="reference internal" href="#monotonic-spline-epistasis-function"><span class="std std-ref">Monotonic spline epistasis function</span></a>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spline_order</strong> (<em>int</em>) – Order of the I-splines defining the global epistasis function.</p></li>
<li><p><strong>meshpoints</strong> (<em>int</em>) – Number of evenly spaced mesh points for the I-spline defining the
global epistasis function.</p></li>
</ul>
</dd>
</dl>
<dl class="method">
<dt id="dms_variants.globalepistasis.MonotonicSplineEpistasis.alpha_ms">
<code class="sig-name descname">alpha_ms</code><span class="sig-paren">(</span><em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#MonotonicSplineEpistasis.alpha_ms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis.alpha_ms" title="Permalink to this definition">¶</a></dt>
<dd><p><span class="math notranslate nohighlight">\(\alpha_m\)</span> in Eq. <a class="reference internal" href="#equation-monotonicspline">(4)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>k</strong> (<em>int</em><em> or </em><em>None</em>) – Which global epistasis function to get I-splines for (1 &lt;= k &lt;=
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.n_latent_phenotypes</span></code></a>). If there
is just one latent phenotype, can also be <cite>None</cite>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><span class="math notranslate nohighlight">\(\alpha_m\)</span> values for global epistasis function <cite>k</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.MonotonicSplineEpistasis.c_alpha">
<code class="sig-name descname">c_alpha</code><span class="sig-paren">(</span><em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#MonotonicSplineEpistasis.c_alpha"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis.c_alpha" title="Permalink to this definition">¶</a></dt>
<dd><p><span class="math notranslate nohighlight">\(c_{\alpha}\)</span> in Eq. <a class="reference internal" href="#equation-monotonicspline">(4)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>k</strong> (<em>int</em><em> or </em><em>None</em>) – Which global epistasis function to get I-splines for (1 &lt;= k &lt;=
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes" title="dms_variants.globalepistasis.AbstractEpistasis.n_latent_phenotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.n_latent_phenotypes</span></code></a>). If there
is just one latent phenotype, can also be <cite>None</cite>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><span class="math notranslate nohighlight">\(c_{\alpha}\)</span> for global epistasis function <cite>k</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.globalepistasis.MonotonicSplineEpistasis.epistasis_func">
<code class="sig-name descname">epistasis_func</code><span class="sig-paren">(</span><em class="sig-param">latent_phenotype</em>, <em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#MonotonicSplineEpistasis.epistasis_func"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis.epistasis_func" title="Permalink to this definition">¶</a></dt>
<dd><p>Global epistasis function <span class="math notranslate nohighlight">\(g\)</span> in Eq. <a class="reference internal" href="#equation-monotonicspline">(4)</a>.</p>
<p>Concrete implementation of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.epistasis_func" title="dms_variants.globalepistasis.AbstractEpistasis.epistasis_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AbstractEpistasis.epistasis_func()</span></code></a>.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.MonotonicSplineEpistasisBottleneckLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">MonotonicSplineEpistasisBottleneckLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">bottleneck</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em>, <em class="sig-param">spline_order=3</em>, <em class="sig-param">meshpoints=4</em>, <em class="sig-param">pseudocount=0.5</em>, <em class="sig-param">base=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#MonotonicSplineEpistasisBottleneckLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.MonotonicSplineEpistasisBottleneckLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.MonotonicSplineEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood" title="dms_variants.globalepistasis.BottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.BottleneckLikelihood</span></code></a></p>
<p>Monotonic spline global epistasis model with bottleneck likelihood.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class implements the <a class="reference internal" href="#monotonic-spline-epistasis-function"><span class="std std-ref">Monotonic spline epistasis function</span></a>
with a <a class="reference internal" href="#bottleneck-likelihood"><span class="std std-ref">Bottleneck likelihood</span></a>. See documentation for the base
classes <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood" title="dms_variants.globalepistasis.BottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">BottleneckLikelihood</span></code></a>,
and <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> for details.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.MonotonicSplineEpistasisCauchyLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">MonotonicSplineEpistasisCauchyLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em>, <em class="sig-param">spline_order=3</em>, <em class="sig-param">meshpoints=4</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#MonotonicSplineEpistasisCauchyLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.MonotonicSplineEpistasisCauchyLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.MonotonicSplineEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.CauchyLikelihood" title="dms_variants.globalepistasis.CauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.CauchyLikelihood</span></code></a></p>
<p>Monotonic spline global epistasis model with Cauchy likelihood.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class implements the <a class="reference internal" href="#monotonic-spline-epistasis-function"><span class="std std-ref">Monotonic spline epistasis function</span></a>
with a <a class="reference internal" href="#cauchy-likelihood"><span class="std std-ref">Cauchy likelihood</span></a>. See documentation for the base
classes <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.CauchyLikelihood" title="dms_variants.globalepistasis.CauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">CauchyLikelihood</span></code></a>,
and <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> for details.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.MonotonicSplineEpistasisGaussianLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">MonotonicSplineEpistasisGaussianLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em>, <em class="sig-param">spline_order=3</em>, <em class="sig-param">meshpoints=4</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#MonotonicSplineEpistasisGaussianLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.MonotonicSplineEpistasisGaussianLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.MonotonicSplineEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.GaussianLikelihood" title="dms_variants.globalepistasis.GaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.GaussianLikelihood</span></code></a></p>
<p>Monotonic spline global epistasis model with Gaussian likelihood.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class implements the <a class="reference internal" href="#monotonic-spline-epistasis-function"><span class="std std-ref">Monotonic spline epistasis function</span></a>
with a <a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a>. See documentation for the base
classes <a class="reference internal" href="#dms_variants.globalepistasis.MonotonicSplineEpistasis" title="dms_variants.globalepistasis.MonotonicSplineEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">MonotonicSplineEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.GaussianLikelihood" title="dms_variants.globalepistasis.GaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">GaussianLikelihood</span></code></a>,
and <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> for details.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.NoEpistasis">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">NoEpistasis</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#NoEpistasis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.NoEpistasis" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.AbstractEpistasis</span></code></a></p>
<p>Non-epistatic model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Subclass of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> that implements the
<a class="reference internal" href="#no-epistasis-function"><span class="std std-ref">No epistasis function</span></a>.</p>
</div>
<dl class="method">
<dt id="dms_variants.globalepistasis.NoEpistasis.epistasis_func">
<code class="sig-name descname">epistasis_func</code><span class="sig-paren">(</span><em class="sig-param">latent_phenotype</em>, <em class="sig-param">k=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#NoEpistasis.epistasis_func"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.NoEpistasis.epistasis_func" title="Permalink to this definition">¶</a></dt>
<dd><p>Global epistasis function <span class="math notranslate nohighlight">\(g\)</span> in Eq. <a class="reference internal" href="#equation-noepistasis">(3)</a>.</p>
<p>Concrete implementation of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis.epistasis_func" title="dms_variants.globalepistasis.AbstractEpistasis.epistasis_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AbstractEpistasis.epistasis_func()</span></code></a>.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.NoEpistasisBottleneckLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">NoEpistasisBottleneckLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">bottleneck</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em>, <em class="sig-param">pseudocount=0.5</em>, <em class="sig-param">base=2.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#NoEpistasisBottleneckLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.NoEpistasisBottleneckLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.NoEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood" title="dms_variants.globalepistasis.BottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.BottleneckLikelihood</span></code></a></p>
<p>No-epistasis model with bottleneck likelihood.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class implements the <a class="reference internal" href="#no-epistasis-function"><span class="std std-ref">No epistasis function</span></a> with a
<a class="reference internal" href="#bottleneck-likelihood"><span class="std std-ref">Bottleneck likelihood</span></a>. See documentation for the base classes
<a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.BottleneckLikelihood" title="dms_variants.globalepistasis.BottleneckLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">BottleneckLikelihood</span></code></a>, and
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> for details.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.NoEpistasisCauchyLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">NoEpistasisCauchyLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#NoEpistasisCauchyLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.NoEpistasisCauchyLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.NoEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.CauchyLikelihood" title="dms_variants.globalepistasis.CauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.CauchyLikelihood</span></code></a></p>
<p>No-epistasis model with Cauchy likelihood.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class implements the <a class="reference internal" href="#no-epistasis-function"><span class="std std-ref">No epistasis function</span></a> with a
<a class="reference internal" href="#cauchy-likelihood"><span class="std std-ref">Cauchy likelihood</span></a>. See documentation for the base classes
<a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.CauchyLikelihood" title="dms_variants.globalepistasis.CauchyLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">CauchyLikelihood</span></code></a>, and
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> for details.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="dms_variants.globalepistasis.NoEpistasisGaussianLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">NoEpistasisGaussianLikelihood</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">*</em>, <em class="sig-param">n_latent_phenotypes=1</em>, <em class="sig-param">model_one_less_latent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#NoEpistasisGaussianLikelihood"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.NoEpistasisGaussianLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.NoEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.GaussianLikelihood" title="dms_variants.globalepistasis.GaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.globalepistasis.GaussianLikelihood</span></code></a></p>
<p>No-epistasis model with Gaussian likelihood.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class implements the <a class="reference internal" href="#no-epistasis-function"><span class="std std-ref">No epistasis function</span></a> with a
<a class="reference internal" href="#gaussian-likelihood"><span class="std std-ref">Gaussian likelihood</span></a>. See documentation for the base classes
<a class="reference internal" href="#dms_variants.globalepistasis.NoEpistasis" title="dms_variants.globalepistasis.NoEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">NoEpistasis</span></code></a>, <a class="reference internal" href="#dms_variants.globalepistasis.GaussianLikelihood" title="dms_variants.globalepistasis.GaussianLikelihood"><code class="xref py py-class docutils literal notranslate"><span class="pre">GaussianLikelihood</span></code></a>, and
<a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a> for details.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_variants.globalepistasis.fit_models">
<code class="sig-prename descclassname">dms_variants.globalepistasis.</code><code class="sig-name descname">fit_models</code><span class="sig-paren">(</span><em class="sig-param">binarymap</em>, <em class="sig-param">likelihood</em>, <em class="sig-param">*</em>, <em class="sig-param">bottleneck=None</em>, <em class="sig-param">max_latent_phenotypes=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/globalepistasis.html#fit_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.globalepistasis.fit_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit and compare global epistasis models.</p>
<p>This function is useful when you want to examine the fit of several
different models to the same data. It does the following:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Fits a non-epistatic model to the data.</p></li>
<li><p>Fits a global epistasis model with <span class="math notranslate nohighlight">\(K = 1\)</span> latent phenotypes
to the data. If the global epistasis model outperforms the no-
epistasis model by <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>, proceed to next step. Otherwise stop.</p></li>
<li><p>Fit a global epistasis model with <span class="math notranslate nohighlight">\(K = 2\)</span> latent phenotypes.
If this model outperforms (by <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>) the model with <span class="math notranslate nohighlight">\(K - 1\)</span>
latent phenotypes, repeat for <span class="math notranslate nohighlight">\(K = 3\)</span> etc until adding more
latent phenotypes no longer improves fit. Note that it only does
continues this process while <span class="math notranslate nohighlight">\(K \le\)</span> <cite>max_latent_phenotypes</cite>,
so set <cite>max_latent_phenotypes</cite> &gt; 1 if you want to fit multiple
latent phenotypes.</p></li>
</ol>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All of the fitting is done with the same likelihood-calculation method
because you can <strong>not</strong> compare models fit with different likelihood-
calculation methods.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binarymap</strong> (<a class="reference internal" href="dms_variants.binarymap.html#dms_variants.binarymap.BinaryMap" title="dms_variants.binarymap.BinaryMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.binarymap.BinaryMap</span></code></a>) – Contains the variants, their functional scores, and score variances.
The models are fit to these data.</p></li>
<li><p><strong>likelihood</strong> (<em>{'Gaussian'</em><em>, </em><em>'Cauchy'</em><em>, </em><em>'Bottleneck'}</em>) – Likelihood calculation method to use when fitting models. See
<a class="reference internal" href="#likelihood-calculation"><span class="std std-ref">Likelihood calculation</span></a>.</p></li>
<li><p><strong>bottleneck</strong> (<em>float</em><em> or </em><em>None</em>) – Required if using ‘Bottleneck’ <cite>likelihood</cite>. In that case, is
the experimentally estimated bottleneck between the pre-
and post-selection conditions.</p></li>
<li><p><strong>max_latent_phenotypes</strong> (<em>int</em>) – Maximum number of latent phenotypes that are potentially be fit.
See the <span class="math notranslate nohighlight">\(K\)</span> parameter in <a class="reference internal" href="#multi-latent"><span class="std std-ref">Multiple latent phenotypes</span></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>Summarizes the results of the model fitting and contains the
fit models. Columns are:</p>
<blockquote>
<div><ul class="simple">
<li><p>’description’: description of model</p></li>
<li><p>’n_latent_phenotypes’: number of latent phenotypes in model</p></li>
<li><p>’AIC’: <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a></p></li>
<li><p>’nparams’: number of parameters</p></li>
<li><p>’log_likelihood’: log likelihood</p></li>
<li><p>’model’: the actual model (subclass of <a class="reference internal" href="#dms_variants.globalepistasis.AbstractEpistasis" title="dms_variants.globalepistasis.AbstractEpistasis"><code class="xref py py-class docutils literal notranslate"><span class="pre">AbstractEpistasis</span></code></a>)</p></li>
<li><p>’fitting_time’: time in seconds that it took to fit model</p></li>
</ul>
</div></blockquote>
<p>The data frame is sorted from best to worst model by <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_variants">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dms_variants.html">dms_variants package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dms_variants.html#dms-variants">dms_variants</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="dms_variants.html#submodules">Submodules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="dms_variants.html">dms_variants package</a><ul>
      <li>Previous: <a href="dms_variants.fastq.html" title="previous chapter">fastq</a></li>
      <li>Next: <a href="dms_variants.illuminabarcodeparser.html" title="next chapter">illuminabarcodeparser</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2020.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_variants.globalepistasis.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>