
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Fitting gobal epistasis models with experimental “noise” &#8212; dms_variants 1.4.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Formatting CodonVariantTable plots" href="codonvariant_plot_formatting.html" />
    <link rel="prev" title="Parse barcodes" href="parsebarcodes_sim_data.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Fitting-gobal-epistasis-models-with-experimental-“noise”">
<h1>Fitting gobal epistasis models with experimental “noise”<a class="headerlink" href="#Fitting-gobal-epistasis-models-with-experimental-“noise”" title="Permalink to this headline">¶</a></h1>
<p>The global epistasis models described by <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a> fits experimentally measured functional scores for the variants using a Gaussian-likelihood function that assumes that deviations from the models are normally distributed (see the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html">globalepistasis module documentation</a> for more details). Those models also accommodate estimates of the noise for each
functional score measurement, which are easy to make if the noise is due to statistical processes that are simple to model (such as finite sequencing depth leading to sampling statistics in the deep sequencing counts).</p>
<p>But often real deep mutational scanning experiments have other sources of experimental “noise.” For instance, the sequencing used to determine the mutations present in a variant might occassionally wrong, leading to mis-calling of variant sequences. Or (more commonly in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/27271655">Bloom lab influenza virus experiments</a>) there may be experimental bottlenecks in the number of variants that are passaged from the pre-selection to post-selection conditions. In
this last case, the noise occurs when the bottleneck randomly influences which pre-selection variants get a “chance” to go to the post-selection condition, not in the sequencing counts estimating the pre- and post-selection frequencies.</p>
<p>When there are bottlenecks or other noise, the Gaussian likelihoods may no longer be a good way to fit the data. Therefore, the global epistasis models allow two other ways to compute the likelihoods. These likelihood calculation methods are described in the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#likelihood-calculation">likelihood calculation section of the globalepistasis module</a>. Briefly, they are:</p>
<ol class="arabic simple">
<li><p>The “bottleneck” likelihood, which explicitly models a bottleneck between the pre- and post-selection conditions. You should use this likelihood when there is a bottleneck at this step that is substantially smaller than the total sequencing depth, and which you think contributes most of the “noise” in the experiment.</p></li>
<li><p>The Cauchy likelihood, which is like the Gaussian likelihood but has a “fatter tail” and so is less affected by outliers.</p></li>
</ol>
<p>Below, we demonstrate these different likelihood calculation methods on simulated data.</p>
<section id="Setup-for-analysis">
<h2>Setup for analysis<a class="headerlink" href="#Setup-for-analysis" title="Permalink to this headline">¶</a></h2>
<p>Import Python modules / packages:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">binarymap</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">plotnine</span> <span class="kn">import</span> <span class="o">*</span>

<span class="kn">import</span> <span class="nn">numpy</span>

<span class="kn">import</span> <span class="nn">dms_variants.bottlenecks</span>
<span class="kn">import</span> <span class="nn">dms_variants.codonvarianttable</span>
<span class="kn">import</span> <span class="nn">dms_variants.globalepistasis</span>
<span class="kn">import</span> <span class="nn">dms_variants.plotnine_themes</span>
<span class="kn">import</span> <span class="nn">dms_variants.simulate</span>
<span class="kn">from</span> <span class="nn">dms_variants.constants</span> <span class="kn">import</span> <span class="n">CBPALETTE</span><span class="p">,</span> <span class="n">CODONS_NOSTOP</span>
</pre></div>
</div>
</div>
<p>Set parameters that define the simulated data. We want to simulate some of the samples with a lot of “noise.” Here that noise takes the form of a very tight bottleneck when transferring the pre-selection library to the selection (such bottlenecks are unavoidable in some experiments). Specifically, we simulate experiments with: - A <em>loose bottleneck</em>: the diversity of the pre-selection library is effectively transferred to the post-selection condition, meaning there is low “noise” from
bottlenecking. - A <em>mid bottleneck</em>: there is a moderate bottleneck in transferring the pre-selection library to the post-selection condition, meaning there is appreciable “noise” from bottlenecking. - A <em>tight bottleneck</em>: there is a very narrow bottleneck when transferring the pre-selection library to the post-selection condition, leading to lots of “noise” from bottlenecking.</p>
<p>The exact simulation parameters are defined below:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">seed</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># random number seed</span>
<span class="n">genelength</span> <span class="o">=</span> <span class="mi">40</span>  <span class="c1"># gene length in codons</span>
<span class="n">libs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;lib_1&quot;</span><span class="p">]</span>  <span class="c1"># distinct libraries of gene</span>
<span class="n">variants_per_lib</span> <span class="o">=</span> <span class="mi">500</span> <span class="o">*</span> <span class="n">genelength</span>  <span class="c1"># variants per library</span>
<span class="n">avgmuts</span> <span class="o">=</span> <span class="mf">2.0</span>  <span class="c1"># average codon mutations per variant</span>
<span class="n">bclen</span> <span class="o">=</span> <span class="mi">16</span>  <span class="c1"># length of nucleotide barcode for each variant</span>
<span class="n">variant_error_rate</span> <span class="o">=</span> <span class="mf">0.005</span>  <span class="c1"># rate at which variant sequence mis-called</span>
<span class="n">avgdepth_per_variant</span> <span class="o">=</span> <span class="mi">100</span>  <span class="c1"># average per-variant sequencing depth</span>
<span class="n">lib_uniformity</span> <span class="o">=</span> <span class="mi">3</span>  <span class="c1"># uniformity of library pre-selection</span>
<span class="n">noise</span> <span class="o">=</span> <span class="mf">0.01</span>  <span class="c1"># random noise in selections</span>
<span class="n">bottlenecks</span> <span class="o">=</span> <span class="p">{</span>  <span class="c1"># bottlenecks from pre- to post-selection</span>
    <span class="s2">&quot;tight_bottle&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">,</span>
    <span class="s2">&quot;mid_bottle&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span>
    <span class="s2">&quot;loose_bottle&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="mi">50</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<p>Seed random number generator for reproducible output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Set pandas display options to show large chunks of Data Frames in this example:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">&quot;display.max_columns&quot;</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">&quot;display.width&quot;</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Hide warnings that clutter output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Set the gray grid <a class="reference external" href="https://plotnine.readthedocs.io/en/stable/generated/plotnine.themes.theme.html?highlight=themes">plotnine theme</a> defined in <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.plotnine_themes.html">dms_variants.plotnine_themes</a> as this gives a nice appearance for the plots:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">theme_set</span><span class="p">(</span><span class="n">dms_variants</span><span class="o">.</span><span class="n">plotnine_themes</span><span class="o">.</span><span class="n">theme_graygrid</span><span class="p">())</span>
</pre></div>
</div>
</div>
</section>
<section id="Simulate-library-of-variants">
<h2>Simulate library of variants<a class="headerlink" href="#Simulate-library-of-variants" title="Permalink to this headline">¶</a></h2>
<p>Simulate a library of codon variants of the gene.</p>
<p>Simulate wildtype gene sequence:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">geneseq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">CODONS_NOSTOP</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">genelength</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wildtype gene of </span><span class="si">{</span><span class="n">genelength</span><span class="si">}</span><span class="s2"> codons:</span><span class="se">\n</span><span class="si">{</span><span class="n">geneseq</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Wildtype gene of 40 codons:
AGATCCGTGATTCTGCGTGCTTACACCAACTCACGGGTGAAACGTGTAATCTTATGCAACAACGACTTACCTATCCGCAACATCCGGCTGATGATGATCCTACACAACTCCGACGCTAGT
</pre></div></div>
</div>
<p>Generate a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a> using <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulate_CodonVariantTable">simulate_CodonVariantTable</a> function:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulate_CodonVariantTable</span><span class="p">(</span>
    <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">,</span>
    <span class="n">bclen</span><span class="o">=</span><span class="n">bclen</span><span class="p">,</span>
    <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span>
        <span class="n">lib</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;avgmuts&quot;</span><span class="p">:</span> <span class="n">avgmuts</span><span class="p">,</span> <span class="s2">&quot;nvariants&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">}</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span>
    <span class="p">},</span>
    <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the number of amino-acid mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_18_0.png" src="_images/narrow_bottleneck_18_0.png" />
</div>
</div>
<p>Average number of codon mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_20_0.png" src="_images/narrow_bottleneck_20_0.png" />
</div>
</div>
<p>Examine how well amino-acid mutations are sampled in the library by looking at the fraction of mutations seen &lt;= some number of times, making separate plots for single mutants and all mutants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">variant_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;single&quot;</span><span class="p">,</span> <span class="s2">&quot;all&quot;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulMutCoverage</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_22_0.png" src="_images/narrow_bottleneck_22_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_22_1.png" src="_images/narrow_bottleneck_22_1.png" />
</div>
</div>
</section>
<section id="Simulate-counts-for-samples">
<h2>Simulate counts for samples<a class="headerlink" href="#Simulate-counts-for-samples" title="Permalink to this headline">¶</a></h2>
<p>Now we simulate the counts of each variant after selection.</p>
<p>First, we define a “phenotype” function using a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator">SigmoidPhenotypeSimulator</a>, which follow the “global epistasis” concepts of <a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al</a> and <a class="reference external" href="https://doi.org/10.1534/genetics.116.195214">Sailer and Harms</a> to define the phenotype in two steps: an underlying latent phenotype that mutations affect additively, and then an
observed phenotype that is a non-linear function of the latent phenotype. The variants are then simulated according to their observed enrichments, which are the exponentials of the observed phenotypes:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">phenosimulator</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">SigmoidPhenotypeSimulator</span><span class="p">(</span><span class="n">geneseq</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the simulated relationship of the latent phenotype with the observed enrichment and phenotype, with a dashed vertical line indicating the wildtype latent phenotype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;enrichment&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype&quot;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotLatentVsObserved</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_26_0.png" src="_images/narrow_bottleneck_26_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_26_1.png" src="_images/narrow_bottleneck_26_1.png" />
</div>
</div>
<p>Plot the latent phenotype, observed phenotype, and observed enrichment of all single amino-acid mutants, with a dashed vertical line indicating the wildtype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observedPhenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observedEnrichment&quot;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotMutsHistogram</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_28_0.png" src="_images/narrow_bottleneck_28_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_28_1.png" src="_images/narrow_bottleneck_28_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_28_2.png" src="_images/narrow_bottleneck_28_2.png" />
</div>
</div>
<p>Now we use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulateSampleCounts">simulateSampleCounts</a> to simulate counts of variants when selection on each variant is proportional to its observed enrichment. We simulate using the several different bottleneck sizes defined above:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">counts</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulateSampleCounts</span><span class="p">(</span>
    <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
    <span class="n">phenotype_func</span><span class="o">=</span><span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedEnrichment</span><span class="p">,</span>
    <span class="n">variant_error_rate</span><span class="o">=</span><span class="n">variant_error_rate</span><span class="p">,</span>
    <span class="n">pre_sample</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;total_count&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="n">avgdepth_per_variant</span><span class="p">,</span>
        <span class="s2">&quot;uniformity&quot;</span><span class="p">:</span> <span class="n">lib_uniformity</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="n">pre_sample_name</span><span class="o">=</span><span class="s2">&quot;pre-selection&quot;</span><span class="p">,</span>
    <span class="n">post_samples</span><span class="o">=</span><span class="p">{</span>
        <span class="n">name</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;noise&quot;</span><span class="p">:</span> <span class="n">noise</span><span class="p">,</span>
            <span class="s2">&quot;total_count&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="n">avgdepth_per_variant</span><span class="p">,</span>
            <span class="s2">&quot;bottleneck&quot;</span><span class="p">:</span> <span class="n">bottle</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">bottle</span> <span class="ow">in</span> <span class="n">bottlenecks</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">},</span>
    <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now add the simulated counts for each library / sample to the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">add_sample_counts_df</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the number of counts for each variant in each sample. The horizontal dashed line shows the total number of variants. The plot shows that all variants are well-sampled in the pre-selection libraries, but that post- selection some variants are sampled more or less. This is expected since selection will decrease and increase the frequency of variants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulVariantCounts</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s2">&quot;v&quot;</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_34_0.png" src="_images/narrow_bottleneck_34_0.png" />
</div>
</div>
<p>Distribution of the number of amino-acid mutations per variant in each sample. As expected, mutations go down after selection:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="n">mut_type</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s2">&quot;v&quot;</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_36_0.png" src="_images/narrow_bottleneck_36_0.png" />
</div>
</div>
<p>Average number of mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s2">&quot;v&quot;</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_38_0.png" src="_images/narrow_bottleneck_38_0.png" />
</div>
</div>
</section>
<section id="Functional-scores-for-variants">
<h2>Functional scores for variants<a class="headerlink" href="#Functional-scores-for-variants" title="Permalink to this headline">¶</a></h2>
<p>Use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.func_scores">CodonVariantTable.func_scores</a> to calculates a functional score for each variant based on its change in frequency from pre- to post-selection:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s2">&quot;pre-selection&quot;</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Also calculate functional scores at the level of amino-acid variants:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_func_scores</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span>
    <span class="s2">&quot;pre-selection&quot;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Classify variants by the “types” of mutations they have using the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.classifyVariants">CodonVariantTable.classifyVariants</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">codonvarianttable</span><span class="o">.</span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span>
    <span class="n">func_scores</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the distributions of scores, coloring by the variant class. As expected, there is more noise with a tighter bottleneck, as can be seen in the spread for wildtype and stop-codon variants. This is expected, as for instance a tight bottleneck will sometimes lead to the random extinction of a wildtype variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">func_scores</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;variant_class&quot;</span><span class="p">,</span> <span class="s2">&quot;func_score&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_violin</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">&quot;variant_class&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s2">&quot;functional score&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~ post_sample&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.75</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">bottlenecks</span><span class="p">),</span> <span class="mi">2</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">libs</span><span class="p">)),</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span> <span class="n">guide</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_46_0.png" src="_images/narrow_bottleneck_46_0.png" />
</div>
</div>
</section>
<section id="Estimate-bottleneck-sizes">
<h2>Estimate bottleneck sizes<a class="headerlink" href="#Estimate-bottleneck-sizes" title="Permalink to this headline">¶</a></h2>
<p>In order to use the bottleneck-based likelihood, we need to know the size of the bottleneck going from the pre- to post-selection conditions. In this simulation, we know the true simulated bottleneck for each sample–but in a real experiment, we may not know the bottleneck. Therefore, we need to estimate it from the data.</p>
<p>We can do that if we examine just the fluctuations in frequency of the wildtype variants. The basic idea is that if there is no bottleneck, then the relative frequencies of the different wildtype variants should remain unchanged–but if there is a bottleneck then some wildtype variants will randomly go up and down in frequency compared to others. The actual calculation is implemented in the
<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.bottlenecks.html#dms_variants.bottlenecks.estimateBottleneck">bottlenecks.estimateBottleneck</a> function.</p>
<p>Below, for each sample we subset on just the wildtype variants. We then estimate the per-variant bottleneck (total bottleneck divided by number of variants), and compare it to the true values. As can be seen below, the estimates are quite accurate for narrow bottlenecks. They become less accurate for the loose bottleneck, because then the bottleneck no longer contributes much of the noise (but in this case, we also no longer really care so much about estimating the bottleneck accurately):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">estimated_bottleneck</span> <span class="o">=</span> <span class="p">{}</span>

<span class="n">estimates</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">scores</span> <span class="ow">in</span> <span class="n">func_scores</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;post_sample&quot;</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">wt_scores</span> <span class="o">=</span> <span class="n">scores</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;n_codon_substitutions == 0&quot;</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">bottlenecks</span><span class="o">.</span><span class="n">estimateBottleneck</span><span class="p">(</span>
        <span class="n">wt_scores</span><span class="p">,</span> <span class="n">n_pre_col</span><span class="o">=</span><span class="s2">&quot;pre_count&quot;</span><span class="p">,</span> <span class="n">n_post_col</span><span class="o">=</span><span class="s2">&quot;post_count&quot;</span>
    <span class="p">)</span>
    <span class="n">estimates</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">n</span><span class="p">,</span> <span class="n">bottlenecks</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">scores</span><span class="p">)))</span>
    <span class="n">estimated_bottleneck</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">n</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">scores</span>
    <span class="p">)</span>  <span class="c1"># estimated total bottleneck (per variant X nvariants)</span>
<span class="n">estimates_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">estimates</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;estimated&quot;</span><span class="p">,</span> <span class="s2">&quot;actual&quot;</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Estimated and actual bottleneck per variant:&quot;</span><span class="p">)</span>
<span class="n">estimates_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Estimated and actual bottleneck per variant:
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>estimated</th>
      <th>actual</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>25.7</td>
      <td>50.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2.8</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.9</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Confirm that the estimates are good when the bottleneck size is much smaller than the average depth per variant:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span>
    <span class="n">estimates_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;actual &lt; 10&quot;</span><span class="p">)[</span><span class="s2">&quot;actual&quot;</span><span class="p">],</span>
    <span class="n">estimates_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;actual &lt; 10&quot;</span><span class="p">)[</span><span class="s2">&quot;estimated&quot;</span><span class="p">],</span>
    <span class="n">rtol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Fit-global-epistasis-models">
<h2>Fit global epistasis models<a class="headerlink" href="#Fit-global-epistasis-models" title="Permalink to this headline">¶</a></h2>
<p>Now we fit global epistasis models using the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.fit_models">fit_models</a> function.</p>
<p>The two types of models are: - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a> models, which model global epistasis in the way described in <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>. - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasis">NoEpistasis</a> models, which assume mutations
contribute additively to the functional scores.</p>
<p>The three ways of calculating the likelihood are: - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.GaussianLikelihood">GaussianLikelihood</a>: assumes deviations from model are normally distributed according to statistical estimates of errors in functional scores. - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.GaussianLikelihood">CauchyLikelihood</a>: a fat-tailed distribution
that is less sensitive to measurements that have large deviations from the model. - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.BottleneckLikelihood">BottleneckLikelihood</a>: explicitly models bottleneck from pre- to post-selection conditions. We use this in conjunction with the bottlenecks estimated above.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fit_df_list</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">scores</span> <span class="ow">in</span> <span class="n">func_scores</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;post_sample&quot;</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>

    <span class="n">bmap</span> <span class="o">=</span> <span class="n">binarymap</span><span class="o">.</span><span class="n">BinaryMap</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">likelihood</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;Gaussian&quot;</span><span class="p">,</span> <span class="s2">&quot;Cauchy&quot;</span><span class="p">,</span> <span class="s2">&quot;Bottleneck&quot;</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">likelihood</span> <span class="o">==</span> <span class="s2">&quot;Bottleneck&quot;</span><span class="p">:</span>
            <span class="n">ifit_df</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">fit_models</span><span class="p">(</span>
                <span class="n">bmap</span><span class="p">,</span> <span class="n">likelihood</span><span class="p">,</span> <span class="n">bottleneck</span><span class="o">=</span><span class="n">estimated_bottleneck</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ifit_df</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">fit_models</span><span class="p">(</span><span class="n">bmap</span><span class="p">,</span> <span class="n">likelihood</span><span class="p">)</span>
        <span class="n">fit_df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ifit_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">,</span> <span class="n">likelihood</span><span class="o">=</span><span class="n">likelihood</span><span class="p">))</span>

<span class="n">fit_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">fit_df_list</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here are the times (in seconds) that it took to fit each model:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="p">(</span>
    <span class="n">fit_df</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;likelihood&quot;</span><span class="p">,</span> <span class="s2">&quot;sample&quot;</span><span class="p">],</span> <span class="n">values</span><span class="o">=</span><span class="s2">&quot;fitting_time&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;description&quot;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>description</th>
      <th>global epistasis</th>
      <th>no epistasis</th>
    </tr>
    <tr>
      <th>likelihood</th>
      <th>sample</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">Bottleneck</th>
      <th>loose_bottle</th>
      <td>22.7</td>
      <td>0.4</td>
    </tr>
    <tr>
      <th>mid_bottle</th>
      <td>12.2</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>tight_bottle</th>
      <td>7.8</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Cauchy</th>
      <th>loose_bottle</th>
      <td>21.1</td>
      <td>1.6</td>
    </tr>
    <tr>
      <th>mid_bottle</th>
      <td>13.5</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>tight_bottle</th>
      <td>10.3</td>
      <td>0.7</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Gaussian</th>
      <th>loose_bottle</th>
      <td>10.6</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>mid_bottle</th>
      <td>8.2</td>
      <td>0.1</td>
    </tr>
    <tr>
      <th>tight_bottle</th>
      <td>6.3</td>
      <td>0.1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we want to see if the no-epistasis or global epistasis model fits the data better in each case. To do this, we compare the <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>. Models with lower AIC are better. We compute the <span class="math notranslate nohighlight">\(\Delta\)</span>AIC between the global epistasis model (<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a>) and no-epistasis model
(<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasis">NoEpistasis</a>): positive <span class="math notranslate nohighlight">\(\Delta\)</span>AIC values indicate that the global epistasis model is better. (Crucially, we can only compare models with likelihoods calculated in the same way on the same data).</p>
<p>Below we see that in all cases, the global epistasis model is better than the no-epistasis model:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">aic_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">fit_df</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;likelihood&quot;</span><span class="p">,</span> <span class="s2">&quot;sample&quot;</span><span class="p">],</span> <span class="n">values</span><span class="o">=</span><span class="s2">&quot;AIC&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;description&quot;</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">deltaAIC</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;no epistasis&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;global epistasis&quot;</span><span class="p">])[[</span><span class="s2">&quot;deltaAIC&quot;</span><span class="p">]]</span>
    <span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="p">)</span>

<span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">aic_df</span><span class="p">[</span><span class="s2">&quot;deltaAIC&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span> <span class="s2">&quot;global epistasis model not better&quot;</span>

<span class="n">aic_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>description</th>
      <th>deltaAIC</th>
    </tr>
    <tr>
      <th>likelihood</th>
      <th>sample</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">Bottleneck</th>
      <th>loose_bottle</th>
      <td>248738.2</td>
    </tr>
    <tr>
      <th>mid_bottle</th>
      <td>16498.1</td>
    </tr>
    <tr>
      <th>tight_bottle</th>
      <td>2690.0</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Cauchy</th>
      <th>loose_bottle</th>
      <td>21072.1</td>
    </tr>
    <tr>
      <th>mid_bottle</th>
      <td>18006.7</td>
    </tr>
    <tr>
      <th>tight_bottle</th>
      <td>16968.0</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Gaussian</th>
      <th>loose_bottle</th>
      <td>36166.7</td>
    </tr>
    <tr>
      <th>mid_bottle</th>
      <td>17981.2</td>
    </tr>
    <tr>
      <th>tight_bottle</th>
      <td>3728.7</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Confirm global epistasis model is always better (<span class="math notranslate nohighlight">\(\Delta\)</span>AIC always positive):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">assert</span> <span class="p">(</span><span class="n">aic_df</span><span class="p">[</span><span class="s2">&quot;deltaAIC&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Because the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a> global epistasis model is always better than the additive <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasis">NoEpistasis</a> model, below we just analyze the results from the global epistasis model.</p>
<p>First, we will examine how the model fits look on all the actual variants used to fit the model. We use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.phenotypes_df">AbstractEpistasis.phenotypes_df</a> to get a data frame of all the variants used to fit each global epistasis model along with their functional scores and the latent and observed phenotypes predicted by each model:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="n">tup</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">phenotypes_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="n">tup</span><span class="o">.</span><span class="n">sample</span><span class="p">,</span> <span class="n">likelihood</span><span class="o">=</span><span class="n">tup</span><span class="o">.</span><span class="n">likelihood</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">fit_df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">tup</span><span class="o">.</span><span class="n">description</span> <span class="o">==</span> <span class="s2">&quot;global epistasis&quot;</span>
    <span class="p">],</span>
    <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Below we plot the relationship between the observed and latent phenotype for each sample and likelihood-calculation method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;latent_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observed_phenotype&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;sample ~ likelihood&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span>
            <span class="mi">2</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;likelihood&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
            <span class="mi">2</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
        <span class="p">),</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_62_0.png" src="_images/narrow_bottleneck_62_0.png" />
</div>
</div>
</section>
<section id="Model-vs-truth-for-amino-acid-variants">
<h2>Model vs truth for amino-acid variants<a class="headerlink" href="#Model-vs-truth-for-amino-acid-variants" title="Permalink to this headline">¶</a></h2>
<p>Because we simulated the variants, we can also use the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator">SigmoidPhenotypeSimulator</a> to get the <strong>true</strong> phenotype of each variant. (Obviously in real non-simulated experiments the true phenotypes are unknown.)</p>
<p>We want to see if the model outperforms the experiments if look at the level of <strong>amino-acid</strong> variants (meaning we pool the counts for all variants with the same amino-acid substitutions together before calculating the functional scores). Recall that above we already got such functional scores into the variable <code class="docutils literal notranslate"><span class="pre">aa_func_scores</span></code>:</p>
<p>First we get a data frame of measured functional scores (in the simulated experiment) calculated at the level of amino-acid variants (pooling all barcoded variants with the same amino-acid mutations):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_variants_df</span> <span class="o">=</span> <span class="n">aa_func_scores</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;post_sample&quot;</span><span class="p">:</span> <span class="s2">&quot;sample&quot;</span><span class="p">})[</span>
    <span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;n_aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;func_score&quot;</span><span class="p">]</span>
<span class="p">]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">measured_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span> <span class="o">**</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;func_score&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>Now we add the model-predicted latent phenotype, observed phenotype, and enrichment for each amino-acid variant using the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.add_phenotypes_to_df">AbstractEpistasis.add_phenotypes_to_df</a> method:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">aa_variants_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">likelihood</span> <span class="ow">in</span> <span class="n">fit_df</span><span class="p">[</span><span class="s2">&quot;likelihood&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">fit_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
            <span class="s1">&#39;(likelihood == @likelihood) &amp; (sample == @sample) &amp; (description == &quot;global epistasis&quot;)&#39;</span>
        <span class="p">)</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="n">model</span><span class="o">.</span><span class="n">add_phenotypes_to_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">predicted_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">model</span><span class="o">.</span><span class="n">enrichments</span><span class="p">(</span>
                    <span class="n">x</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">]</span>
                <span class="p">),</span>
                <span class="n">likelihood</span><span class="o">=</span><span class="n">likelihood</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>
<span class="n">aa_variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_list</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now we add the “true” values from the simulator:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_variants_df</span> <span class="o">=</span> <span class="n">aa_variants_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">true_latent_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="n">phenosimulator</span><span class="o">.</span><span class="n">latentPhenotype</span>
    <span class="p">),</span>
    <span class="n">true_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedEnrichment</span>
    <span class="p">),</span>
    <span class="n">true_observed_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedPhenotype</span>
    <span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now we calculate the correlations between the true enrichments from the simulations and the measured values and the ones predicted by the model. We do this using enrichment rather than observed phenotype as we expect the observed phenotypes to be very noisy at the “low end” (since experiments lose the sensitivity to distinguish among bad and really-bad variants):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">enrichments_corr</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_variants_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;predicted_enrichment&quot;</span><span class="p">:</span> <span class="s2">&quot;model prediction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;measured_enrichment&quot;</span><span class="p">:</span> <span class="s2">&quot;measurement&quot;</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span>
        <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;likelihood&quot;</span><span class="p">,</span> <span class="s2">&quot;true_enrichment&quot;</span><span class="p">],</span>
        <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;model prediction&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement&quot;</span><span class="p">],</span>
        <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;enrichment_type&quot;</span><span class="p">,</span>
        <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;enrichment&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;likelihood&quot;</span><span class="p">,</span> <span class="s2">&quot;enrichment_type&quot;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;enrichment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;true_enrichment&quot;</span><span class="p">]))</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;correlation&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;likelihood&quot;</span><span class="p">],</span> <span class="n">values</span><span class="o">=</span><span class="s2">&quot;correlation&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;enrichment_type&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">enrichments_corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enrichment_type</th>
      <th>measurement</th>
      <th>model prediction</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>likelihood</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">loose_bottle</th>
      <th>Bottleneck</th>
      <td>0.97</td>
      <td>0.97</td>
    </tr>
    <tr>
      <th>Cauchy</th>
      <td>0.97</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>Gaussian</th>
      <td>0.97</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">mid_bottle</th>
      <th>Bottleneck</th>
      <td>0.80</td>
      <td>0.98</td>
    </tr>
    <tr>
      <th>Cauchy</th>
      <td>0.80</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>Gaussian</th>
      <td>0.80</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">tight_bottle</th>
      <th>Bottleneck</th>
      <td>0.58</td>
      <td>0.96</td>
    </tr>
    <tr>
      <th>Cauchy</th>
      <td>0.58</td>
      <td>0.90</td>
    </tr>
    <tr>
      <th>Gaussian</th>
      <td>0.58</td>
      <td>0.76</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Above we see that the model predictions are always better than the actual measurements, and furthermore that when there is a very tight experimental bottleneck (lots of “noise”) the Bottleneck likelihood (and to a lesser degree the Cauchy likelihood) outperforms the Gaussian likelihood.</p>
<p>Below we plot the correlations between the model predictions and the true enrichments for each likelihood calculation method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">aa_variants_df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;true_enrichment&quot;</span><span class="p">,</span> <span class="s2">&quot;predicted_enrichment&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;sample ~ likelihood&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span>
            <span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;likelihood&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
            <span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
        <span class="p">)</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_72_0.png" src="_images/narrow_bottleneck_72_0.png" />
</div>
</div>
</section>
<section id="Model-vs-truth-for-single-mutants">
<h2>Model vs truth for single mutants<a class="headerlink" href="#Model-vs-truth-for-single-mutants" title="Permalink to this headline">¶</a></h2>
<p>Now we repeat the above analysis for just <strong>single</strong> amino-acid mutants.</p>
<p>First, get the single amino-acid mutant variants:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">single_aa_variants_df</span> <span class="o">=</span> <span class="n">aa_variants_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;n_aa_substitutions == 1&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span>
    <span class="n">drop</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Tabulate how the measurements and model predictions compare to the true enrichments:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">enrichments_corr</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">single_aa_variants_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;predicted_enrichment&quot;</span><span class="p">:</span> <span class="s2">&quot;model prediction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;measured_enrichment&quot;</span><span class="p">:</span> <span class="s2">&quot;measurement&quot;</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span>
        <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;likelihood&quot;</span><span class="p">,</span> <span class="s2">&quot;true_enrichment&quot;</span><span class="p">],</span>
        <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;model prediction&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement&quot;</span><span class="p">],</span>
        <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;enrichment_type&quot;</span><span class="p">,</span>
        <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;enrichment&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;likelihood&quot;</span><span class="p">,</span> <span class="s2">&quot;enrichment_type&quot;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;enrichment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;true_enrichment&quot;</span><span class="p">]))</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;correlation&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;likelihood&quot;</span><span class="p">],</span> <span class="n">values</span><span class="o">=</span><span class="s2">&quot;correlation&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;enrichment_type&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">enrichments_corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enrichment_type</th>
      <th>measurement</th>
      <th>model prediction</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>likelihood</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">loose_bottle</th>
      <th>Bottleneck</th>
      <td>0.99</td>
      <td>0.98</td>
    </tr>
    <tr>
      <th>Cauchy</th>
      <td>0.99</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>Gaussian</th>
      <td>0.99</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">mid_bottle</th>
      <th>Bottleneck</th>
      <td>0.91</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>Cauchy</th>
      <td>0.91</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>Gaussian</th>
      <td>0.91</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">tight_bottle</th>
      <th>Bottleneck</th>
      <td>0.78</td>
      <td>0.98</td>
    </tr>
    <tr>
      <th>Cauchy</th>
      <td>0.78</td>
      <td>0.92</td>
    </tr>
    <tr>
      <th>Gaussian</th>
      <td>0.78</td>
      <td>0.77</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Once again, the model predictions outperform the measurements, and the Bottleneck likelihood performs best for the tight bottleneck.</p>
<p>Plot the predictions for each likelihood-calculation method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">single_aa_variants_df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;true_enrichment&quot;</span><span class="p">,</span> <span class="s2">&quot;predicted_enrichment&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;sample ~ likelihood&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span>
            <span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;likelihood&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
            <span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
        <span class="p">)</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_78_0.png" src="_images/narrow_bottleneck_78_0.png" />
</div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data.html">Analyze simulated library of codon variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data_w_gaps.html">Variants with deletion mutations</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data_multi_targets.html">Libraries with secondary target genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="prob_escape.html">Compute probabilities of escape</a></li>
<li class="toctree-l2"><a class="reference internal" href="parsebarcodes_sim_data.html">Parse barcodes</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Fitting gobal epistasis models with experimental “noise”</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_plot_formatting.html">Formatting <code class="docutils literal notranslate"><span class="pre">CodonVariantTable</span></code> plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dms_variants.html">dms_variants package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="examples.html">Examples</a><ul>
      <li>Previous: <a href="parsebarcodes_sim_data.html" title="previous chapter">Parse barcodes</a></li>
      <li>Next: <a href="codonvariant_plot_formatting.html" title="next chapter">Formatting <code class="docutils literal notranslate"><span class="pre">CodonVariantTable</span></code> plots</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2022.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/narrow_bottleneck.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>