<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Parse barcodes &#8212; dms_variants 1.5.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=420b4293" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css?v=7f9a90b1" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css" />
    <script src="_static/documentation_options.js?v=e0a75244"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Fitting gobal epistasis models with experimental “noise”" href="narrow_bottleneck.html" />
    <link rel="prev" title="Compute probabilities of escape" href="prob_escape.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="Parse-barcodes">
<h1>Parse barcodes<a class="headerlink" href="#Parse-barcodes" title="Link to this heading">¶</a></h1>
<p>Simulate some Illumina reads giving barcode counts of variants, parse them from the simulated FASTQ files, and add them to a variant table. Primarily designed to illustrate how to use an <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.illuminabarcodeparser.html#dms_variants.illuminabarcodeparser.IlluminaBarcodeParser">IlluminaBarcodeParser</a> to parse barcode counts and add them to a
<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a>.</p>
<section id="Setup-for-analysis">
<h2>Setup for analysis<a class="headerlink" href="#Setup-for-analysis" title="Link to this heading">¶</a></h2>
<p>Import Python modules / packages. We use <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a> for ggplot2-like plotting:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">plotnine</span> <span class="kn">import</span> <span class="o">*</span>

<span class="kn">from</span> <span class="nn">dms_variants.codonvarianttable</span> <span class="kn">import</span> <span class="n">CodonVariantTable</span>
<span class="kn">from</span> <span class="nn">dms_variants.constants</span> <span class="kn">import</span> <span class="n">CBPALETTE</span>
<span class="kn">import</span> <span class="nn">dms_variants.illuminabarcodeparser</span>
<span class="kn">import</span> <span class="nn">dms_variants.plotnine_themes</span>
<span class="kn">from</span> <span class="nn">dms_variants.simulate</span> <span class="kn">import</span> <span class="n">mutate_seq</span><span class="p">,</span> <span class="n">rand_seq</span><span class="p">,</span> <span class="n">simulate_CodonVariantTable</span>
<span class="kn">from</span> <span class="nn">dms_variants.utils</span> <span class="kn">import</span> <span class="n">reverse_complement</span>
</pre></div>
</div>
</div>
<p>Seed random number generator:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Hide warnings that clutter output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Set the gray grid <a class="reference external" href="https://plotnine.readthedocs.io/en/stable/generated/plotnine.themes.theme.html?highlight=themes">plotnine theme</a> defined in <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.plotnine_themes.html">dms_variants.plotnine_themes</a> as this gives a nice appearance for the plots:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">theme_set</span><span class="p">(</span><span class="n">dms_variants</span><span class="o">.</span><span class="n">plotnine_themes</span><span class="o">.</span><span class="n">theme_graygrid</span><span class="p">())</span>
</pre></div>
</div>
</div>
</section>
<section id="Create-small-variant-table">
<h2>Create small variant table<a class="headerlink" href="#Create-small-variant-table" title="Link to this heading">¶</a></h2>
<p>We simulate a small <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a> for one libraries with three variants, with short 8-nucleotide barcodes. We simulate the table using using <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulate_CodonVariantTable">simulate_CodonVariantTable</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bclen</span> <span class="o">=</span> <span class="mi">8</span>

<span class="n">variants</span> <span class="o">=</span> <span class="n">simulate_CodonVariantTable</span><span class="p">(</span>
    <span class="n">geneseq</span><span class="o">=</span><span class="n">rand_seq</span><span class="p">(</span><span class="mi">30</span><span class="p">),</span>
    <span class="n">bclen</span><span class="o">=</span><span class="n">bclen</span><span class="p">,</span>
    <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;lib_1&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;avgmuts&quot;</span><span class="p">:</span> <span class="mf">1.5</span><span class="p">,</span> <span class="s2">&quot;nvariants&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">}},</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here are the variants in the simulated table:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>variant_call_support</th>
      <th>codon_substitutions</th>
      <th>aa_substitutions</th>
      <th>n_codon_substitutions</th>
      <th>n_aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>ATTCCCGT</td>
      <td>1</td>
      <td>CAG1CTA AGA7GCT</td>
      <td>Q1L R7A</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>TACGATTA</td>
      <td>1</td>
      <td>TCT9AAA</td>
      <td>S9K</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>TCACAACC</td>
      <td>1</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Simulate-barcode-reads">
<h2>Simulate barcode reads<a class="headerlink" href="#Simulate-barcode-reads" title="Link to this heading">¶</a></h2>
<p>We are now going to simulate barcode reads that we will then parse with the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.illuminabarcodeparser.html#dms_variants.illuminabarcodeparser.IlluminaBarcodeParser">IlluminaBarcodeParser</a>. This parser is designed to parse reads that have this structure:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>5&#39;-[R2_start]-upstream-barcode-downstream-[R1_start]-3&#39;
</pre></div>
</div>
<p>where the R2 read is optional. Note that we use the convention that the variant table defines the <em>barcode</em> sequence in the <strong>orientation read by R1</strong>, which is the reverse complement of the orientation shown in the schematic immediately above.</p>
<p>Here we will simulate a common setup where there is no R2 read, and the <em>downstream</em> sequence between the barcode and the R1 start is non-existent (R1 is flush against the barcode). That means we still have to define an upstream sequence (we’ll make this 10 nucleotides long). We also get the reverse complement of this upstream sequence, which is what is found in the R1 reads:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">upstream</span> <span class="o">=</span> <span class="n">rand_seq</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="n">upstream_rc</span> <span class="o">=</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">upstream</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`upstream` sequence is </span><span class="si">{</span><span class="n">upstream</span><span class="si">}</span><span class="s2">; reverse complement is </span><span class="si">{</span><span class="n">upstream_rc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
`upstream` sequence is CCTGATACGA; reverse complement is TCGTATCAGG
</pre></div></div>
</div>
<p>We also simulate how many counts of each barcode we’ll have:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">simulated_barcode_counts</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span>
<span class="p">)</span>

<span class="n">simulated_barcode_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>barcode</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>ATTCCCGT</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>TACGATTA</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>TCACAACC</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we simulate the actual Illumina barcode reads that give these counts. When we parse the barcodes, we’ll allow up to one mismatch in the upstream sequence, so sometimes add such a mismatch. Also, since these reads are in the orientation read by R1, we generate them by adding the barcodes (which we are already representing in the R1 orientation) with the reverse-complement of the <em>upstream</em> sequence (since R1 reads into the upstream sequence backwards):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">barcode_seqs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">simulated_barcode_counts</span><span class="o">.</span><span class="n">itertuples</span><span class="p">():</span>
    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">tup</span><span class="o">.</span><span class="n">count</span><span class="p">):</span>
        <span class="n">barcode_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tup</span><span class="o">.</span><span class="n">barcode</span> <span class="o">+</span> <span class="n">mutate_seq</span><span class="p">(</span><span class="n">upstream_rc</span><span class="p">,</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<p>We will also now add a few sequences that are <strong>not</strong> do not parse to valid barcodes. First, add a few where the barcode sequence is invalid (not among the true barcodes in the variant table):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_invalid</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">valid_barcodes</span> <span class="o">=</span> <span class="n">simulated_barcode_counts</span><span class="p">[</span><span class="s2">&quot;barcode&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_invalid</span><span class="p">):</span>
    <span class="n">barcode_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rand_seq</span><span class="p">(</span><span class="n">bclen</span><span class="p">,</span> <span class="n">exclude</span><span class="o">=</span><span class="n">valid_barcodes</span><span class="p">)</span> <span class="o">+</span> <span class="n">upstream_rc</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Next add a few that are unparseable because the <em>upstream</em> sequence has too many mutations to be recognized:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_unparseable</span> <span class="o">=</span> <span class="mi">2</span>
<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_unparseable</span><span class="p">):</span>
    <span class="n">barcode_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">valid_barcodes</span><span class="p">)</span> <span class="o">+</span> <span class="n">mutate_seq</span><span class="p">(</span><span class="n">upstream_rc</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
</pre></div>
</div>
</div>
<p>Finally, we will specify a few low-quality sequences. First, define the minimum quality score in the barcode region that we will consider acceptable:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">minq</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">minq_char</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="n">minq</span> <span class="o">+</span> <span class="mi">33</span><span class="p">)</span>  <span class="c1"># corresponding character for Q score</span>
</pre></div>
</div>
</div>
<p>Add a sufficient quality Q-score string to all the reads in <code class="docutils literal notranslate"><span class="pre">barcode_seqs</span></code>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">barcode_seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">seq</span><span class="p">,</span> <span class="n">minq_char</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">barcode_seqs</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p>Finally, we add a few sequences with excessively low-quality barcode reads by taking the first few valid entries in <code class="docutils literal notranslate"><span class="pre">barcode_seqs</span></code> and adding a copy with a lower Q-score:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lowq_char</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="n">minq</span> <span class="o">+</span> <span class="mi">33</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># low-quality Q-score</span>
<span class="n">n_low_quality</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="n">q</span> <span class="ow">in</span> <span class="n">barcode_seqs</span><span class="p">[:</span><span class="n">n_low_quality</span><span class="p">]:</span>
    <span class="n">barcode_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">seq</span><span class="p">,</span> <span class="n">lowq_char</span> <span class="o">+</span> <span class="n">q</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>
</pre></div>
</div>
</div>
<p>Finally, we will generate the contents of a FASTQ file holding these sequences. In order to do that, we need to specify the quality scores for the sequences</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fastq_text</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
    <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;@read</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="se">\n</span><span class="s2">+</span><span class="se">\n</span><span class="si">{</span><span class="n">q</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">barcode_seqs</span><span class="p">)]</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Here is the simulated FASTQ file:</span><span class="se">\n</span><span class="s2">-------------</span><span class="se">\n</span><span class="si">{</span><span class="n">fastq_text</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Here is the simulated FASTQ file:
-------------
@read1
ATTCCCGTTCGTATCATG
+
555555555555555555
@read2
ATTCCCGTTCGTATCAGG
+
555555555555555555
@read3
TACGATTATCGTTTCAGG
+
555555555555555555
@read4
TCACAACCTCGTATCACG
+
555555555555555555
@read5
TCACAACCTCGTATCAGG
+
555555555555555555
@read6
TCTTCGGATCGTATCAGG
+
555555555555555555
@read7
TACTGTATTCGTATCAGG
+
555555555555555555
@read8
AGTCCCACTCGTATCAGG
+
555555555555555555
@read9
TCACAACCTCGGATCACG
+
555555555555555555
@read10
TACGATTATCGTACCAGC
+
555555555555555555
@read11
ATTCCCGTTCGTATCATG
+
455555555555555555
</pre></div></div>
</div>
</section>
<section id="Parse-barcode-reads">
<h2>Parse barcode reads<a class="headerlink" href="#Parse-barcode-reads" title="Link to this heading">¶</a></h2>
<p>Now we initialize a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.illuminabarcodeparser.html#dms_variants.illuminabarcodeparser.IlluminaBarcodeParser">IlluminaBarcodeParser</a> to parse the barcodes, allowing one mismatch in the upstream sequence (note if we had a <em>downstream</em> sequence we could specify it too). We specify the set of valid barcodes, which we get using the
<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.illuminabarcodeparser.html#dms_variants.illuminabarcodeparser.IlluminaBarcodeParser.valid_barcodes">CodonVariantTable.valid_barcodes</a> method of our variant table. Note that if we had multiple libraries, we’d need a different parser for each one as each would have it’s own set of valid barcodes:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">parser</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">illuminabarcodeparser</span><span class="o">.</span><span class="n">IlluminaBarcodeParser</span><span class="p">(</span>
    <span class="n">bclen</span><span class="o">=</span><span class="n">bclen</span><span class="p">,</span>
    <span class="n">valid_barcodes</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">(</span><span class="s2">&quot;lib_1&quot;</span><span class="p">),</span>
    <span class="n">upstream</span><span class="o">=</span><span class="n">upstream</span><span class="p">,</span>
    <span class="n">upstream_mismatch</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">minq</span><span class="o">=</span><span class="n">minq</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now parse the barcodes. First, write the FASTQ to a temporary file, then parse it. We add the name of the library and the sample to the parsed data frames. Then we parse the barcodes using <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.illuminabarcodeparser.html#dms_variants.illuminabarcodeparser.IlluminaBarcodeParser.parse">IlluminaBarcodeParser.parse</a>, which returns a 2-tuple of data frames giving the counts of valid barcodes and the fates of all reads that it attempted to parse:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="s2">&quot;r+&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fastq</span><span class="p">:</span>
    <span class="n">fastq</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fastq_text</span><span class="p">)</span>
    <span class="n">fastq</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
    <span class="n">counts</span><span class="p">,</span> <span class="n">fates</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span>
        <span class="n">r1files</span><span class="o">=</span><span class="n">fastq</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
        <span class="n">add_cols</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;library&quot;</span><span class="p">:</span> <span class="s2">&quot;lib_1&quot;</span><span class="p">,</span> <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="s2">&quot;sample_1&quot;</span><span class="p">},</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here are the parsed barcode counts:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>barcode</th>
      <th>count</th>
      <th>library</th>
      <th>sample</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ATTCCCGT</td>
      <td>2</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TCACAACC</td>
      <td>2</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TACGATTA</td>
      <td>1</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Here are the fates of the reads we attempted to parse:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fates</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>fate</th>
      <th>count</th>
      <th>library</th>
      <th>sample</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>valid barcode</td>
      <td>5</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>invalid barcode</td>
      <td>3</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>unparseable barcode</td>
      <td>2</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>low quality barcode</td>
      <td>1</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>failed chastity filter</td>
      <td>0</td>
      <td>lib_1</td>
      <td>sample_1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Add-barcodes-to-variant-table">
<h2>Add barcodes to variant table<a class="headerlink" href="#Add-barcodes-to-variant-table" title="Link to this heading">¶</a></h2>
<p>We add the barcodes <code class="docutils literal notranslate"><span class="pre">count</span></code> data frame to the variant table using <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df">CodonVariantTable.add_sample_counts_df</a>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants</span><span class="o">.</span><span class="n">add_sample_counts_df</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Here are the counts added to the variant table:&quot;</span><span class="p">)</span>
<span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Here are the counts added to the variant table:
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>barcode</th>
      <th>count</th>
      <th>variant_call_support</th>
      <th>codon_substitutions</th>
      <th>aa_substitutions</th>
      <th>n_codon_substitutions</th>
      <th>n_aa_substitutions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>sample_1</td>
      <td>ATTCCCGT</td>
      <td>2</td>
      <td>1</td>
      <td>CAG1CTA AGA7GCT</td>
      <td>Q1L R7A</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>sample_1</td>
      <td>TCACAACC</td>
      <td>2</td>
      <td>1</td>
      <td></td>
      <td></td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>sample_1</td>
      <td>TACGATTA</td>
      <td>1</td>
      <td>1</td>
      <td>TCT9AAA</td>
      <td>S9K</td>
      <td>1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Make sure these counts are as expected from the simulation parameters:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">testing</span><span class="o">.</span><span class="n">assert_frame_equal</span><span class="p">(</span>
    <span class="n">simulated_barcode_counts</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;barcode&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
    <span class="p">(</span>
        <span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span><span class="p">[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">]]</span>
        <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;barcode&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Look-at-the-barcode-read-fates">
<h2>Look at the barcode-read fates<a class="headerlink" href="#Look-at-the-barcode-read-fates" title="Link to this heading">¶</a></h2>
<p>Now we look in more detail at the barcode-read <code class="docutils literal notranslate"><span class="pre">fates</span></code> data frame that tells what happened to the barcode reads. First, we make sure the correct number of reads fall in each category based on the simulation parameters:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">testing</span><span class="o">.</span><span class="n">assert_frame_equal</span><span class="p">(</span>
    <span class="n">fates</span><span class="p">[[</span><span class="s2">&quot;fate&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">]],</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot;valid barcode&quot;</span><span class="p">,</span> <span class="n">simulated_barcode_counts</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
            <span class="p">(</span><span class="s2">&quot;invalid barcode&quot;</span><span class="p">,</span> <span class="n">n_invalid</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;unparseable barcode&quot;</span><span class="p">,</span> <span class="n">n_unparseable</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;low quality barcode&quot;</span><span class="p">,</span> <span class="n">n_low_quality</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;failed chastity filter&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
        <span class="p">],</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;fate&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">],</span>
    <span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now plot the barcode-read fates using <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a> for ggplot2-like plotting. The plot below shows the valid barcodes in orange and all others in gray:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">barcode_fate_plot</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">fates</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">sample</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                <span class="n">x</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">),</span>
            <span class="n">fate</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;fate&quot;</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;fate&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
            <span class="n">is_valid</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;fate&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;valid barcode&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;fate&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s2">&quot;is_valid&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;sample ~ library&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;sample ~ library&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">CBPALETTE</span><span class="p">,</span> <span class="n">guide</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span>
            <span class="mf">1.4</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">fates</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())),</span>
            <span class="mf">1.4</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.2</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">fates</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())),</span>
        <span class="p">),</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_y_continuous</span><span class="p">(</span>
        <span class="n">labels</span><span class="o">=</span><span class="n">dms_variants</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">latex_sci_not</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;number of reads&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">barcode_fate_plot</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/parsebarcodes_sim_data_47_0.png" src="_images/parsebarcodes_sim_data_47_0.png" />
</div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo" />
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data.html">Analyze simulated library of codon variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data_w_gaps.html">Variants with deletion mutations</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data_multi_targets.html">Libraries with secondary target genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="prob_escape.html">Compute probabilities of escape</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Parse barcodes</a></li>
<li class="toctree-l2"><a class="reference internal" href="narrow_bottleneck.html">Fitting gobal epistasis models with experimental “noise”</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_plot_formatting.html">Formatting <code class="docutils literal notranslate"><span class="pre">CodonVariantTable</span></code> plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dms_variants.html">dms_variants package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="examples.html">Examples</a><ul>
      <li>Previous: <a href="prob_escape.html" title="previous chapter">Compute probabilities of escape</a></li>
      <li>Next: <a href="narrow_bottleneck.html" title="next chapter">Fitting gobal epistasis models with experimental “noise”</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2019--2024.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/parsebarcodes_sim_data.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://github.blog/wp-content/uploads/2008/12/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>