<!DOCTYPE html>

<html lang="en" data-content_root="../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dms_variants.simulate &#8212; dms_variants 1.5.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css?v=420b4293" />
    <link rel="stylesheet" type="text/css" href="../../_static/plot_directive.css?v=7f9a90b1" />
    <script src="../../_static/documentation_options.js?v=e0a75244"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for dms_variants.simulate</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">========</span>
<span class="sd">simulate</span>
<span class="sd">========</span>

<span class="sd">Simulate data.</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">tempfile</span>

<span class="kn">import</span> <span class="nn">numpy</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">plotnine</span> <span class="k">as</span> <span class="nn">p9</span>

<span class="kn">import</span> <span class="nn">dms_variants.codonvarianttable</span>
<span class="kn">from</span> <span class="nn">dms_variants.constants</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">AAS_WITHSTOP</span><span class="p">,</span>
    <span class="n">AAS_WITHSTOP_WITHGAP</span><span class="p">,</span>
    <span class="n">CBPALETTE</span><span class="p">,</span>
    <span class="n">CODONS</span><span class="p">,</span>
    <span class="n">CODONS_WITHGAP</span><span class="p">,</span>
    <span class="n">CODON_TO_AA</span><span class="p">,</span>
    <span class="n">CODON_TO_AA_WITHGAP</span><span class="p">,</span>
    <span class="n">NTS</span><span class="p">,</span>
<span class="p">)</span>


<div class="viewcode-block" id="rand_seq">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.rand_seq">[docs]</a>
<span class="k">def</span> <span class="nf">rand_seq</span><span class="p">(</span><span class="n">seqlen</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">exclude</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nseqs</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Random nucleotide sequence(s).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seqlen : int</span>
<span class="sd">        Length of sequence(s).</span>
<span class="sd">    exclude : None or list</span>
<span class="sd">        Do not generate any of the sequences in this list.</span>
<span class="sd">    nseqs : int</span>
<span class="sd">        If &gt;1, return a list of this many sequences.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str or list</span>
<span class="sd">        A random sequence or list of them depending on value of `nseqs`.</span>
<span class="sd">        If multiple sequences, they are all unique.</span>

<span class="sd">    Seed `random.seed` for reproducible output.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; random.seed(1)</span>
<span class="sd">    &gt;&gt;&gt; rand_seq(4)</span>
<span class="sd">    &#39;CAGA&#39;</span>

<span class="sd">    &gt;&gt;&gt; random.seed(1)</span>
<span class="sd">    &gt;&gt;&gt; rand_seq(5, nseqs=3)</span>
<span class="sd">    [&#39;CAGAT&#39;, &#39;TTTCA&#39;, &#39;TATTA&#39;]</span>

<span class="sd">    &gt;&gt;&gt; random.seed(1)</span>
<span class="sd">    &gt;&gt;&gt; rand_seq(5, nseqs=3, exclude=[&#39;TTTCA&#39;])</span>
<span class="sd">    [&#39;CAGAT&#39;, &#39;TATTA&#39;, &#39;TGCAG&#39;]</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">exclude</span><span class="p">:</span>
        <span class="n">exclude</span> <span class="o">=</span> <span class="p">{</span><span class="n">s</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">exclude</span><span class="p">}</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">exclude</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">NTS</span><span class="p">)</span> <span class="o">**</span> <span class="n">seqlen</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;excluding too many sequences, none valid.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">exclude</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">seqs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">nseqs</span><span class="p">:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">NTS</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">seqlen</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">seq</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">exclude</span><span class="p">:</span>
            <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">exclude</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">seqs</span></div>



<div class="viewcode-block" id="mutate_seq">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.mutate_seq">[docs]</a>
<span class="k">def</span> <span class="nf">mutate_seq</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">nmuts</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Mutate a nucleotide sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq : str</span>
<span class="sd">        Sequence to mutate.</span>
<span class="sd">    nmuts: int</span>
<span class="sd">        Number of mutations to make to sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Mutagenized sequence, all upper-case.</span>

<span class="sd">    Seed `random.seed` for reproducible output.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; random.seed(1)</span>
<span class="sd">    &gt;&gt;&gt; mutate_seq(&#39;ATGCAA&#39;, 2)</span>
<span class="sd">    &#39;AAGCGA&#39;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">nmuts</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Can not make </span><span class="si">{</span><span class="n">nmuts</span><span class="si">}</span><span class="s2"> mutations to sequence of &quot;</span> <span class="sa">f</span><span class="s2">&quot;length </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
    <span class="n">sites</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)),</span> <span class="n">nmuts</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">:</span>
        <span class="n">seq</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="n">nt</span> <span class="k">for</span> <span class="n">nt</span> <span class="ow">in</span> <span class="n">NTS</span> <span class="k">if</span> <span class="n">nt</span> <span class="o">!=</span> <span class="n">seq</span><span class="p">[</span><span class="n">r</span><span class="p">]])</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span></div>



<div class="viewcode-block" id="codon_muts">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.codon_muts">[docs]</a>
<span class="k">def</span> <span class="nf">codon_muts</span><span class="p">(</span><span class="n">codonseq</span><span class="p">,</span> <span class="n">nmuts</span><span class="p">,</span> <span class="n">nvariants</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get random codon mutations to sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    codonseq : str</span>
<span class="sd">        A valid codon sequence</span>
<span class="sd">    nmuts : int</span>
<span class="sd">        Number of mutations to make per variant.</span>
<span class="sd">    nvariants : int</span>
<span class="sd">        Number of variants to generate. There is no guarantee that the</span>
<span class="sd">        variants are unique.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list</span>
<span class="sd">        List of mutations in each of the `nvariants` mutated variants</span>
<span class="sd">        in the form &#39;ATG1TAC GGC3GGG&#39; where the number refers to the</span>
<span class="sd">        **codon** position.</span>

<span class="sd">    Seed `random.seed` for reproducible output.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; codonseq = &#39;ATGGAAGGGCATGGA&#39;</span>
<span class="sd">    &gt;&gt;&gt; random.seed(1)</span>
<span class="sd">    &gt;&gt;&gt; codon_muts(codonseq, nmuts=2, nvariants=3)</span>
<span class="sd">    [&#39;ATG1CAC GAA2ACT&#39;, &#39;CAT4CTT GGA5GGG&#39;, &#39;GAA2ACG CAT4GAA&#39;]</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;(?:</span><span class="si">{</span><span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">CODONS</span><span class="p">)</span><span class="si">}</span><span class="s2">)+&quot;</span><span class="p">,</span> <span class="n">codonseq</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;`codonseq` not valid codon sequence:</span><span class="se">\n</span><span class="si">{</span><span class="n">codonseq</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">codonseq</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">==</span> <span class="mi">0</span>
    <span class="n">codons</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span> <span class="n">codonseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="n">i</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">codonseq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span><span class="p">)}</span>
    <span class="k">if</span> <span class="n">nmuts</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">codons</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`nmuts` greater than number of codons in `codonseq`&quot;</span><span class="p">)</span>
    <span class="n">mutlist</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvariants</span><span class="p">):</span>
        <span class="n">sitemuts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">site</span><span class="p">,</span> <span class="n">wt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">codons</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span> <span class="n">nmuts</span><span class="p">)):</span>
            <span class="n">mut</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">CODONS</span> <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">wt</span><span class="p">])</span>
            <span class="n">sitemuts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wt</span><span class="si">}{</span><span class="n">site</span><span class="si">}{</span><span class="n">mut</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">mutlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sitemuts</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">mutlist</span></div>



<div class="viewcode-block" id="simulate_CodonVariantTable">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.simulate_CodonVariantTable">[docs]</a>
<span class="k">def</span> <span class="nf">simulate_CodonVariantTable</span><span class="p">(</span>
    <span class="o">*</span><span class="p">,</span>
    <span class="n">geneseq</span><span class="p">,</span>
    <span class="n">bclen</span><span class="p">,</span>
    <span class="n">library_specs</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">variant_call_support</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
    <span class="n">allowed_aa_muts</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">allowgaps</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate :class:`dms_variants.codonvarianttable.CodonVariantTable`.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">    Only simulates the variants, not counts for samples. To simulate counts,</span>
<span class="sd">    use :func:`simulateSampleCounts`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geneseq : str</span>
<span class="sd">        Sequence of wildtype protein-coding gene.</span>
<span class="sd">    bclen : int</span>
<span class="sd">        Length of the barcodes; must enable complexity at least</span>
<span class="sd">        10-fold greater than max number of variants.</span>
<span class="sd">    library_specs : dict</span>
<span class="sd">        Specifications for each simulated library. Keys are &#39;avgmuts&#39;</span>
<span class="sd">        and &#39;nvariants&#39;. Mutations per variant are Poisson distributed.</span>
<span class="sd">    seed : int or None</span>
<span class="sd">        Random number seed or `None` to set no seed.</span>
<span class="sd">    variant_call_support : int &gt; 0</span>
<span class="sd">        Support is floor of 1 plus draw from exponential distribution</span>
<span class="sd">        with this mean.</span>
<span class="sd">    allowed_aa_muts : None or array-like</span>
<span class="sd">        If we only allow some amino-acid substitutions, provide them.</span>
<span class="sd">        Specify mutations like &#39;E1E&#39; if you want to allow synonymous</span>
<span class="sd">        codon mutations.</span>
<span class="sd">    allowgaps : bool</span>
<span class="sd">        Allow gaps as a codon character.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    :class:`dms_variants.codonvarianttable.CodonVariantTable`</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>

<span class="sd">    &gt;&gt;&gt; geneseq = &#39;ATGGGCAGC&#39;</span>
<span class="sd">    &gt;&gt;&gt; bclen = 4</span>
<span class="sd">    &gt;&gt;&gt; library_specs = {&#39;lib1&#39;: {&#39;avgmuts&#39;: 3, &#39;nvariants&#39;: 5}}</span>
<span class="sd">    &gt;&gt;&gt; variants =  simulate_CodonVariantTable(geneseq=geneseq,</span>
<span class="sd">    ...                                        bclen=bclen,</span>
<span class="sd">    ...                                        library_specs=library_specs)</span>
<span class="sd">    &gt;&gt;&gt; variants.barcode_variant_df[[&#39;barcode&#39;, &#39;aa_substitutions&#39;]]</span>
<span class="sd">      barcode aa_substitutions</span>
<span class="sd">    0    AGTC      M1L G2Y S3F</span>
<span class="sd">    1    ATTC              S3R</span>
<span class="sd">    2    CAAA      M1C G2S S3A</span>
<span class="sd">    3    CGAT              M1N</span>
<span class="sd">    4    TAAT              M1V</span>

<span class="sd">    &gt;&gt;&gt; allowed_aa_muts = [&#39;M1M&#39;, &#39;M1L&#39;, &#39;M1A&#39;, &#39;G2A&#39;, &#39;G2K&#39;, &#39;S3C&#39;, &#39;S3S&#39;]</span>
<span class="sd">    &gt;&gt;&gt; variants =  simulate_CodonVariantTable(geneseq=geneseq,</span>
<span class="sd">    ...                                        bclen=bclen,</span>
<span class="sd">    ...                                        library_specs=library_specs,</span>
<span class="sd">    ...                                        allowed_aa_muts=allowed_aa_muts,</span>
<span class="sd">    ...                                        seed=2)</span>
<span class="sd">    &gt;&gt;&gt; variants.barcode_variant_df[[&#39;barcode&#39;, &#39;aa_substitutions&#39;]]</span>
<span class="sd">      barcode aa_substitutions</span>
<span class="sd">    0    AAAA          M1L G2A</span>
<span class="sd">    1    CAAC          M1A G2A</span>
<span class="sd">    2    GTTG              M1A</span>
<span class="sd">    3    TTAA              G2A</span>
<span class="sd">    4    TTCA          M1L G2A</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
        <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">library_specs</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;empty `library_specs`&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">variant_call_support</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`variant_call_support` must be int &gt; 0&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;length of `geneseq` not multiple of 3&quot;</span><span class="p">)</span>
    <span class="n">genelength</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span>

    <span class="k">if</span> <span class="n">allowgaps</span><span class="p">:</span>
        <span class="n">aas</span> <span class="o">=</span> <span class="n">AAS_WITHSTOP_WITHGAP</span>
        <span class="n">codons</span> <span class="o">=</span> <span class="n">CODONS_WITHGAP</span>
        <span class="n">codon_to_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA_WITHGAP</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">aas</span> <span class="o">=</span> <span class="n">AAS_WITHSTOP</span>
        <span class="n">codons</span> <span class="o">=</span> <span class="n">CODONS</span>
        <span class="n">codon_to_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span>

    <span class="k">if</span> <span class="n">allowed_aa_muts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">allowed_aa_muts_at_site</span> <span class="o">=</span> <span class="p">{</span><span class="n">site</span><span class="p">:</span> <span class="n">aas</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">genelength</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">wt_aas</span> <span class="o">=</span> <span class="p">[</span><span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="n">i</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">genelength</span><span class="p">)]</span>
        <span class="n">allowed_aa_muts_at_site</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">mut_str</span> <span class="ow">in</span> <span class="n">allowed_aa_muts</span><span class="p">:</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(?P&lt;wt&gt;\S)(?P&lt;site&gt;\d+)(?P&lt;mut&gt;\S)&quot;</span><span class="p">,</span> <span class="n">mut_str</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cannot match </span><span class="si">{</span><span class="n">mut_str</span><span class="si">}</span><span class="s2"> in allowed_aa_muts&quot;</span><span class="p">)</span>
            <span class="n">wt</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s2">&quot;wt&quot;</span><span class="p">)</span>
            <span class="n">site</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s2">&quot;site&quot;</span><span class="p">))</span>
            <span class="n">mut</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s2">&quot;mut&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">site</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="n">site</span> <span class="o">&gt;</span> <span class="n">genelength</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;bad site in </span><span class="si">{</span><span class="n">mut_str</span><span class="si">}</span><span class="s2"> in allowed_aa_muts&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt</span> <span class="o">!=</span> <span class="n">wt_aas</span><span class="p">[</span><span class="n">site</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;bad wt in </span><span class="si">{</span><span class="n">mut_str</span><span class="si">}</span><span class="s2"> in allowed_aa_muts&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">mut</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">aas</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;bad mut in </span><span class="si">{</span><span class="n">mut_str</span><span class="si">}</span><span class="s2"> in allowed_aa_muts&quot;</span><span class="p">)</span>
            <span class="n">allowed_aa_muts_at_site</span><span class="p">[</span><span class="n">site</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
    <span class="n">allowed_codons_at_site</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">site</span><span class="p">:</span> <span class="p">[</span>
            <span class="n">c</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">codons</span>
            <span class="k">if</span> <span class="n">codon_to_aa</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="ow">in</span> <span class="n">allowed_aa_muts_at_site</span><span class="p">[</span><span class="n">site</span><span class="p">]</span>
            <span class="ow">and</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">geneseq</span><span class="p">[</span><span class="n">site</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">-</span> <span class="mi">3</span> <span class="p">:</span> <span class="n">site</span> <span class="o">*</span> <span class="mi">3</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">allowed_aa_muts_at_site</span>
    <span class="p">}</span>
    <span class="n">allowed_codons_at_site</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">site</span><span class="p">:</span> <span class="n">codons</span> <span class="k">for</span> <span class="n">site</span><span class="p">,</span> <span class="n">codons</span> <span class="ow">in</span> <span class="n">allowed_codons_at_site</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">codons</span>
    <span class="p">}</span>
    <span class="n">allowed_sites</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">allowed_codons_at_site</span><span class="p">)</span>

    <span class="c1"># probability to draw each site, proportional to number of mutated codons</span>
    <span class="n">ps</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span> <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">allowed_codons_at_site</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
    <span class="n">ps</span> <span class="o">=</span> <span class="n">ps</span> <span class="o">/</span> <span class="n">ps</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

    <span class="c1"># for backward compatibility, we have to define the random_sample function</span>
    <span class="c1"># to be random.sample if all amino acids allowed at site</span>
    <span class="k">def</span> <span class="nf">random_sample</span><span class="p">(</span><span class="n">sites</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">p</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">allowed_aa_muts</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">sites</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">sites</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">p</span><span class="p">)</span>

    <span class="n">barcode_variant_dict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">lib</span><span class="p">,</span> <span class="n">specs_dict</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">library_specs</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="n">nvariants</span> <span class="o">=</span> <span class="n">specs_dict</span><span class="p">[</span><span class="s2">&quot;nvariants&quot;</span><span class="p">]</span>
        <span class="n">avgmuts</span> <span class="o">=</span> <span class="n">specs_dict</span><span class="p">[</span><span class="s2">&quot;avgmuts&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">nvariants</span> <span class="o">&gt;</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">NTS</span><span class="p">))</span> <span class="o">**</span> <span class="n">bclen</span><span class="p">:</span>  <span class="c1"># safety factor 10</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;barcode too short for nvariants&quot;</span><span class="p">)</span>
        <span class="n">existing_barcodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">_ivariant</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvariants</span><span class="p">):</span>
            <span class="n">barcode</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">NTS</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">bclen</span><span class="p">))</span>
            <span class="k">while</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">existing_barcodes</span><span class="p">:</span>
                <span class="n">barcode</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">NTS</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">bclen</span><span class="p">))</span>
            <span class="n">existing_barcodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">barcode</span><span class="p">)</span>

            <span class="n">support</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">random</span><span class="o">.</span><span class="n">expovariate</span><span class="p">(</span><span class="n">variant_call_support</span><span class="p">))</span>

            <span class="c1"># get mutations</span>
            <span class="n">substitutions</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">nmuts</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">allowed_sites</span><span class="p">),</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">avgmuts</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">icodon</span> <span class="ow">in</span> <span class="n">random_sample</span><span class="p">(</span><span class="n">allowed_sites</span><span class="p">,</span> <span class="n">nmuts</span><span class="p">,</span> <span class="n">ps</span><span class="p">):</span>
                <span class="n">wtcodon</span> <span class="o">=</span> <span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">icodon</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">icodon</span><span class="p">]</span>
                <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">allowed_codons_at_site</span><span class="p">[</span><span class="n">icodon</span><span class="p">])</span>
                <span class="k">assert</span> <span class="n">mutcodon</span> <span class="o">!=</span> <span class="n">wtcodon</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mutcodon</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">wtcodon</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">for</span> <span class="n">i_nt</span><span class="p">,</span> <span class="p">(</span><span class="n">wt_nt</span><span class="p">,</span> <span class="n">mut_nt</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">wtcodon</span><span class="p">,</span> <span class="n">mutcodon</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="n">wt_nt</span> <span class="o">!=</span> <span class="n">mut_nt</span><span class="p">:</span>
                        <span class="n">igene</span> <span class="o">=</span> <span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">icodon</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">i_nt</span> <span class="o">+</span> <span class="mi">1</span>
                        <span class="n">substitutions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wt_nt</span><span class="si">}{</span><span class="n">igene</span><span class="si">}{</span><span class="n">mut_nt</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">substitutions</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">substitutions</span><span class="p">)</span>

            <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s2">&quot;barcode&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">barcode</span><span class="p">)</span>
            <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s2">&quot;substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">substitutions</span><span class="p">)</span>
            <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span>
            <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s2">&quot;variant_call_support&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">support</span><span class="p">)</span>

    <span class="n">barcode_variants</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">barcode_variant_dict</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">barcode_variants</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="n">cvt</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">codonvarianttable</span><span class="o">.</span><span class="n">CodonVariantTable</span><span class="p">(</span>
            <span class="n">barcode_variant_file</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">,</span> <span class="n">allowgaps</span><span class="o">=</span><span class="n">allowgaps</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">cvt</span></div>



<div class="viewcode-block" id="simulateSampleCounts">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.simulateSampleCounts">[docs]</a>
<span class="k">def</span> <span class="nf">simulateSampleCounts</span><span class="p">(</span>
    <span class="o">*</span><span class="p">,</span>
    <span class="n">variants</span><span class="p">,</span>
    <span class="n">phenotype_func</span><span class="p">,</span>
    <span class="n">variant_error_rate</span><span class="p">,</span>
    <span class="n">pre_sample</span><span class="p">,</span>
    <span class="n">post_samples</span><span class="p">,</span>
    <span class="n">secondary_target_phenotypes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pre_sample_name</span><span class="o">=</span><span class="s2">&quot;pre-selection&quot;</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate pre- and post-selection variant counts.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">    Add to :class:`dms_variants.codonvarianttable.CodonVariantTable` using</span>
<span class="sd">    :meth:`dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts`</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    variants : class:`dms_variants.codonvarianttable.CodonVariantTable`</span>
<span class="sd">        Holds variants used in simulation.</span>
<span class="sd">    phenotype_func : function</span>
<span class="sd">        Takes a space-delimited string of amino-acid substitutions and returns</span>
<span class="sd">        a number giving enrichment of variant relative to wildtype. For</span>
<span class="sd">        instance, :meth:`SigmoidPhenotypeSimulator.observedEnrichment`.</span>
<span class="sd">        If there are multiple targets in `variants`, this function is only</span>
<span class="sd">        applied to the primary target. See `secondary_target_phenotypes` for</span>
<span class="sd">        the secondary targets.</span>
<span class="sd">    variant_error_rate : float</span>
<span class="sd">        Rate at which variants mis-called. Provide the probability that a</span>
<span class="sd">        variant has a spuriously called (or missing) codon mutation; each</span>
<span class="sd">        variant then has a random codon mutation added or removed with this</span>
<span class="sd">        probability before being passed to `phenotype_func`.</span>
<span class="sd">    pre_sample : pandas.DataFrame or dict</span>
<span class="sd">        Counts of each variant pre-selection. To specify counts, provide</span>
<span class="sd">        data frame with columns &quot;library&quot;, &quot;barcode&quot;, and &quot;count&quot;. To</span>
<span class="sd">        simulate, provide dict with keys &quot;total_count&quot; and &quot;uniformity&quot;,</span>
<span class="sd">        and for each library, we simulate pre-selection counts as a draw</span>
<span class="sd">        of &quot;total_count&quot; counts from a multinomial parameterized by pre-</span>
<span class="sd">        selection frequencies drawn from a Dirichlet distribution with</span>
<span class="sd">        concentration parameter of &quot;uniformity&quot; (5 is reasonable value).</span>
<span class="sd">    post_samples : dict</span>
<span class="sd">        Keyed by name of each sample with value another dict keyed by</span>
<span class="sd">        &#39;total_count&#39;, &#39;noise&#39;, and &#39;bottleneck&#39;. Counts drawn from</span>
<span class="sd">        multinomial parameterized by pre-selection frerquencies after</span>
<span class="sd">        passing through bottleneck of indicated size, and adding noise</span>
<span class="sd">        by mutiplying phenotype by a random variable with mean 1 and</span>
<span class="sd">        standard deviation specified by &#39;noise&#39; (0 is no noise).</span>
<span class="sd">    secondary_target_phenotypes : None or dict</span>
<span class="sd">        If there are secondary targets in `variants`, this should be keyed</span>
<span class="sd">        by each secondary target name and give the associated phenotypes</span>
<span class="sd">        as values.</span>
<span class="sd">    pre_sample_name : str</span>
<span class="sd">        Name used for the pre-selection sample.</span>
<span class="sd">    seed : None or int</span>
<span class="sd">        If not `None`, random number seed.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        Data frame with the following columns:</span>
<span class="sd">            - &quot;library&quot;</span>
<span class="sd">            - &quot;barcode&quot;</span>
<span class="sd">            - &quot;sample&quot;</span>
<span class="sd">            - &quot;count&quot;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
        <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">pre_sample_name</span> <span class="ow">in</span> <span class="n">post_samples</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`pre_sample_name` is in `post_samples`&quot;</span><span class="p">)</span>

    <span class="c1"># -----------------------------------------</span>
    <span class="c1"># internal function</span>
    <span class="k">def</span> <span class="nf">_add_variant_errors</span><span class="p">(</span><span class="n">codon_substitutions</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add errors to variant according to `variant_error_rate`.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">variant_error_rate</span><span class="p">:</span>
            <span class="n">muts</span> <span class="o">=</span> <span class="n">codon_substitutions</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mf">0.5</span><span class="p">:</span>
                <span class="c1"># add mutation</span>
                <span class="n">mutatedsites</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="nb">int</span><span class="p">(</span>
                        <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;^(</span><span class="si">{</span><span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">CODONS</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span>
                            <span class="sa">r</span><span class="s2">&quot;(?P&lt;r&gt;\d+)&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot;(</span><span class="si">{</span><span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">CODONS</span><span class="p">)</span><span class="si">}</span><span class="s2">)$&quot;</span><span class="p">,</span>
                            <span class="n">mut</span><span class="p">,</span>
                        <span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">muts</span>
                <span class="p">}</span>
                <span class="n">unmutatedsites</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">sites</span> <span class="k">if</span> <span class="n">r</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mutatedsites</span><span class="p">]</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">unmutatedsites</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;variant already has all mutations&quot;</span><span class="p">)</span>
                <span class="n">errorsite</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">unmutatedsites</span><span class="p">)</span>
                <span class="n">wtcodon</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">errorsite</span><span class="p">]</span>
                <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">CODONS</span> <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">wtcodon</span><span class="p">])</span>
                <span class="n">muts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wtcodon</span><span class="si">}{</span><span class="n">errorsite</span><span class="si">}{</span><span class="n">mutcodon</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># remove mutation</span>
                <span class="n">_</span> <span class="o">=</span> <span class="n">muts</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
                <span class="k">return</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">codon_substitutions</span>

    <span class="c1"># -----------------------------------------</span>

    <span class="k">if</span> <span class="n">variants</span><span class="o">.</span><span class="n">primary_target</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">primary_df</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">secondary_df</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">primary_df</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
            <span class="s2">&quot;target == @variants.primary_target&quot;</span>
        <span class="p">)</span>
        <span class="n">secondary_df</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
            <span class="s2">&quot;target != @variants.primary_target&quot;</span>
        <span class="p">)</span>

    <span class="n">primary_df</span> <span class="o">=</span> <span class="n">primary_df</span><span class="p">[</span>
        <span class="p">[</span><span class="s2">&quot;target&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;codon_substitutions&quot;</span><span class="p">]</span>
    <span class="p">]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">codon_substitutions</span><span class="o">=</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;codon_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">_add_variant_errors</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">aa_substitutions</span><span class="o">=</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;codon_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">dms_variants</span><span class="o">.</span><span class="n">codonvarianttable</span><span class="o">.</span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">codonToAAMuts</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">phenotype_func</span><span class="p">),</span>
    <span class="p">)[</span>
        <span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype&quot;</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="n">secondary_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">secondary_target_phenotypes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">secondary_target_phenotypes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">secondary_target_phenotypes</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="n">secondary_df</span><span class="p">[</span><span class="s2">&quot;target&quot;</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;`secondary_target_phenotypes` lacks some &quot;</span> <span class="s2">&quot;secondary targets&quot;</span>
            <span class="p">)</span>
        <span class="n">secondary_df</span> <span class="o">=</span> <span class="n">secondary_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;target&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">secondary_target_phenotypes</span><span class="p">)</span>
        <span class="p">)[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype&quot;</span><span class="p">]]</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">primary_df</span><span class="p">,</span> <span class="n">secondary_df</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="n">primary_df</span>

    <span class="n">libraries</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">libraries</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pre_sample</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="c1"># pre-sample counts specified</span>
        <span class="n">req_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">req_cols</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pre_sample</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;pre_sample lacks cols </span><span class="si">{</span><span class="n">req_cols</span><span class="si">}</span><span class="s2">:&quot;</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">pre_sample</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span>
            <span class="n">pre_sample</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
            <span class="o">!=</span> <span class="n">barcode_variant_df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;pre_sample DataFrame lacks required &quot;</span> <span class="s2">&quot;library and barcode columns&quot;</span>
            <span class="p">)</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">pre_sample</span><span class="p">[</span><span class="n">req_cols</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;count&quot;</span><span class="p">:</span> <span class="n">pre_sample_name</span><span class="p">})</span>
        <span class="c1"># &quot;true&quot; pre-selection freqs are just input counts</span>
        <span class="n">nperlib</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;library&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s2">&quot;total_count&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">nperlib</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;library&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">pre_freqs</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="n">pre_sample_name</span><span class="p">]</span> <span class="o">/</span> <span class="n">x</span><span class="o">.</span><span class="n">total_count</span><span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;total_count&quot;</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pre_sample</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="n">pre_req_keys</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;uniformity&quot;</span><span class="p">,</span> <span class="s2">&quot;total_count&quot;</span><span class="p">}</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">pre_sample</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">!=</span> <span class="n">pre_req_keys</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;pre_sample lacks required keys </span><span class="si">{</span><span class="n">pre_req_keys</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">pre_df_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libraries</span><span class="p">:</span>  <span class="c1"># noqa: B007</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;library == @lib&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">pre_freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">dirichlet</span><span class="p">(</span>
                    <span class="n">pre_sample</span><span class="p">[</span><span class="s2">&quot;uniformity&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
                <span class="p">),</span>
                <span class="n">count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span>
                    <span class="n">pre_sample</span><span class="p">[</span><span class="s2">&quot;total_count&quot;</span><span class="p">],</span> <span class="n">x</span><span class="o">.</span><span class="n">pre_freq</span>
                <span class="p">),</span>
                <span class="n">sample</span><span class="o">=</span><span class="n">pre_sample_name</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">pre_df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">pre_df_list</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pre_sample not DataFrame / dict: &quot;</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pre_sample</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="s2">&quot;pre_freq&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype&quot;</span><span class="p">]</span>
    <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">cols</span><span class="p">),</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;cols = </span><span class="si">{</span><span class="nb">set</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s2">barcode_variant_df.columns = &quot;</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">set</span><span class="p">(</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">post_samples</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;post_samples can&#39;t have key </span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">; &quot;</span> <span class="s2">&quot;choose another sample name&quot;</span>
            <span class="p">)</span>

    <span class="n">df_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="n">cols</span><span class="p">[:</span><span class="mi">4</span><span class="p">]]]</span>

    <span class="k">def</span> <span class="nf">_bottleneck_freqs</span><span class="p">(</span><span class="n">pre_freq</span><span class="p">,</span> <span class="n">bottleneck</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">bottleneck</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pre_freq</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span><span class="n">bottleneck</span><span class="p">,</span> <span class="n">pre_freq</span><span class="p">)</span> <span class="o">/</span> <span class="n">bottleneck</span>

    <span class="n">post_req_keys</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;bottleneck&quot;</span><span class="p">,</span> <span class="s2">&quot;noise&quot;</span><span class="p">,</span> <span class="s2">&quot;total_count&quot;</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">lib</span><span class="p">,</span> <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">sample_dict</span><span class="p">)</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>  <span class="c1"># noqa: B007</span>
        <span class="n">libraries</span><span class="p">,</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">post_samples</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">!=</span> <span class="n">post_req_keys</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;post_samples </span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2"> lacks </span><span class="si">{</span><span class="n">post_req_keys</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">lib_df</span> <span class="o">=</span> <span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;library == @lib&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">)</span>
        <span class="c1"># simulated pre-selection freqs after bottleneck</span>
        <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;bottleneck_freq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_bottleneck_freqs</span><span class="p">(</span>
            <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;pre_freq&quot;</span><span class="p">],</span>
            <span class="n">sample_dict</span><span class="p">[</span><span class="s2">&quot;bottleneck&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;noise&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span>
            <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">sample_dict</span><span class="p">[</span><span class="s2">&quot;noise&quot;</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lib_df</span><span class="p">))],</span>
            <span class="mi">0</span><span class="p">,</span>
            <span class="kc">None</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;post_freq_nonorm&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;bottleneck_freq&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;phenotype&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;noise&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;post_freq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;post_freq_nonorm&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;post_freq_nonorm&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span>
            <span class="n">sample_dict</span><span class="p">[</span><span class="s2">&quot;total_count&quot;</span><span class="p">],</span> <span class="n">lib_df</span><span class="p">[</span><span class="s2">&quot;post_freq&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">lib_df</span> <span class="o">=</span> <span class="n">lib_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;post_counts&quot;</span><span class="p">:</span> <span class="n">sample</span><span class="p">})</span>
        <span class="n">lib_df</span> <span class="o">=</span> <span class="n">lib_df</span><span class="p">[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;barcode&quot;</span><span class="p">,</span> <span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">]]</span>

        <span class="n">df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lib_df</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_list</span><span class="p">)</span></div>



<div class="viewcode-block" id="SigmoidPhenotypeSimulator">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator">[docs]</a>
<span class="k">class</span> <span class="nc">SigmoidPhenotypeSimulator</span><span class="p">:</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;Simulate phenotypes under sigmoid global epistasis model.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">    Mutational effects on latent phenotype are simulated to follow</span>
<span class="sd">    compound normal distribution; latent phenotype maps to observed</span>
<span class="sd">    enrichment via sigmoid, and the observed phenotype is the</span>
<span class="sd">    :math:`\log_2` of the enrichment. This distinction between latent and</span>
<span class="sd">    observed phenotype parallels the &quot;global epistasis&quot; models of</span>
<span class="sd">    `Otwinoski et al &lt;https://doi.org/10.1073/pnas.1804015115&gt;`_ and</span>
<span class="sd">    `Sailer and Harms &lt;http://www.genetics.org/content/205/3/1079&gt;`_.</span>

<span class="sd">    Specifically, let :math:`p_{\rm{latent}}` be the latent phenotype</span>
<span class="sd">    of a variant, let :math:`p_{\rm{observed}}` be the observed phenotype,</span>
<span class="sd">    and let :math:`E_{\rm{observed}}` be the observed enrichment. Also</span>
<span class="sd">    let :math:`p_{\rm{latent, wt}}` be the latent phenotype of the</span>
<span class="sd">    wildtype. Then:</span>

<span class="sd">    .. math::</span>

<span class="sd">       E_{\rm{observed}}</span>
<span class="sd">       &amp;=&amp; \frac{\left(1 + \exp\left(-p_{\rm{latent, wt}}\right)\right)</span>
<span class="sd">                 \left(1 - E_{\rm{min}}\right)}</span>
<span class="sd">                {1 + \exp\left(-p_{\rm{latent}}\right)} +</span>
<span class="sd">                E_{\rm{min}} \\</span>
<span class="sd">       p_{\rm{observed}} &amp;=&amp; \log_2 E_{\rm{observed}}</span>

<span class="sd">    where :math:`0 \le E_{\rm{min}} &lt; 1` is the minimum possible observed</span>
<span class="sd">    enrichment. Typically :math:`E_{\rm{min}} &gt; 0` in real experiments</span>
<span class="sd">    as pseudocounts are used so estimates of enrichments can&#39;t be zero.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geneseq : str</span>
<span class="sd">        Codon sequence of wild-type gene.</span>
<span class="sd">    seed : int or None</span>
<span class="sd">        Random number seed.</span>
<span class="sd">    wt_latent : float</span>
<span class="sd">        Latent phenotype of wildtype.</span>
<span class="sd">    norm_weights : list or tuple of tuples</span>
<span class="sd">        Specify compound normal distribution of mutational effects on</span>
<span class="sd">        latent phenotype as `(weight, mean, sd)` for each Gaussian.</span>
<span class="sd">    stop_effect : float</span>
<span class="sd">        Effect of stop codon at any position.</span>
<span class="sd">    min_observed_enrichment : float</span>
<span class="sd">        Minimum possible observed enrichment.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    wt_latent : float</span>
<span class="sd">        Wildtype latent phenotype.</span>
<span class="sd">    muteffects : dict</span>
<span class="sd">        Effect on latent phenotype of each amino-acid mutation.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">geneseq</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">wt_latent</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
        <span class="n">norm_weights</span><span class="o">=</span><span class="p">((</span><span class="mf">0.4</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.7</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.6</span><span class="p">,</span> <span class="o">-</span><span class="mi">7</span><span class="p">,</span> <span class="mf">3.5</span><span class="p">)),</span>
        <span class="n">stop_effect</span><span class="o">=-</span><span class="mi">15</span><span class="p">,</span>
        <span class="n">min_observed_enrichment</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;See main class docstring for how to initialize.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">wt_latent</span> <span class="o">=</span> <span class="n">wt_latent</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">min_observed_enrichment</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;not 0 &lt;= `min_observed_enrichment` &lt; 1&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">min_observed_enrichment</span> <span class="o">=</span> <span class="n">min_observed_enrichment</span>

        <span class="c1"># simulate muteffects from compound normal distribution</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
        <span class="n">weights</span><span class="p">,</span> <span class="n">means</span><span class="p">,</span> <span class="n">sds</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">norm_weights</span><span class="p">)</span>
        <span class="n">cumweights</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">icodon</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span><span class="p">):</span>
            <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="n">icodon</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">icodon</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]]</span>
            <span class="k">for</span> <span class="n">mut_aa</span> <span class="ow">in</span> <span class="n">AAS_WITHSTOP</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mut_aa</span> <span class="o">!=</span> <span class="n">wt_aa</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">mut_aa</span> <span class="o">==</span> <span class="s2">&quot;*&quot;</span><span class="p">:</span>
                        <span class="n">muteffect</span> <span class="o">=</span> <span class="n">stop_effect</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># choose Gaussian from compound normal</span>
                        <span class="n">i</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">cumweights</span> <span class="o">&lt;</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">())</span>
                        <span class="c1"># draw mutational effect from chosen Gaussian</span>
                        <span class="n">muteffect</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">gauss</span><span class="p">(</span><span class="n">means</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">sds</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wt_aa</span><span class="si">}{</span><span class="n">icodon</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}{</span><span class="n">mut_aa</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">muteffect</span>

<div class="viewcode-block" id="SigmoidPhenotypeSimulator.latentPhenotype">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator.latentPhenotype">[docs]</a>
    <span class="k">def</span> <span class="nf">latentPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Latent phenotype of a variant.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        subs : str</span>
<span class="sd">            Space-delimited list of amino-acid substitutions.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Latent phenotype of variant.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">wt_latent</span> <span class="o">+</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">subs</span><span class="o">.</span><span class="n">split</span><span class="p">())</span></div>


<div class="viewcode-block" id="SigmoidPhenotypeSimulator.observedPhenotype">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator.observedPhenotype">[docs]</a>
    <span class="k">def</span> <span class="nf">observedPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Observed phenotype of a variant.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        subs : str</span>
<span class="sd">            Space-delimited list of amino-acid substitutions.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Observed phenotype of variant.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">latentToObserved</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">(</span><span class="n">subs</span><span class="p">),</span> <span class="s2">&quot;phenotype&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="SigmoidPhenotypeSimulator.observedEnrichment">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator.observedEnrichment">[docs]</a>
    <span class="k">def</span> <span class="nf">observedEnrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Observed enrichment of a variant.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        subs : str</span>
<span class="sd">            Space-delimited list of amino-acid substitutions.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Observed enrichment relative to wildtype.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">latentToObserved</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">(</span><span class="n">subs</span><span class="p">),</span> <span class="s2">&quot;enrichment&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="SigmoidPhenotypeSimulator.latentToObserved">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator.latentToObserved">[docs]</a>
    <span class="k">def</span> <span class="nf">latentToObserved</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">latent</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;Observed enrichment or observed phenotype from latent phenotype.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        latent : float or numpy.ndarray</span>
<span class="sd">            The latent phenotype.</span>
<span class="sd">        value : {&#39;enrichment&#39;, &#39;phenotype&#39;}</span>
<span class="sd">            Get the observed enrichment or the observed phenotype?</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float or numpy.ndarray</span>
<span class="sd">            The observed enrichment or phenotype.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">enrichment</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">))</span> <span class="o">*</span> <span class="p">(</span>
            <span class="mi">1</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_observed_enrichment</span>
        <span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">latent</span><span class="p">))</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_observed_enrichment</span>
        <span class="k">if</span> <span class="n">value</span> <span class="o">==</span> <span class="s2">&quot;enrichment&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">enrichment</span>
        <span class="k">elif</span> <span class="n">value</span> <span class="o">==</span> <span class="s2">&quot;phenotype&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">enrichment</span><span class="p">)</span> <span class="o">/</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid `value` of </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="SigmoidPhenotypeSimulator.plotLatentVsObserved">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator.plotLatentVsObserved">[docs]</a>
    <span class="k">def</span> <span class="nf">plotLatentVsObserved</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">latent_min</span><span class="o">=-</span><span class="mi">15</span><span class="p">,</span> <span class="n">latent_max</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">npoints</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">wt_vline</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Plot observed enrichment/phenotype as function of latent phenotype.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        value : {&#39;enrichment&#39;, &#39;phenotype&#39;}</span>
<span class="sd">            Do we plot observed enrichment or observed phenotype?</span>
<span class="sd">        latent_min : float</span>
<span class="sd">            Smallest value of latent phenotype on plot.</span>
<span class="sd">        latent_max : float</span>
<span class="sd">            Largest value of latent phenotype on plot.</span>
<span class="sd">        npoints : int</span>
<span class="sd">            Plot a line fit to this many points.</span>
<span class="sd">        wt_vline : bool</span>
<span class="sd">            Draw a vertical line at the wildtype latent phenotype.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        plotnine.ggplot.ggplot</span>
<span class="sd">            Plot of observed enrichment or phenotype as function of latent</span>
<span class="sd">            phenotype.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">latent</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">latent_min</span><span class="p">,</span> <span class="n">latent_max</span><span class="p">,</span> <span class="n">npoints</span><span class="p">)</span>
        <span class="n">observed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">latentToObserved</span><span class="p">(</span><span class="n">latent</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">p9</span><span class="o">.</span><span class="n">ggplot</span><span class="p">(</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;latent&quot;</span><span class="p">:</span> <span class="n">latent</span><span class="p">,</span> <span class="s2">&quot;observed&quot;</span><span class="p">:</span> <span class="n">observed</span><span class="p">}),</span>
                <span class="n">p9</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="s2">&quot;latent&quot;</span><span class="p">,</span> <span class="s2">&quot;observed&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">geom_line</span><span class="p">()</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">3.5</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">))</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;latent phenotype&quot;</span><span class="p">)</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">ylab</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;observed </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">wt_vline</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">geom_vline</span><span class="p">(</span>
                <span class="n">xintercept</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">linetype</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="SigmoidPhenotypeSimulator.plotMutsHistogram">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator.plotMutsHistogram">[docs]</a>
    <span class="k">def</span> <span class="nf">plotMutsHistogram</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">mutant_order</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">wt_vline</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Plot distribution of phenotype for all mutants of a given order.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        value : {&#39;latentPhenotype&#39;, &#39;observedPhenotype&#39;, &#39;observedEnrichment&#39;}</span>
<span class="sd">            What value to plot.</span>
<span class="sd">        mutant_order : int</span>
<span class="sd">            Plot mutations of this order. Currently only works for 1</span>
<span class="sd">            (single mutants).</span>
<span class="sd">        bins : int</span>
<span class="sd">            Number of bins in histogram.</span>
<span class="sd">        wt_vline : bool</span>
<span class="sd">            Draw a vertical line at the wildtype value.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        plotnine.ggplot.ggplot</span>
<span class="sd">            Histogram of phenotype for all mutants.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutant_order</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;only implemented for `mutant_order` of 1&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">value</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observedPhenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observedEnrichment&quot;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid `value` of </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">func</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>

        <span class="n">xlist</span> <span class="o">=</span> <span class="p">[</span><span class="n">func</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">p9</span><span class="o">.</span><span class="n">ggplot</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">value</span><span class="p">:</span> <span class="n">xlist</span><span class="p">}),</span> <span class="n">p9</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">value</span><span class="p">))</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">3.5</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">))</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">ylab</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;number of </span><span class="si">{</span><span class="n">mutant_order</span><span class="si">}</span><span class="s2">-mutants&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">wt_vline</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">geom_vline</span><span class="p">(</span>
                <span class="n">xintercept</span><span class="o">=</span><span class="n">func</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">linetype</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>
</div>



<div class="viewcode-block" id="MultiLatentSigmoidPhenotypeSimulator">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.MultiLatentSigmoidPhenotypeSimulator">[docs]</a>
<span class="k">class</span> <span class="nc">MultiLatentSigmoidPhenotypeSimulator</span><span class="p">:</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;Simulate phenotype that derives from several latent phenotypes.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">    The observed phenotype is the sum of several sigmoid-transformed latent</span>
<span class="sd">    phenotypes. Specifically, let there be :math:`k = 1, \ldots, K` latent</span>
<span class="sd">    phenotypes, each of which is transformed into an observed phenotype</span>
<span class="sd">    :math:`p_{\rm{observed}}^k as described in the docs for</span>
<span class="sd">    :class:`SigmoidPhenotypeSimulator`. The overall observed phenotype is:</span>

<span class="sd">    .. math::</span>

<span class="sd">       p_{\rm{observed}} = \sum_k p_{\rm{observed}}^k</span>

<span class="sd">    and the observed enrichment is:</span>

<span class="sd">    .. math::</span>

<span class="sd">       E_{\rm{observed}} = 2^{p_{\rm{observed}}}.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sigmoid_phenotype_simulators : list</span>
<span class="sd">        A list of :class:`SigmoidPhenotypeSimulator` objects, each of which</span>
<span class="sd">        transforms one of the latent phenotypes to a component of the</span>
<span class="sd">        observed phenotype.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    n_latent_phenotypes : int</span>
<span class="sd">        The number of latent phenotypes.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sigmoid_phenotype_simulators</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;See main class docstring.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_latent_phenotypes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sigmoid_phenotype_simulators</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_latent_phenotypes</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`sigmoid_phenotype_simulators` cannot be empty&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">SigmoidPhenotypeSimulator</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">sigmoid_phenotype_simulators</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;entries in `sigmoid_phenotype_simulators` not &quot;</span>
                <span class="s2">&quot;all of SigmoidPhenotypeSimulator type&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sigmoid_phenotype_simulators</span> <span class="o">=</span> <span class="n">sigmoid_phenotype_simulators</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_all_subs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_sigmoid_phenotype_simulators</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">muteffects</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sigmoid_phenotype_simulators</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_all_subs</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">muteffects</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;entries in `sigmoid_phenotype_simulators` &quot;</span>
                    <span class="s2">&quot;do not all represent same gene sequence&quot;</span>
                <span class="p">)</span>

<div class="viewcode-block" id="MultiLatentSigmoidPhenotypeSimulator.observedEnrichment">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.MultiLatentSigmoidPhenotypeSimulator.observedEnrichment">[docs]</a>
    <span class="k">def</span> <span class="nf">observedEnrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Observed enrichment of a variant.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        subs : str</span>
<span class="sd">            Space-delimited list of amino-acid substitutions.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Observed enrichment relative to wildtype.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mi">2</span> <span class="o">**</span> <span class="bp">self</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">(</span><span class="n">subs</span><span class="p">)</span></div>


<div class="viewcode-block" id="MultiLatentSigmoidPhenotypeSimulator.observedPhenotype">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.MultiLatentSigmoidPhenotypeSimulator.observedPhenotype">[docs]</a>
    <span class="k">def</span> <span class="nf">observedPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Observed phenotype of a variant.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        subs : str</span>
<span class="sd">            Space-delimited list of amino-acid substitutions.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Observed phenotype of a variant.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span>
            <span class="n">m</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">(</span><span class="n">subs</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sigmoid_phenotype_simulators</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="MultiLatentSigmoidPhenotypeSimulator.latentPhenotype">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.MultiLatentSigmoidPhenotypeSimulator.latentPhenotype">[docs]</a>
    <span class="k">def</span> <span class="nf">latentPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subs</span><span class="p">,</span> <span class="n">k</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Latent phenotype :math:`k` of a variant.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        subs : str</span>
<span class="sd">            Space-delimited list of amino-acid substitutions.</span>
<span class="sd">        k : int</span>
<span class="sd">            Latent phenotype to get (1 &lt;= `k` &lt;=</span>
<span class="sd">            :attr:`MultiLatentSigmoidPhenotypeSimulator.n_latent_phenotypes`).</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Latent phenotype `k`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="mi">1</span> <span class="o">&lt;=</span> <span class="n">k</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_latent_phenotypes</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sigmoid_phenotype_simulators</span><span class="p">[</span><span class="n">k</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">(</span><span class="n">subs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid `k` of </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="MultiLatentSigmoidPhenotypeSimulator.plotMutsHistogram">
<a class="viewcode-back" href="../../dms_variants.simulate.html#dms_variants.simulate.MultiLatentSigmoidPhenotypeSimulator.plotMutsHistogram">[docs]</a>
    <span class="k">def</span> <span class="nf">plotMutsHistogram</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">value</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">k</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">mutant_order</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
        <span class="n">wt_vline</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Plot distribution of phenotype for all mutants of a given order.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        value : {&#39;latentPhenotype&#39;, &#39;observedPhenotype&#39;, &#39;observedEnrichment&#39;}</span>
<span class="sd">            What value to plot.</span>
<span class="sd">        k : int or None</span>
<span class="sd">            If value is `latentPhenotype, which phenotype (1 &lt;= `k` &lt;=</span>
<span class="sd">            :attr:`MultiLatentSigmoidPhenotypeSimulator.n_latent_phenotypes`)</span>
<span class="sd">            to plot.</span>
<span class="sd">        mutant_order : int</span>
<span class="sd">            Plot mutations of this order. Currently only works for 1</span>
<span class="sd">            (single mutants).</span>
<span class="sd">        bins : int</span>
<span class="sd">            Number of bins in histogram.</span>
<span class="sd">        wt_vline : bool</span>
<span class="sd">            Draw a vertical line at the wildtype value.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        plotnine.ggplot.ggplot</span>
<span class="sd">            Histogram of phenotype for all mutants.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutant_order</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;only implemented for `mutant_order` of 1&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">value</span> <span class="o">==</span> <span class="s2">&quot;latentPhenotype&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">and</span> <span class="mi">1</span> <span class="o">&lt;=</span> <span class="n">k</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_latent_phenotypes</span><span class="p">:</span>
                <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;k&quot;</span><span class="p">:</span> <span class="n">k</span><span class="p">}</span>
                <span class="n">xlabel</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;latentPhenotype </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid `k` of </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="n">value</span>

        <span class="k">if</span> <span class="n">value</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observedPhenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observedEnrichment&quot;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;invalid `value` of </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">func</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>

        <span class="n">xlist</span> <span class="o">=</span> <span class="p">[</span><span class="n">func</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_all_subs</span><span class="p">]</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">p9</span><span class="o">.</span><span class="n">ggplot</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">value</span><span class="p">:</span> <span class="n">xlist</span><span class="p">}),</span> <span class="n">p9</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">value</span><span class="p">))</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">3.5</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">))</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">ylab</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;number of </span><span class="si">{</span><span class="n">mutant_order</span><span class="si">}</span><span class="s2">-mutants&quot;</span><span class="p">)</span>
            <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">xlab</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">wt_vline</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">p9</span><span class="o">.</span><span class="n">geom_vline</span><span class="p">(</span>
                <span class="n">xintercept</span><span class="o">=</span><span class="n">func</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">linetype</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>
</div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">doctest</span>

    <span class="n">doctest</span><span class="o">.</span><span class="n">testmod</span><span class="p">()</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../../index.html">
    <img class="logo" src="../../_static/BloomLogo.jpg" alt="Logo" />
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">dms_variants documentation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../dms_variants.html">dms_variants package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2019--2024.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://github.blog/wp-content/uploads/2008/12/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>